// @sanity/runtime-cli/dist/actions/blueprints/assets.js
import crypto from 'node:crypto';
import fs from 'node:fs';
import AdmZip from 'adm-zip';
import config from '../../config.js';
import { CONVERT_BYTES_TO_MB, MAX_ASSET_SIZE } from '../../constants.js';
import { prepareAsset as prepareFunctionAsset } from '../../utils/functions/prepare-asset.js';
import getHeaders from '../../utils/get-headers.js';
import { createTracedFetch } from '../../utils/traced-fetch.js';
import { isLocalFunctionResource, isStudioResource } from '../../utils/types.js';
const { apiUrl } = config;
const ASSETS_URL = `${apiUrl}vX/blueprints/assets`;
export const ASSET_CHECK_URL = `${ASSETS_URL}/check`;
export const ASSET_STASH_URL = `${ASSETS_URL}/stash`;
export async function stashAsset({ resource, auth, logger, installer, }) {
    try {
        let b64;
        let hash;
        let extension;
        let cleanup;
        if (isLocalFunctionResource(resource)) {
            const prepResult = await prepareFunctionAsset({ resource }, { installer });
            if (prepResult.success === false) {
                return {
                    success: false,
                    error: prepResult.error,
                };
            }
            const zipResult = await pathToB64ZipHash(prepResult.outputPath);
            b64 = zipResult.b64;
            hash = zipResult.hash;
            extension = 'zip';
            cleanup = prepResult.cleanup;
        }
        else if (isStudioResource(resource)) {
            const prepResult = await prepareStudioAsset({ resource });
            if (prepResult.success === false) {
                return {
                    success: false,
                    error: prepResult.error,
                };
            }
            const buffer = await fs.promises.readFile(prepResult.outputPath);
            b64 = buffer.toString('base64');
            hash = hashBuffer(buffer);
            extension = 'tar.gz';
            cleanup = prepResult.cleanup;
        }
        else {
            return {
                success: false,
                error: `Unexpected resource for asset processing`,
            };
        }
        return await stashSingleAsset({
            b64,
            hash,
            extension,
            cleanup,
            resource,
            auth,
            logger,
        });
    }
    catch (err) {
        return { success: false, error: err instanceof Error ? err.message : `${err}` };
    }
}
async function prepareStudioAsset({ resource, }) {
    if (!fs.existsSync(resource.src)) {
        return {
            success: false,
            async cleanup() {
                // no op
            },
            error: `Property 'src' on resource with type '${resource.type}' must be a directory`,
        };
    }
    return {
        success: true,
        async cleanup() {
            // no op
        },
        // studio must be pre built (e.g. with `npm run build`) and tarred/gzipped
        outputPath: resource.src,
    };
}
async function stashSingleAsset({ b64, hash, cleanup, resource, auth, logger, extension = 'zip', }) {
    const fetchFn = createTracedFetch(logger);
    try {
        try {
            const checkResponse = await fetchFn(`${ASSET_CHECK_URL}/${hash}`, {
                method: 'GET',
                headers: getHeaders(auth),
            });
            const checkJson = await checkResponse.json();
            if (checkResponse.ok && checkJson.exists && checkJson.id) {
                return { success: true, assetId: checkJson.id, exists: true, hash };
            }
        }
        catch { } // Continue to stash the Asset
        // Asset doesn't exist, so stash it
        const assetResponse = await fetchFn(ASSET_STASH_URL, {
            method: 'POST',
            headers: getHeaders(auth),
            body: JSON.stringify({
                file: b64,
                filename: `${resource.name}.${extension}`,
            }),
        });
        const assetJson = await assetResponse.json();
        if (assetResponse.ok) {
            return { success: true, assetId: assetJson.id, exists: false, hash };
        }
        return {
            success: false,
            error: `Failed to stash asset for ${resource.name}: ${assetJson.message || 'Unknown error'}`,
        };
    }
    catch (err) {
        let error = '';
        if (err instanceof Error)
            error = err.message;
        return { success: false, error };
    }
    finally {
        await cleanup();
    }
}
export function hashBuffer(buffer) {
    const hash = crypto.createHash('sha256');
    hash.update(buffer);
    return hash.digest('hex');
}
export async function pathToZip(path) {
    const stats = await fs.promises.stat(path);
    const zip = new AdmZip(undefined, {
        fs: {
            ...fs,
            // @ts-expect-error monkey patching to return matching mtime for builds
            statSync(path, options) {
                const s = fs.statSync(path, options);
                if (s)
                    s.mtime = new Date(0);
                return s;
            },
        },
    });
    if (stats.isDirectory())
        zip.addLocalFolder(path);
    else
        zip.addLocalFile(path, '', 'index.js');
    return zip;
}
export async function pathToB64ZipHash(path) {
    const zip = await pathToZip(path);
    // Add up the uncompressed sizes of all entries to get a more accurate size of the asset being uploaded
    const size = zip
        .getEntries()
        .map((entry) => entry.header.size)
        .reduce((a, b) => a + b, 0);
    if (size > MAX_ASSET_SIZE) {
        throw new Error(`Asset size of ${(size / CONVERT_BYTES_TO_MB).toFixed(2)} MB exceeds the maximum allowed size of 200 MB.`);
    }
    const buffer = zip.toBuffer();
    const b64 = buffer.toString('base64');
    const hash = hashBuffer(buffer);
    return { b64, buffer, hash, zip };
}
