// @sanity/runtime-cli/dist/actions/blueprints/blueprint.js
import { existsSync, mkdirSync, readFileSync, statSync, writeFileSync } from 'node:fs';
import { basename, dirname, extname } from 'node:path';
import { cwd } from 'node:process';
import { pathToFileURL } from 'node:url';
import blueprintParserValidator from '@sanity/blueprints-parser';
import * as find from 'empathic/find';
import { createJiti } from 'jiti';
import { SANITY_FUNCTION_DOCUMENT, SANITY_FUNCTION_MEDIA_LIBRARY_ASSET, SANITY_FUNCTION_SCHEDULED, } from '../../constants.js';
import { validateResources } from '../../utils/validate/index.js';
import { backfillProjectBasedStackId, readConfigFile, } from './config.js';
import { resolveIds } from './resolve.js';
const SUPPORTED_FILE_EXTENSIONS = ['.json', '.js', '.mjs', '.ts'];
let SUPPORTED_FILE_NAMES = SUPPORTED_FILE_EXTENSIONS.map((ext) => `blueprint${ext}`);
SUPPORTED_FILE_NAMES = [
    ...SUPPORTED_FILE_NAMES,
    ...SUPPORTED_FILE_NAMES.map((name) => `sanity.${name}`),
];
export const JSON_BLUEPRINT_CONTENT = {
    blueprintVersion: '2024-10-01',
    resources: [],
};
export const TS_BLUEPRINT_CONTENT = `
import {defineBlueprint, defineDocumentFunction} from '@sanity/blueprints'

export default defineBlueprint({
  resources: [
    // defineDocumentFunction({name: 'my-function'}),
  ],
})
`.trim();
/**
 * Finds the blueprint file in the given path or current working directory
 * @returns The path, file name, and extension of the blueprint file
 */
export function findBlueprintFile(blueprintPath) {
    let dirToSearch = cwd();
    if (blueprintPath) {
        const pathExists = existsSync(blueprintPath);
        if (!pathExists)
            return null;
        const stat = statSync(blueprintPath);
        if (stat.isFile()) {
            return {
                blueprintFilePath: blueprintPath,
                fileName: basename(blueprintPath),
                extension: extname(blueprintPath),
            };
        }
        if (stat.isDirectory()) {
            dirToSearch = blueprintPath;
        }
    }
    const blueprintFile = find.any(SUPPORTED_FILE_NAMES, { cwd: dirToSearch });
    if (!blueprintFile)
        return null;
    return {
        blueprintFilePath: blueprintFile,
        fileName: basename(blueprintFile),
        extension: extname(blueprintFile),
    };
}
/**
 * Load a blueprint file from disk: read JSON or import+execute a JS/TS module.
 * Returns the raw blueprint data and the module reference (if dynamic).
 */
export async function loadBlueprintFile(fileInfo) {
    const { blueprintFilePath, fileName, extension } = fileInfo;
    let rawBlueprint;
    let blueprintModule;
    try {
        switch (extension) {
            case '.json': {
                const blueprintString = readFileSync(blueprintFilePath, 'utf8').toString();
                rawBlueprint = JSON.parse(blueprintString);
                break;
            }
            case '.js':
            case '.mjs': {
                const mod = await import(blueprintFilePath);
                blueprintModule = mod.default;
                break;
            }
            case '.ts': {
                const jiti = createJiti(dirname(blueprintFilePath));
                const modDefault = await jiti.import(pathToFileURL(blueprintFilePath).href, { default: true });
                blueprintModule = modDefault;
                break;
            }
            default:
                throw Error(`Unsupported Blueprint manifest extension: ${extension}`);
        }
    }
    catch (err) {
        throw Error(`Could not parse Blueprint manifest: ${fileName}\n${err}`);
    }
    // Dynamic modules: extract attached properties, then execute to get raw blueprint
    if (blueprintModule) {
        if (typeof blueprintModule === 'function') {
            try {
                rawBlueprint = blueprintModule();
            }
            catch {
                throw Error(`Error executing Blueprint manifest: ${fileName}`);
            }
        }
        else {
            throw Error(`Blueprint manifest ${fileName} must export a default function.`);
        }
    }
    if (typeof rawBlueprint === 'undefined') {
        throw Error('Could not read Blueprint manifest.');
    }
    return { rawBlueprint, module: blueprintModule };
}
/**
 * Run the blueprint parser/validator on raw blueprint data.
 * Optionally validates function resources.
 */
export function parseBlueprintContent(rawBlueprint, options) {
    const parserResult = blueprintParserValidator(rawBlueprint, {
        invalidReferenceTypes: [
            SANITY_FUNCTION_DOCUMENT,
            SANITY_FUNCTION_MEDIA_LIBRARY_ASSET,
            SANITY_FUNCTION_SCHEDULED,
        ],
    });
    const parsedBlueprint = parserResult.result === 'valid' ? parserResult.blueprint : undefined;
    const errors = parserResult.result !== 'valid' ? parserResult.errors : [];
    if (parsedBlueprint?.resources && options.validateResources) {
        errors.push(...validateResources(parsedBlueprint.resources));
    }
    // Widen from the parser's Resource type to BlueprintResource
    const parserResources = parsedBlueprint?.resources ?? [];
    const resources = parserResources.filter((r) => !!r.type);
    return {
        parsedBlueprint: parsedBlueprint || rawBlueprint,
        errors,
        resources,
    };
}
/**
 * Find, load, and parse a local blueprint file in one step.
 * Does not read the config file or resolve IDs -- callers that need
 * scope/stack information should use `readConfigFile` + `resolveIds` separately.
 *
 * @param blueprintPath Path to a blueprint file or directory containing one
 * @param options.validateResources Whether to validate function resources
 */
export async function loadAndParseBlueprint(blueprintPath, options = { validateResources: true }) {
    const fileInfo = findBlueprintFile(blueprintPath);
    if (!fileInfo) {
        throw Error('Could not find a Blueprint manifest (sanity.blueprint.ts, .js, or .json). Run `blueprints init` to create one.');
    }
    const loaded = await loadBlueprintFile(fileInfo);
    const parsed = parseBlueprintContent(loaded.rawBlueprint, options);
    return {
        fileInfo,
        rawBlueprint: loaded.rawBlueprint,
        module: loaded.module,
        ...parsed,
    };
}
/**
 * Reads the blueprint file from disk and parses it.
 * Resolves IDs from environment > blueprint module > config file.
 * Can infer stackId from projectId for legacy ST-<projectId> stacks.
 * @see {@link loadBlueprintFile} {@link readConfigFile} {@link resolveIds} {@link parseBlueprintContent}
 * @todo deprecate in favor of decomposed functions
 * @returns Known information about the Blueprint, config, and Stack
 */
export async function readLocalBlueprint(logger, validate, blueprintPath) {
    const fileInfo = findBlueprintFile(blueprintPath);
    if (!fileInfo) {
        throw Error('Could not find a Blueprint manifest (sanity.blueprint.ts, .js, or .json). Run `blueprints init` to create one.');
    }
    // 1. Load blueprint from disk
    const loaded = await loadBlueprintFile(fileInfo);
    const { rawBlueprint } = loaded;
    // 2. Read the config file
    const blueprintConfig = readConfigFile(fileInfo.blueprintFilePath);
    // 3. Resolve IDs: env > module > config
    const resolved = resolveIds({
        module: loaded.module,
        config: blueprintConfig,
    });
    // 4. Legacy stack ID inference: ST-<projectId>
    let { stackId } = resolved;
    if (!stackId && resolved.projectId) {
        try {
            stackId = await backfillProjectBasedStackId({
                blueprintFilePath: fileInfo.blueprintFilePath,
                projectId: resolved.projectId,
                logger,
            });
            if (stackId)
                resolved.sources.stackId = 'inferred';
        }
        catch {
            // assumption was wrong; leave stackId undefined
        }
    }
    // 5. Parse + validate blueprint content
    const parsed = parseBlueprintContent(rawBlueprint, {
        validateResources: validate.resources,
    });
    return {
        fileInfo,
        blueprintConfig,
        rawBlueprint,
        ...parsed,
        ...resolved,
        stackId: stackId ?? resolved.stackId,
    };
}
export function writeBlueprintToDisk({ blueprintFilePath, jsonContent = JSON_BLUEPRINT_CONTENT, }) {
    const dir = dirname(blueprintFilePath);
    const extension = extname(blueprintFilePath);
    let blueprintContent;
    switch (extension) {
        case '.json': {
            blueprintContent = JSON.stringify(jsonContent, null, 2);
            break;
        }
        case '.js':
        case '.mjs':
        case '.ts': {
            blueprintContent = TS_BLUEPRINT_CONTENT;
            break;
        }
        default: {
            throw Error(`Unsupported Blueprint manifest extension: ${extension}`);
        }
    }
    mkdirSync(dir, { recursive: true });
    writeFileSync(blueprintFilePath, blueprintContent);
    return blueprintContent;
}
export function addResourceToBlueprint({ blueprintFilePath, resource, }) {
    const blueprintFile = findBlueprintFile(blueprintFilePath);
    if (!blueprintFile)
        throw Error('Could not find a Blueprint manifest.');
    const { blueprintFilePath: foundPath, extension } = blueprintFile;
    // modify .json files directly
    if (extension === '.json') {
        const blueprintString = readFileSync(foundPath, 'utf8').toString();
        const blueprint = JSON.parse(blueprintString);
        blueprint.resources = blueprint.resources || [];
        blueprint.resources.push(resource);
        writeFileSync(foundPath, JSON.stringify(blueprint, null, 2));
        return;
    }
    return resource;
}
