// @sanity/runtime-cli/dist/actions/blueprints/resources.js
import { spawn } from 'node:child_process';
import { existsSync, mkdirSync, writeFileSync } from 'node:fs';
import { dirname, join } from 'node:path';
import { cwd } from 'node:process';
import { MAP_EVENT_TO_FUNCTION_TYPE, SANITY_FUNCTION_DOCUMENT, SANITY_FUNCTION_MEDIA_LIBRARY_ASSET, SANITY_FUNCTION_SCHEDULED, SANITY_FUNCTION_SYNC_TAG_INVALIDATE, } from '../../constants.js';
import { styleText } from '../../utils/style-text.js';
import { writeOrUpdateNodeDependency } from '../node.js';
import { addResourceToBlueprint } from './blueprint.js';
const DEFAULT_FUNCTION_TEMPLATE = /*js*/ `export async function handler({context, event}) {
  const time = new Date().toLocaleTimeString()
  console.log(\`👋 Your Sanity Function was called at \${time}\`)
}`;
const DEFAULT_HELPER_FUNCTION_TEMPLATE = /*ts*/ `import { documentEventHandler } from '@sanity/functions'

export const handler = documentEventHandler(async ({ context, event }) => {
  const time = new Date().toLocaleTimeString()
  console.log(\`👋 Your Sanity Function was called at \${time}\`)
})`;
const DEFAULT_SCHEDULED_HELPER_FUNCTION_TEMPLATE = /*ts*/ `import { scheduledEventHandler } from '@sanity/functions'

export const handler = scheduledEventHandler(async ({ context }) => {
  const time = new Date().toLocaleTimeString()
  console.log(\`Your scheduled Sanity Function was called at \${time}\`)
})`;
const DEFAULT_SYNC_TAG_HELPER_FUNCTION_TEMPLATE = /*ts*/ `import { syncTagInvalidateEventHandler } from '@sanity/functions'

export const handler = syncTagInvalidateEventHandler(async ({ context, event, done }) => {
  const time = new Date().toLocaleTimeString()
  console.log(\`Your sync tag invalidate Sanity Function was called at \${time}\`)
  // TODO: add code to do something with the invalidated sync tags provided to you in \`event.data.syncTags\`
  try {
    // notify Sanity that you have completed invalidation
    const response = await done(event.data.syncTags)
    console.log('Invalidation complete, Sanity responded with an HTTP', response.status)
  } catch (e) {
    console.error('Error invoking Sanity invalidation done endpoint!', e)
  }
})`;
/**
 * Creates a new function resource file and adds it to the blueprint
 */
export async function createFunctionResource(options, logger) {
    const { name, type, lang, blueprintFilePath, addHelpers = false, installCommand } = options;
    let workingDir = cwd();
    if (blueprintFilePath) {
        if (!existsSync(blueprintFilePath)) {
            throw Error(`Blueprint file not found: ${blueprintFilePath}`);
        }
        workingDir = dirname(blueprintFilePath);
    }
    // Ensure functions directory exists
    const functionsDir = join(workingDir, 'functions');
    if (!existsSync(functionsDir)) {
        mkdirSync(functionsDir, { recursive: true });
    }
    // Create function directory
    const functionDir = join(functionsDir, name);
    if (existsSync(functionDir))
        throw Error(`${functionDir} already exists`);
    mkdirSync(functionDir, { recursive: true });
    if (!['ts', 'js'].includes(lang))
        throw Error(`Unsupported language: ${lang}`);
    // type looks like 'document-publish', 'media-library-asset-delete' or 'scheduled-function'
    // and we are guaranteed to have the same leading words (typeName below) for all provided type strings (via guards in the call site for this method).
    const functionType = MAP_EVENT_TO_FUNCTION_TYPE[type[0]];
    // Create index.<lang> with default template
    const indexPath = join(functionDir, `index.${lang}`);
    let template = DEFAULT_FUNCTION_TEMPLATE;
    if (addHelpers) {
        switch (functionType) {
            case SANITY_FUNCTION_SCHEDULED:
                template = DEFAULT_SCHEDULED_HELPER_FUNCTION_TEMPLATE;
                break;
            case SANITY_FUNCTION_SYNC_TAG_INVALIDATE:
                template = DEFAULT_SYNC_TAG_HELPER_FUNCTION_TEMPLATE;
                break;
            default:
                template = DEFAULT_HELPER_FUNCTION_TEMPLATE;
                break;
        }
    }
    writeFileSync(indexPath, template);
    if (addHelpers && blueprintFilePath) {
        try {
            await writeOrUpdateNodeDependency(blueprintFilePath, '@sanity/functions', logger);
        }
        catch (err) {
            throw new Error('Unable to add @sanity/functions to your project.', { cause: err });
        }
    }
    if (installCommand) {
        const success = await runPackageInstall(workingDir, installCommand);
        if (!success) {
            throw new Error(`Failed to install dependencies using \`${installCommand}\``);
        }
    }
    const eventsOn = type.map((t) => t.substring(t.lastIndexOf('-') + 1));
    // Create resource definition
    let resourceJson;
    switch (functionType) {
        case SANITY_FUNCTION_DOCUMENT:
            resourceJson = {
                name,
                src: `functions/${name}`,
                type: SANITY_FUNCTION_DOCUMENT,
                event: {
                    on: eventsOn,
                },
            };
            break;
        case SANITY_FUNCTION_MEDIA_LIBRARY_ASSET:
            resourceJson = {
                name,
                src: `functions/${name}`,
                type: SANITY_FUNCTION_MEDIA_LIBRARY_ASSET,
                event: {
                    on: eventsOn,
                    resource: { type: 'media-library', id: 'my-media-library-id' },
                },
            };
            break;
        case SANITY_FUNCTION_SCHEDULED:
            resourceJson = {
                name,
                src: `functions/${name}`,
                type: SANITY_FUNCTION_SCHEDULED,
                event: {
                    // @ts-expect-error use shorthand
                    expression: '0 0 * * *',
                },
            };
            break;
        case SANITY_FUNCTION_SYNC_TAG_INVALIDATE:
            resourceJson = {
                name,
                src: `functions/${name}`,
                type: SANITY_FUNCTION_SYNC_TAG_INVALIDATE,
            };
            break;
    }
    if (!resourceJson) {
        throw new Error('Could not create function resource based on selections');
    }
    // Add to blueprint or return for manual addition
    const resource = addResourceToBlueprint({ blueprintFilePath, resource: resourceJson });
    return {
        filePath: indexPath,
        resourceAdded: !resource, // If resource is null, it was added to blueprint
        resource: resource || resourceJson,
    };
}
async function runPackageInstall(cwd, command) {
    return new Promise((resolve) => {
        const install = spawn(command, ['install'], { cwd });
        const formatOutput = (data) => {
            const lines = data.toString().split('\n');
            return lines
                .filter(Boolean)
                .map((line) => `  ${styleText('magenta', '│')} ${styleText('dim', line)}`)
                .join('\n');
        };
        install.stdout?.on('data', (data) => {
            console.log(formatOutput(data));
        });
        install.stderr?.on('data', (data) => {
            console.error(formatOutput(data));
        });
        install.on('close', (code) => {
            resolve(code === 0);
        });
    });
}
