// @sanity/runtime-cli/dist/cores/index.js
import { readLocalBlueprint } from '../actions/blueprints/blueprint.js';
import { getStack, resolveStackIdByNameOrId } from '../actions/blueprints/stacks.js';
import { presentBlueprintParserErrors } from '../utils/display/errors.js';
import { validTokenOrErrorMessage } from '../utils/validated-token.js';
export async function initBlueprintConfig({ bin, log, token, validateResources = false, validateToken = true, blueprintPath, }) {
    let checkedToken = token;
    if (!token || (token && validateToken)) {
        const tokenCheck = await validTokenOrErrorMessage(log, token);
        if (!tokenCheck.ok) {
            return { ok: false, error: tokenCheck.error.message };
        }
        checkedToken = tokenCheck.value;
    }
    if (!checkedToken) {
        return { ok: false, error: 'A valid token is required but was not provided.' };
    }
    const blueprint = await readLocalBlueprint(log, { resources: validateResources }, blueprintPath);
    if (blueprint.errors.length > 0) {
        log(presentBlueprintParserErrors(blueprint.errors));
        return { ok: false, error: 'Blueprint manifest contains errors.' };
    }
    return {
        ok: true,
        value: {
            bin,
            blueprint,
            log,
            token: checkedToken,
            validateResources,
        },
    };
}
export async function initDeployedBlueprintConfig(config) {
    if (!config.blueprint) {
        const blueprintResult = await initBlueprintConfig(config);
        if (!blueprintResult.ok)
            return blueprintResult;
        config.blueprint = blueprintResult.value.blueprint;
        config.token = blueprintResult.value.token;
    }
    const { scopeType, scopeId, stackId: blueprintStackId } = config.blueprint;
    if (!scopeType || !scopeId) {
        return { ok: false, error: 'Missing scope configuration for Blueprint' };
    }
    const auth = { token: config.token, scopeType, scopeId };
    let stackId = blueprintStackId;
    if (config.stackOverride) {
        stackId = await resolveStackIdByNameOrId(config.stackOverride, auth, config.log);
    }
    if (!stackId) {
        return { ok: false, error: 'Missing Stack deployment configuration for Blueprint' };
    }
    const spinner = config.log.ora('Loading Stack deployment...').start();
    const stackResponse = await getStack({ stackId, auth, logger: config.log });
    if (!stackResponse.ok) {
        spinner.fail('Could not load Stack deployment');
        return { ok: false, error: 'Missing Stack deployment' };
    }
    spinner.stop().clear();
    return {
        ok: true,
        value: {
            bin: config.bin,
            log: config.log,
            blueprint: config.blueprint,
            token: config.token,
            scopeType,
            scopeId,
            stackId,
            auth,
            deployedStack: stackResponse.stack,
            validateResources: config.validateResources,
        },
    };
}
