// @sanity/runtime-cli/dist/actions/sanity/access.js
import config from '../../config.js';
import getHeaders from '../../utils/get-headers.js';
import { createTracedFetch } from '../../utils/traced-fetch.js';
const { apiUrl } = config;
export const accessApiVersion = 'v2025-07-11';
export const accessApiPath = `${apiUrl}${accessApiVersion}/access`;
/** Create a robot token via the Access API. */
export async function createRobotToken({ auth, resourceType, resourceId, body, logger, }) {
    const fetchFn = createTracedFetch(logger);
    const url = `${accessApiPath}/${resourceType}/${encodeURIComponent(resourceId)}/robots`;
    const response = await fetchFn(url, {
        method: 'POST',
        headers: getHeaders(auth),
        body: JSON.stringify(body),
    });
    const raw = await response.json().catch(() => null);
    const data = (raw && typeof raw === 'object' ? raw : {});
    if (!response.ok) {
        const errMessage = data.message || data.error?.message || response.statusText;
        return { ok: false, error: errMessage, robot: null };
    }
    return { ok: true, error: null, robot: raw };
}
