// @sanity/cli/dist/actions/init/types.js
const VALID_PACKAGE_MANAGERS = new Set([
    'npm',
    'pnpm',
    'yarn'
]);
function narrowPackageManager(value) {
    return value !== undefined && VALID_PACKAGE_MANAGERS.has(value) ? value : undefined;
}
const VALID_VISIBILITIES = new Set([
    'private',
    'public'
]);
function narrowVisibility(value) {
    return value !== undefined && VALID_VISIBILITIES.has(value) ? value : undefined;
}
export function flagsToInitOptions(flags, isUnattended, args, mcpMode) {
    return {
        argType: args?.type,
        autoUpdates: flags['auto-updates'],
        bare: flags.bare,
        coupon: flags.coupon,
        dataset: flags.dataset,
        datasetDefault: flags['dataset-default'],
        env: flags.env,
        fromCreate: flags['from-create'],
        git: flags['no-git'] ? false : flags.git,
        importDataset: flags['import-dataset'],
        mcpMode,
        nextjsAddConfigFiles: flags['nextjs-add-config-files'],
        nextjsAppendEnv: flags['nextjs-append-env'],
        nextjsEmbedStudio: flags['nextjs-embed-studio'],
        organization: flags.organization,
        outputPath: flags['output-path'],
        overwriteFiles: flags['overwrite-files'],
        packageManager: narrowPackageManager(flags['package-manager']),
        project: flags.project,
        projectName: flags['project-name'] ?? flags['create-project'],
        projectPlan: flags['project-plan'],
        provider: flags.provider,
        reconfigure: flags.reconfigure,
        template: flags.template,
        templateToken: flags['template-token'],
        typescript: flags.typescript,
        unattended: isUnattended,
        visibility: narrowVisibility(flags.visibility)
    };
}
