// @sanity/cli/oclif.config.js
export default {
  bin: 'sanity',
  commands: './dist/commands',
  dirname: 'sanity',
  helpClass: './dist/SanityHelp',
  hooks: {
    command_not_found: ['./dist/hooks/commandNotFound/topicAliases.js'],
    init: ['./dist/hooks/init/checkForUpdates.js'],
    prerun: [
      './dist/hooks/prerun/injectEnvVariables.js',
      './dist/hooks/prerun/setupTelemetry.js',
      './dist/hooks/prerun/warnings.js',
    ],
  },
  plugins: ['@oclif/plugin-help', '@sanity/runtime-cli', '@sanity/migrate', '@sanity/codegen'],
  topics: {
    backups: {description: 'Manage dataset backups'},
    cors: {description: 'Manage CORS origins for your project'},
    datasets: {description: 'Manage datasets in your project'},
    docs: {description: 'Browse and search Sanity documentation'},
    documents: {description: 'Manage documents in a dataset'},
    graphql: {description: 'Manage GraphQL APIs for your project'},
    hooks: {description: 'Manage webhooks for your project'},
    manifest: {description: 'Extract studio configuration as JSON manifests'},
    mcp: {description: 'Configure Sanity MCP server for AI editors'},
    media: {description: 'Manage media assets and aspect definitions'},
    openapi: {description: 'Manage OpenAPI specifications'},
    projects: {description: 'Manage Sanity projects'},
    schemas: {description: 'Manage and validate schemas'},
    telemetry: {description: 'Manage telemetry consent'},
    tokens: {description: 'Manage API tokens for your project'},
    users: {description: 'Manage project users and invitations'},
  },
  topicSeparator: ' ',
}
