// @sanity/runtime-cli/dist/commands/blueprints/destroy.js
import { Flags } from '@oclif/core';
import { organizationIdFlagConfig, projectIdFlagConfig, ResolvedCommand } from '../../baseCommands.js';
import { blueprintDestroyCore } from '../../cores/blueprints/destroy.js';
import { Logger } from '../../utils/logger.js';
export default class DestroyCommand extends ResolvedCommand {
    static needs = ['token', 'blueprint'];
    static summary = 'Destroy a remote Stack deployment and its resources';
    static description = `Permanently removes the remote Stack and all its provisioned resources. Your local Blueprint files remain untouched, allowing you to redeploy later with 'blueprints init' + 'blueprints deploy'.

This is a destructive operation. You will be prompted to confirm unless --force is specified.

Use this to clean up test environments or decommission a Stack you no longer need.`;
    static examples = [
        '<%= config.bin %> <%= command.id %>',
        '<%= config.bin %> <%= command.id %> --stack <name-or-id> --project-id <projectId> --force --no-wait',
    ];
    static flags = {
        force: Flags.boolean({
            description: 'Force Stack destruction (skip confirmation)',
            aliases: ['f'],
            default: false,
        }),
        'project-id': Flags.string({
            ...projectIdFlagConfig,
            dependsOn: ['stack', 'force'],
        }),
        'organization-id': Flags.string({
            ...organizationIdFlagConfig,
            dependsOn: ['stack', 'force'],
        }),
        stack: Flags.string({
            description: 'Stack name or ID to destroy (defaults to the locally configured Stack)',
            aliases: ['stack-id', 'stackId'],
        }),
        'no-wait': Flags.boolean({
            description: 'Do not wait for Stack destruction to complete',
            default: false,
        }),
    };
    async run() {
        const result = await blueprintDestroyCore({
            bin: this.config.bin,
            log: Logger(this.log.bind(this), this.flags),
            token: this.sanityToken,
            blueprint: this.blueprint,
            validateResources: this.flags['validate-resources'],
            flags: this.flags,
        });
        if (!result.success)
            this.coreError(result);
        return result.json;
    }
}
