// @sanity/runtime-cli/dist/commands/blueprints/info.js
import { Flags } from '@oclif/core';
import { organizationIdFlagConfig, projectIdFlagConfig, ResolvedCommand } from '../../baseCommands.js';
import { blueprintInfoCore } from '../../cores/blueprints/info.js';
import { Logger } from '../../utils/logger.js';
export default class InfoCommand extends ResolvedCommand {
    static needs = ['deployedStack'];
    static summary = 'Display the status and resources of the remote Stack deployment';
    static description = `Displays the current state and metadata of your remote Stack deployment, including deployed resources, status, and configuration.

Use this command to verify a deployment succeeded, check what resources are live, or confirm which Stack your local Blueprint is connected to.

Run 'blueprints stacks' to see all available Stacks in your project or organization.`;
    static examples = [
        '<%= config.bin %> <%= command.id %>',
        '<%= config.bin %> <%= command.id %> --stack <name-or-id>',
        '<%= config.bin %> <%= command.id %> --project-id <id> --stack <name-or-id>',
        '<%= config.bin %> <%= command.id %> --organization-id <orgId> --stack <name-or-id>',
    ];
    static flags = {
        stack: Flags.string({
            description: 'Stack name or ID',
            aliases: ['id'],
        }),
        'project-id': Flags.string({ ...projectIdFlagConfig }),
        'organization-id': Flags.string({ ...organizationIdFlagConfig }),
    };
    async run() {
        const result = await blueprintInfoCore({
            bin: this.config.bin,
            log: Logger(this.log.bind(this), this.flags),
            stackId: this.stackId,
            deployedStack: this.deployedStack,
            validateResources: this.flags['validate-resources'],
            flags: this.flags,
        });
        if (!result.success)
            this.coreError(result);
        return result.json;
    }
}
