// @sanity/runtime-cli/dist/commands/blueprints/logs.js
import { Flags } from '@oclif/core';
import { organizationIdFlagConfig, projectIdFlagConfig, ResolvedCommand, stackFlagConfig, } from '../../baseCommands.js';
import { blueprintLogsCore } from '../../cores/blueprints/logs.js';
import { Logger } from '../../utils/logger.js';
export default class LogsCommand extends ResolvedCommand {
    static needs = ['deployedStack'];
    static summary = "Display logs for the current Blueprint's Stack deployment";
    static description = `Retrieves Stack deployment logs, useful for debugging and monitoring deployment activity.

Use --watch (-w) to tail logs in real-time.

Use --limit, --since, or --before to narrow the result set when not watching.

If you're not seeing expected logs, verify your Stack is deployed with 'blueprints info'.`;
    static examples = [
        '<%= config.bin %> <%= command.id %>',
        '<%= config.bin %> <%= command.id %> --watch',
        '<%= config.bin %> <%= command.id %> --stack <name-or-id>',
        '<%= config.bin %> <%= command.id %> --limit 500',
        '<%= config.bin %> <%= command.id %> --since 2026-05-01T00:00:00Z',
        '<%= config.bin %> <%= command.id %> --before 2026-05-01T00:00:00Z',
    ];
    static flags = {
        stack: Flags.string({ ...stackFlagConfig }),
        'project-id': Flags.string({ ...projectIdFlagConfig }),
        'organization-id': Flags.string({ ...organizationIdFlagConfig }),
        watch: Flags.boolean({
            char: 'w',
            description: 'Watch for new Stack logs',
            aliases: ['follow'],
        }),
        limit: Flags.integer({
            char: 'l',
            description: 'Maximum number of log entries to retrieve (1-500)',
            min: 1,
            max: 500,
            exclusive: ['watch'],
        }),
        since: Flags.string({
            description: 'Only show logs after this ISO 8601 timestamp',
            exclusive: ['watch'],
        }),
        before: Flags.string({
            description: 'Only show logs before this ISO 8601 timestamp',
            exclusive: ['watch'],
        }),
    };
    async run() {
        const result = await blueprintLogsCore({
            bin: this.config.bin,
            log: Logger(this.log.bind(this), this.flags),
            auth: this.auth,
            stackId: this.stackId,
            deployedStack: this.deployedStack,
            validateResources: this.flags['validate-resources'],
            flags: this.flags,
        });
        if (result.streaming)
            return result.streaming;
        if (!result.success)
            this.coreError(result);
        return result.json;
    }
}
