// @sanity/runtime-cli/dist/commands/blueprints/stacks.js
import { Flags } from '@oclif/core';
import { organizationIdFlagConfig, projectIdFlagConfig, ResolvedCommand } from '../../baseCommands.js';
import { blueprintStacksCore } from '../../cores/blueprints/stacks.js';
import { Logger } from '../../utils/logger.js';
export default class StacksCommand extends ResolvedCommand {
    static needs = ['token', 'blueprint'];
    static summary = 'List remote Stack deployments for your project or organization';
    static description = `Shows all Stacks associated with a project or organization. By default, lists Stacks scoped to the local Blueprint.

Use this to discover existing Stacks you can scope a local Blueprint to (using 'blueprints config --edit'), or to audit what's deployed across your project.

Use --include-projects with --organization-id to also list Stacks from all projects within the organization.`;
    static examples = [
        '<%= config.bin %> <%= command.id %>',
        '<%= config.bin %> <%= command.id %> --project-id <projectId>',
        '<%= config.bin %> <%= command.id %> --organization-id <organizationId>',
        '<%= config.bin %> <%= command.id %> --organization-id <organizationId> --include-projects',
    ];
    static flags = {
        'project-id': Flags.string({
            ...projectIdFlagConfig,
            exclusive: ['organization-id', 'include-projects'],
        }),
        'organization-id': Flags.string({ ...organizationIdFlagConfig }),
        'include-projects': Flags.boolean({
            description: 'Include Stacks from all projects within the organization. Requires --organization-id.',
            dependsOn: ['organization-id'],
        }),
    };
    async run() {
        const result = await blueprintStacksCore({
            bin: this.config.bin,
            log: Logger(this.log.bind(this), this.flags),
            token: this.sanityToken,
            blueprint: this.blueprint,
            validateResources: this.flags['validate-resources'],
            flags: this.flags,
        });
        if (!result.success)
            this.coreError(result);
        return result.json;
    }
}
