// @sanity/sdk/dist/index.js
import { switchMap, from, firstValueFrom, EMPTY, asapScheduler, distinctUntilChanged, map as map$1, combineLatest, of, concatMap, withLatestFrom, filter as filter$1, concat, timer, throwError, first as first$1, Subject, takeUntil, share, partition, merge, shareReplay, tap as tap$1, catchError as catchError$1, startWith as startWith$1, pairwise as pairwise$1, groupBy as groupBy$1, mergeMap as mergeMap$1, throttle, race, skip, Observable, NEVER, fromEvent, Subscription, debounceTime, defer } from "rxjs";
import { createLogger, pickProperties, insecureRandomId, getClientState, bindActionGlobally, createStateSourceAction, setCleanupTimeout, omitProperty, defineStore, authStore, AuthStateType, getCleanedUrl, getTokenFromLocation, createLoggedInAuthState, getAuthCode, REQUEST_TAG_PREFIX, DEFAULT_API_VERSION, getDefaultLocation, isDeepEqual, configureLogging as configureLogging$1, isReleasePerspective, bindActionByResource, isDatasetResource, isMediaLibraryResource, isCanvasResource, getUsersKey, addSubscription, parseUsersKey, getClient, PROJECT_API_VERSION, setUsersError, setUsersData, API_VERSION as API_VERSION$8, getDashboardOrganizationId as getDashboardOrganizationId$1, USERS_STATE_CLEAR_DELAY, removeSubscription, updateLastLoadMoreRequest, cancelRequest, initializeRequest, getTokenState, getQueryState, resolveQuery, bindActionByResourceAndPerspective, PREVIEW_PROJECTION, transformProjectionToPreview } from "./_chunks-es/createGroqSearchFilter.js";
import { createGroqSearchFilter, getActiveReleasesState, getAuthState, getClientErrorApiBody, getClientErrorApiDescription, getClientErrorApiType, getCurrentUserState, getIsInDashboardState, getLoginUrlState, getPerspectiveState, getQueryKey, isCanvasSource, isDatasetSource, isMediaLibrarySource, isProjectUserNotFoundClientError, isStudioConfig, parseQueryKey, setAuthToken } from "./_chunks-es/createGroqSearchFilter.js";
import { first, switchMap as switchMap$1, groupBy, mergeMap, startWith, pairwise, filter, map, delay, tap, catchError, scan, share as share$1 } from "rxjs/operators";
import { createController, createNode } from "@sanity/comlink";
import { createSelector } from "reselect";
import { SanityEncoder } from "@sanity/mutate";
import { getVersionId, DocumentId, getPublishedId, getDraftId, isDraftId, isVersionId, isPublishedId } from "@sanity/id-utils";
import { jsonMatch, stringifyPath, slicePath, getIndexForKey } from "@sanity/json-match";
import { getIndexForKey as getIndexForKey2, getPathDepth, joinPaths, jsonMatch as jsonMatch2, slicePath as slicePath2, stringifyPath as stringifyPath2 } from "@sanity/json-match";
import { evaluateSync, parse } from "groq-js";
import { diffValue } from "@sanity/diff-patch";
import { applyPatches, parsePatch } from "@sanity/diff-match-patch";
import { isKeySegment, isKeyedObject } from "@sanity/types";
import { createDocumentLoaderFromClient } from "@sanity/mutate/_unstable_store";
import { SDK_CHANNEL_NAME, SDK_NODE_NAME } from "@sanity/message-protocol";
import { fromUrl } from "@sanity/bifur-client";
import { CorsOriginError } from "@sanity/client";
import { CORE_SDK_VERSION } from "./_chunks-es/version.js";
function createSanityInstance(config = {}) {
  const instanceId = crypto.randomUUID(), disposeListeners = /* @__PURE__ */ new Map(), disposed = { current: !1 }, instanceContext = {
    instanceId,
    projectId: config.projectId,
    dataset: config.dataset
  }, logger = createLogger("sdk", { instanceContext });
  logger.info("Sanity instance created", {
    hasProjectId: !!config.projectId,
    hasDataset: !!config.dataset,
    hasAuth: !!config.auth,
    hasPerspective: !!config.perspective
  }), logger.debug("Instance configuration", {
    projectId: config.projectId,
    dataset: config.dataset,
    perspective: config.perspective,
    hasStudioConfig: !!config.studio,
    hasStudioTokenSource: !!config.studio?.auth?.token,
    legacyStudioMode: config.studioMode?.enabled,
    hasAuthProviders: !!config.auth?.providers,
    hasAuthToken: !!config.auth?.token
  });
  const instance = {
    instanceId,
    config,
    isDisposed: () => disposed.current,
    dispose: () => {
      if (disposed.current) {
        logger.trace("Dispose called on already disposed instance", { internal: !0 });
        return;
      }
      logger.trace("Disposing instance", {
        internal: !0,
        listenerCount: disposeListeners.size
      }), disposed.current = !0, disposeListeners.forEach((listener) => listener()), disposeListeners.clear(), logger.info("Instance disposed");
    },
    onDispose: (cb) => {
      const listenerId = insecureRandomId();
      return disposeListeners.set(listenerId, cb), () => {
        disposeListeners.delete(listenerId);
      };
    },
    getParent: () => {
    },
    createChild: (next) => {
      logger.debug("Creating child instance", {
        parentInstanceId: instanceId.slice(0, 8),
        overridingProjectId: !!next.projectId,
        overridingDataset: !!next.dataset,
        overridingAuth: !!next.auth
      });
      const child = Object.assign(
        createSanityInstance({
          ...config,
          ...next,
          ...config.auth === next.auth ? config.auth : config.auth && next.auth && { auth: { ...config.auth, ...next.auth } }
        }),
        { getParent: () => instance }
      );
      return logger.trace("Child instance created", {
        internal: !0,
        childInstanceId: child.instanceId.slice(0, 8)
      }), child;
    },
    match: (targetConfig) => {
      if (Object.entries(pickProperties(targetConfig, ["auth", "projectId", "dataset"])).every(
        ([key, value]) => config[key] === value
      ))
        return instance;
      const parent = instance.getParent();
      if (parent) return parent.match(targetConfig);
    }
  };
  return instance;
}
const API_VERSION$7 = "vX";
function agentGenerate(instance, options, resource) {
  return getClientState(instance, { apiVersion: API_VERSION$7, resource }).observable.pipe(
    switchMap((client) => client.observable.agent.action.generate(options))
  );
}
function agentTransform(instance, options, resource) {
  return getClientState(instance, { apiVersion: API_VERSION$7, resource }).observable.pipe(
    switchMap((client) => client.observable.agent.action.transform(options))
  );
}
function agentTranslate(instance, options, resource) {
  return getClientState(instance, { apiVersion: API_VERSION$7, resource }).observable.pipe(
    switchMap((client) => client.observable.agent.action.translate(options))
  );
}
function agentPrompt(instance, options, resource) {
  return getClientState(instance, { apiVersion: API_VERSION$7, resource }).observable.pipe(
    switchMap((client) => from(client.agent.action.prompt(options)))
  );
}
function agentPatch(instance, options, resource) {
  return getClientState(instance, { apiVersion: API_VERSION$7, resource }).observable.pipe(
    switchMap((client) => from(client.agent.action.patch(options)))
  );
}
function compareProjectOrganization(projectId, projectOrganizationId, currentDashboardOrgId) {
  return projectOrganizationId !== currentDashboardOrgId ? {
    error: `Project ${projectId} belongs to Organization ${projectOrganizationId ?? "unknown"}, but the Dashboard has Organization ${currentDashboardOrgId} selected`
  } : { error: null };
}
function createFetcherStore({
  name,
  fetcher: getObservable,
  getKey,
  fetchThrottleInternal = 1e3,
  stateExpirationDelay = 5e3
}) {
  const store = defineStore({
    name,
    getInitialState: () => ({
      stateByParams: {}
    }),
    initialize: (context) => {
      const subscription = subscribeToSubscriptionsAndFetch(context);
      return () => subscription.unsubscribe();
    }
  }), subscribeToSubscriptionsAndFetch = ({
    state
  }) => state.observable.pipe(
    // Map the state to an array of [serialized, entry] pairs.
    switchMap$1((s) => {
      const entries = Object.entries(s.stateByParams);
      return entries.length > 0 ? from(entries) : EMPTY;
    }),
    // Group by the serialized key.
    groupBy(([key]) => key),
    mergeMap(
      (group$) => group$.pipe(
        // Emit an initial value for pairwise comparisons.
        startWith([group$.key, void 0]),
        pairwise(),
        // Trigger only when the subscriptions array grows.
        filter(([[, prevEntry], [, currEntry]]) => {
          const prevSubs = prevEntry?.subscriptions ?? [];
          return (currEntry?.subscriptions ?? []).length > prevSubs.length;
        }),
        map(([, [, currEntry]]) => currEntry),
        // Only trigger if we haven't fetched recently.
        filter((entry) => {
          const lastFetch = entry?.lastFetchInitiatedAt;
          return lastFetch ? Date.now() - new Date(lastFetch).getTime() >= fetchThrottleInternal : !0;
        }),
        switchMap$1((entry) => entry ? (state.set("setLastFetchInitiatedAt", (prev) => ({
          stateByParams: {
            ...prev.stateByParams,
            [entry.key]: {
              ...entry,
              ...prev.stateByParams[entry.key],
              lastFetchInitiatedAt: (/* @__PURE__ */ new Date()).toISOString()
            }
          }
        })), getObservable(entry.instance)(...entry.params).pipe(
          // the `createStateSourceAction` util requires the update
          // to
          delay(0, asapScheduler),
          tap(
            (data) => state.set("setData", (prev) => ({
              stateByParams: {
                ...prev.stateByParams,
                [entry.key]: {
                  ...omitProperty(entry, "error"),
                  ...prev.stateByParams[entry.key] ? omitProperty(prev.stateByParams[entry.key], "error") : {},
                  data
                }
              }
            }))
          ),
          catchError((error) => (state.set("setError", (prev) => ({
            stateByParams: {
              ...prev.stateByParams,
              [entry.key]: {
                ...entry,
                ...prev.stateByParams[entry.key],
                error
              }
            }
          })), EMPTY))
        )) : EMPTY)
      )
    )
  ).subscribe({
    error: (error) => state.set("setError", { error })
  }), getState = bindActionGlobally(
    store,
    createStateSourceAction({
      selector: ({
        instance,
        state: { stateByParams, error }
      }, ...params) => {
        if (error) throw error;
        const key = getKey(instance, ...params), entry = stateByParams[key];
        if (entry?.error) throw entry.error;
        return entry?.data;
      },
      onSubscribe: ({ instance, state }, ...params) => {
        const subscriptionId = insecureRandomId(), key = getKey(instance, ...params);
        return state.set("addSubscription", (prev) => ({
          stateByParams: {
            ...prev.stateByParams,
            [key]: {
              ...prev.stateByParams[key],
              instance,
              key,
              params: prev.stateByParams[key]?.params || params,
              subscriptions: [...prev.stateByParams[key]?.subscriptions || [], subscriptionId]
            }
          }
        })), () => {
          setCleanupTimeout(() => {
            state.set("removeSubscription", (prev) => {
              const entry = prev.stateByParams[key];
              if (!entry) return prev;
              const newSubs = (entry.subscriptions || []).filter((id) => id !== subscriptionId);
              return newSubs.length === 0 ? { stateByParams: omitProperty(prev.stateByParams, key) } : {
                stateByParams: {
                  ...prev.stateByParams,
                  [key]: {
                    ...entry,
                    subscriptions: newSubs
                  }
                }
              };
            });
          }, stateExpirationDelay);
        };
      }
    })
  ), resolveState = bindActionGlobally(
    store,
    ({ instance }, ...params) => firstValueFrom(getState(instance, ...params).observable.pipe(first((i) => i !== void 0)))
  );
  return { getState, resolveState };
}
const API_VERSION$6 = "v2025-02-19";
function normalizeProjectOptions(options) {
  return {
    includeMembers: options?.includeMembers ?? !0,
    includeFeatures: options?.includeFeatures ?? !0
  };
}
function resolveProjectId(instance, options) {
  const projectId = options?.projectId ?? instance.config.projectId;
  if (!projectId)
    throw new Error("A projectId is required to use the project API.");
  return projectId;
}
function getProjectCacheKey(instance, options) {
  const projectId = resolveProjectId(instance, options), { includeMembers, includeFeatures } = normalizeProjectOptions(options);
  return `project:${projectId}${includeMembers ? ":members" : ""}${includeFeatures ? ":features" : ""}`;
}
const project = createFetcherStore({
  name: "Project",
  getKey: getProjectCacheKey,
  fetcher: (instance) => (options) => {
    const projectId = resolveProjectId(instance, options);
    return getClientState(instance, {
      apiVersion: API_VERSION$6,
      scope: "global"
    }).observable.pipe(
      switchMap((client) => {
        const normalized = normalizeProjectOptions(options), query = Object.fromEntries(
          Object.entries(normalized).filter(([, value]) => value !== void 0).map(([key, value]) => [key, String(value)])
        );
        return client.observable.request({
          uri: `/projects/${projectId}`,
          query,
          tag: "project.get"
        });
      })
    );
  }
}), getProjectState = project.getState, resolveProject = project.resolveState, getDashboardOrganizationId = bindActionGlobally(
  authStore,
  createStateSourceAction(({ state: { dashboardContext } }) => dashboardContext?.orgId)
);
function observeOrganizationVerificationState(instance, projectIds) {
  const dashboardOrgId$ = getDashboardOrganizationId(instance).observable.pipe(distinctUntilChanged()), projectOrgIdObservables = projectIds.map(
    (id) => getProjectState(instance, { projectId: id }).observable.pipe(
      map$1((project2) => ({ projectId: id, orgId: project2?.organizationId ?? null })),
      // Ensure we only proceed if the orgId is loaded, distinct prevents unnecessary checks
      distinctUntilChanged((prev, curr) => prev.orgId === curr.orgId)
    )
  ), allProjectOrgIds$ = projectOrgIdObservables.length > 0 ? combineLatest(projectOrgIdObservables) : of([]);
  return combineLatest([dashboardOrgId$, allProjectOrgIds$]).pipe(
    switchMap(([dashboardOrgId, projectOrgDataArray]) => {
      if (!dashboardOrgId || projectOrgDataArray.length === 0)
        return of({ error: null });
      for (const projectData of projectOrgDataArray) {
        if (!projectData.orgId)
          continue;
        const result = compareProjectOrganization(
          projectData.projectId,
          projectData.orgId,
          dashboardOrgId
        );
        if (result.error)
          return of(result);
      }
      return of({ error: null });
    }),
    // Only emit when the overall error status actually changes
    distinctUntilChanged((prev, curr) => prev.error === curr.error)
  );
}
const handleAuthCallback = bindActionGlobally(
  authStore,
  async ({ state }, locationHref = getDefaultLocation()) => {
    const { providedToken, callbackUrl, clientFactory, apiHost, storageArea, storageKey } = state.get().options;
    if (providedToken) return !1;
    const { authState } = state.get();
    if (authState.type === AuthStateType.LOGGING_IN && authState.isExchangingToken) return !1;
    const cleanedUrl = getCleanedUrl(locationHref), tokenFromUrl = getTokenFromLocation(locationHref);
    if (tokenFromUrl)
      return state.set("setTokenFromUrl", {
        authState: createLoggedInAuthState(tokenFromUrl, null)
      }), cleanedUrl;
    const authCode = getAuthCode(callbackUrl, locationHref);
    if (!authCode) return !1;
    const parsedUrl = new URL(locationHref);
    let dashboardContext = {};
    try {
      const contextParam = parsedUrl.searchParams.get("_context");
      if (contextParam) {
        const parsedContext = JSON.parse(contextParam);
        parsedContext && typeof parsedContext == "object" && (delete parsedContext.sid, dashboardContext = parsedContext);
      }
    } catch (err) {
      console.error("Failed to parse dashboard context:", err);
    }
    state.set("exchangeSessionForToken", {
      authState: { type: AuthStateType.LOGGING_IN, isExchangingToken: !0 },
      dashboardContext
    });
    try {
      const client = clientFactory({
        apiVersion: DEFAULT_API_VERSION,
        requestTagPrefix: REQUEST_TAG_PREFIX,
        useProjectHostname: !1,
        useCdn: !1,
        ...apiHost && { apiHost }
      }), { token } = await client.request({
        method: "GET",
        uri: "/auth/fetch",
        query: { sid: authCode },
        tag: "fetch-token"
      });
      return storageArea?.setItem(storageKey, JSON.stringify({ token })), state.set("setToken", { authState: createLoggedInAuthState(token, null) }), cleanedUrl;
    } catch (error) {
      return state.set("exchangeSessionForTokenError", { authState: { type: AuthStateType.ERROR, error } }), cleanedUrl;
    }
  }
), logout = bindActionGlobally(authStore, async ({ state }) => {
  const { clientFactory, apiHost, providedToken, storageArea, storageKey } = state.get().options;
  if (providedToken) return;
  const { authState } = state.get();
  if (authState.type === AuthStateType.LOGGED_OUT && authState.isDestroyingSession) return;
  const token = authState.type === AuthStateType.LOGGED_IN && authState.token;
  try {
    token && (state.set("loggingOut", {
      authState: { type: AuthStateType.LOGGED_OUT, isDestroyingSession: !0 }
    }), await clientFactory({
      token,
      requestTagPrefix: REQUEST_TAG_PREFIX,
      apiVersion: DEFAULT_API_VERSION,
      ...apiHost && { apiHost },
      useProjectHostname: !1,
      useCdn: !1
    }).request({ uri: "/auth/logout", method: "POST", tag: "logout" }));
  } finally {
    state.set("logoutSuccess", {
      authState: { type: AuthStateType.LOGGED_OUT, isDestroyingSession: !1 }
    }), storageArea?.removeItem(storageKey), storageArea?.removeItem(`${storageKey}_last_refresh`);
  }
}), destroyController$1 = ({ state }) => {
  const { controller } = state.get();
  controller && (controller.destroy(), state.set("destroyController", {
    controller: null,
    channels: /* @__PURE__ */ new Map()
  }));
}, getOrCreateChannel$1 = ({ state }, options) => {
  const controller = state.get().controller;
  if (!controller)
    throw new Error("Controller must be initialized before using or creating channels");
  const channels = state.get().channels, existing = channels.get(options.name);
  if (existing) {
    if (!isDeepEqual(existing.options, options))
      throw new Error(`Channel "${options.name}" already exists with different options`);
    return state.set("incrementChannelRefCount", {
      channels: new Map(channels).set(options.name, {
        ...existing,
        refCount: existing.refCount + 1
      })
    }), existing.channel.start(), existing.channel;
  }
  const channel = controller.createChannel(options);
  return channel.start(), state.set("createChannel", {
    channels: new Map(channels).set(options.name, {
      channel,
      options,
      refCount: 1
    })
  }), channel;
}, getOrCreateController$1 = ({ state, instance }, targetOrigin) => {
  const { controller, controllerOrigin } = state.get();
  if (controller && controllerOrigin === targetOrigin)
    return controller;
  controller && destroyController$1({ state });
  const newController = createController({ targetOrigin });
  return state.set("initializeController", {
    controllerOrigin: targetOrigin,
    controller: newController
  }), newController;
}, releaseChannel$1 = ({ state }, name) => {
  const channels = state.get().channels, channelEntry = channels.get(name);
  if (channelEntry) {
    const newRefCount = channelEntry.refCount === 0 ? 0 : channelEntry.refCount - 1;
    newRefCount === 0 ? (channelEntry.channel.stop(), channels.delete(name), state.set("releaseChannel", { channels: new Map(channels) })) : state.set("releaseChannel", {
      channels: new Map(channels).set(name, {
        ...channelEntry,
        refCount: newRefCount
      })
    });
  }
}, comlinkControllerStore = defineStore({
  name: "connectionStore",
  getInitialState: () => ({
    controller: null,
    controllerOrigin: null,
    channels: /* @__PURE__ */ new Map()
  }),
  initialize({ instance }) {
    return () => {
      destroyController(instance);
    };
  }
}), destroyController = bindActionGlobally(
  comlinkControllerStore,
  destroyController$1
), getOrCreateChannel = bindActionGlobally(
  comlinkControllerStore,
  getOrCreateChannel$1
), getOrCreateController = bindActionGlobally(
  comlinkControllerStore,
  getOrCreateController$1
), releaseChannel = bindActionGlobally(comlinkControllerStore, releaseChannel$1), getOrCreateNode$1 = ({ state }, options) => {
  const nodes = state.get().nodes, existing = nodes.get(options.name);
  if (existing) {
    if (!isDeepEqual(existing.options, options))
      throw new Error(`Node "${options.name}" already exists with different options`);
    return existing.node.start(), existing.node;
  }
  const node = createNode(options);
  node.start();
  const statusUnsub = node.onStatus((status) => {
    const currentNodes = state.get().nodes, currentEntry = currentNodes.get(options.name);
    if (!currentEntry) return;
    const updatedEntry = {
      ...currentEntry,
      status
    };
    state.set("updateNodeStatus", {
      nodes: new Map(currentNodes).set(options.name, updatedEntry)
    });
  }), entry = {
    node,
    options,
    status: "idle",
    statusUnsub
  };
  return nodes.set(options.name, entry), state.set("createNode", { nodes }), node;
}, releaseNode$1 = ({ state }, name) => {
  const nodes = state.get().nodes, existing = nodes.get(name);
  if (existing) {
    existing.statusUnsub && existing.statusUnsub(), existing.node.stop(), nodes.delete(name), state.set("removeNode", { nodes });
    return;
  }
}, comlinkNodeStore = defineStore({
  name: "nodeStore",
  getInitialState: () => ({
    nodes: /* @__PURE__ */ new Map(),
    subscriptions: /* @__PURE__ */ new Map()
  }),
  initialize({ state }) {
    return () => {
      state.get().nodes.forEach(({ node }) => {
        node.stop();
      });
    };
  }
}), releaseNode = bindActionGlobally(comlinkNodeStore, releaseNode$1), getOrCreateNode = bindActionGlobally(comlinkNodeStore, getOrCreateNode$1), NODE_RELEASE_TIME = 5e3, selectNode = (context, nodeInput) => context.state.nodes.get(nodeInput.name), getNodeState = bindActionGlobally(
  comlinkNodeStore,
  createStateSourceAction({
    selector: createSelector([selectNode], (nodeEntry) => nodeEntry?.status === "connected" ? {
      node: nodeEntry.node,
      status: nodeEntry.status
    } : void 0),
    onSubscribe: ({ state, instance }, nodeInput) => {
      const nodeName = nodeInput.name, subscriberId = Symbol("comlink-node-subscriber");
      getOrCreateNode(instance, nodeInput);
      let subs = state.get().subscriptions.get(nodeName);
      return subs || (subs = /* @__PURE__ */ new Set(), state.get().subscriptions.set(nodeName, subs)), subs.add(subscriberId), () => {
        setCleanupTimeout(() => {
          const activeSubs = state.get().subscriptions.get(nodeName);
          activeSubs && (activeSubs.delete(subscriberId), activeSubs.size === 0 && (state.get().subscriptions.delete(nodeName), releaseNode(instance, nodeName)));
        }, NODE_RELEASE_TIME);
      };
    }
  })
);
function createDocumentHandle(handle) {
  return handle;
}
function createDocumentTypeHandle(handle) {
  return handle;
}
function createProjectHandle(handle) {
  return handle;
}
function createDatasetHandle(handle) {
  return handle;
}
function configureLogging(config) {
  configureLogging$1(config);
  const configLevel = config.level || "warn", shouldLog = ["info", "debug", "trace"].includes(configLevel) || configLevel === "warn";
  shouldLog && config.handler?.info ? config.handler.info(`[${(/* @__PURE__ */ new Date()).toISOString()}] [INFO] [sdk] Logging configured`, {
    level: configLevel,
    namespaces: config.namespaces || [],
    internal: config.internal || !1,
    source: "programmatic"
  }) : shouldLog && console.info(`[${(/* @__PURE__ */ new Date()).toISOString()}] [INFO] [sdk] Logging configured`, {
    level: configLevel,
    namespaces: config.namespaces || [],
    internal: config.internal || !1,
    source: "programmatic"
  });
}
const API_VERSION$5 = "v2025-02-19", datasets = createFetcherStore({
  name: "Datasets",
  getKey: (instance, options) => {
    const projectId = options?.projectId ?? instance.config.projectId;
    if (!projectId)
      throw new Error("A projectId is required to use the project API.");
    return projectId;
  },
  fetcher: (instance) => (options) => getClientState(instance, {
    apiVersion: API_VERSION$5,
    // non-null assertion is fine because we check above
    projectId: options?.projectId ?? instance.config.projectId,
    useProjectHostname: !0
  }).observable.pipe(switchMap((client) => client.observable.datasets.list()))
}), getDatasetsState = datasets.getState, resolveDatasets = datasets.resolveState;
function getEffectiveDocumentId(doc) {
  return doc.liveEdit ? doc.documentId : isReleasePerspective(doc.perspective) ? getVersionId(DocumentId(doc.documentId), doc.perspective.releaseName) : getPublishedId(DocumentId(doc.documentId));
}
const isSanityMutatePatch = (value) => !(typeof value != "object" || !value || !("type" in value) || typeof value.type != "string" || value.type !== "patch" || !("id" in value) || typeof value.id != "string" || !("patches" in value) || !Array.isArray(value.patches));
function createDocument(doc, initialValue) {
  let effectiveDocumentId;
  return typeof doc.documentId == "string" && (effectiveDocumentId = getEffectiveDocumentId({ ...doc, documentId: doc.documentId })), {
    type: "document.create",
    ...doc,
    ...effectiveDocumentId && { documentId: effectiveDocumentId },
    ...initialValue && { initialValue }
  };
}
function deleteDocument(doc) {
  const effectiveDocumentId = getEffectiveDocumentId(doc);
  return {
    type: "document.delete",
    ...doc,
    documentId: effectiveDocumentId
  };
}
function convertSanityMutatePatch(sanityPatchMutation) {
  return SanityEncoder.encode(sanityPatchMutation).map((i) => {
    const copy = { ...i.patch };
    return "id" in copy && delete copy.id, copy;
  });
}
function editDocument(doc, patches) {
  const effectiveDocumentId = getEffectiveDocumentId(doc);
  if (isSanityMutatePatch(patches)) {
    const converted = convertSanityMutatePatch(patches) ?? [];
    return {
      ...doc,
      type: "document.edit",
      documentId: effectiveDocumentId,
      patches: converted
    };
  }
  return {
    ...doc,
    type: "document.edit",
    documentId: effectiveDocumentId,
    ...patches && { patches: Array.isArray(patches) ? patches : [patches] }
  };
}
function publishDocument(doc) {
  const effectiveDocumentId = getEffectiveDocumentId(doc);
  return {
    type: "document.publish",
    ...doc,
    documentId: effectiveDocumentId
  };
}
function unpublishDocument(doc) {
  const effectiveDocumentId = getEffectiveDocumentId(doc);
  return {
    type: "document.unpublish",
    ...doc,
    documentId: effectiveDocumentId
  };
}
function discardDocument(doc) {
  const effectiveDocumentId = getEffectiveDocumentId(doc);
  return {
    type: "document.discard",
    ...doc,
    documentId: effectiveDocumentId
  };
}
const DOCUMENT_STATE_CLEAR_DELAY = 1e3, INITIAL_OUTGOING_THROTTLE_TIME = 1e3, API_VERSION$4 = "v2025-05-06";
function generateArrayKey(length = 12) {
  const numBytes = Math.ceil(length / 2), bytes = crypto.getRandomValues(new Uint8Array(numBytes));
  return Array.from(bytes, (byte) => byte.toString(16).padStart(2, "0")).join("").slice(0, length);
}
function memoize(fn) {
  const cache = /* @__PURE__ */ new WeakMap();
  return ((input) => {
    if (!input || typeof input != "object") return fn(input);
    const cached = cache.get(input);
    if (cached) return cached;
    const result = fn(input);
    return cache.set(input, result), result;
  });
}
const ensureArrayKeysDeep = memoize((input) => {
  if (!input || typeof input != "object") return input;
  if (Array.isArray(input))
    return !input.length || typeof input[0] != "object" || input.every(isKeyedObject) ? input : input.map((item) => !item || typeof item != "object" ? item : isKeyedObject(item) ? ensureArrayKeysDeep(item) : { ...ensureArrayKeysDeep(item), _key: generateArrayKey() });
  const entries = Object.entries(input).map(([key, value]) => [key, ensureArrayKeysDeep(value)]);
  return entries.every(([key, value]) => input[key] === value) ? input : Object.fromEntries(entries);
});
function set(input, pathExpressionValues) {
  const result = Object.entries(pathExpressionValues).flatMap(
    ([pathExpression, replacementValue]) => Array.from(jsonMatch(input, pathExpression)).map((matchEntry) => ({
      ...matchEntry,
      replacementValue
    }))
  ).reduce((acc, { path, replacementValue }) => setDeep(acc, path, replacementValue), input);
  return ensureArrayKeysDeep(result);
}
function setIfMissing(input, pathExpressionValues) {
  const result = Object.entries(pathExpressionValues).flatMap(([pathExpression, replacementValue]) => Array.from(jsonMatch(input, pathExpression)).map((matchEntry) => ({
    ...matchEntry,
    replacementValue
  }))).filter((matchEntry) => matchEntry.value === null || matchEntry.value === void 0).reduce((acc, { path, replacementValue }) => setDeep(acc, path, replacementValue), input);
  return ensureArrayKeysDeep(result);
}
function unset(input, pathExpressions) {
  const result = pathExpressions.flatMap((pathExpression) => Array.from(jsonMatch(input, pathExpression))).reverse().reduce((acc, { path }) => unsetDeep(acc, path), input);
  return ensureArrayKeysDeep(result);
}
function insert(input, { items, ...insertPatch }) {
  let operation, pathExpression;
  if ("before" in insertPatch ? (operation = "before", pathExpression = insertPatch.before) : "after" in insertPatch ? (operation = "after", pathExpression = insertPatch.after) : "replace" in insertPatch && (operation = "replace", pathExpression = insertPatch.replace), !operation || typeof pathExpression != "string" || !pathExpression.length) return input;
  const arrayPath = slicePath(pathExpression, 0, -1), positionPath = slicePath(pathExpression, -1);
  let result = input;
  for (const { path, value } of jsonMatch(input, arrayPath)) {
    if (!Array.isArray(value)) continue;
    let arr = value;
    switch (operation) {
      case "replace": {
        const indexesToRemove = /* @__PURE__ */ new Set();
        let position = 1 / 0;
        for (const itemMatch of jsonMatch(arr, positionPath)) {
          if (itemMatch.path.length !== 1) continue;
          const [segment] = itemMatch.path;
          if (typeof segment == "string") continue;
          let index;
          typeof segment == "number" && (index = segment), typeof index == "number" && index < 0 && (index = arr.length + index), isKeySegment(segment) && (index = getIndexForKey(arr, segment._key)), typeof index == "number" && (index < 0 && (index = arr.length + index), indexesToRemove.add(index), index < position && (position = index));
        }
        if (position === 1 / 0) continue;
        arr = arr.map((item, index) => ({ item, index })).filter(({ index }) => !indexesToRemove.has(index)).map(({ item }) => item), arr = [...arr.slice(0, position), ...items, ...arr.slice(position, arr.length)];
        break;
      }
      case "before": {
        let position = 1 / 0;
        for (const itemMatch of jsonMatch(arr, positionPath)) {
          if (itemMatch.path.length !== 1) continue;
          const [segment] = itemMatch.path;
          if (typeof segment == "string") continue;
          let index;
          typeof segment == "number" && (index = segment), typeof index == "number" && index < 0 && (index = arr.length + index), isKeySegment(segment) && (index = getIndexForKey(arr, segment._key)), typeof index == "number" && (index < 0 && (index = arr.length - index), index < position && (position = index));
        }
        if (position === 1 / 0) continue;
        arr = [...arr.slice(0, position), ...items, ...arr.slice(position, arr.length)];
        break;
      }
      case "after": {
        let position = -1 / 0;
        for (const itemMatch of jsonMatch(arr, positionPath)) {
          if (itemMatch.path.length !== 1) continue;
          const [segment] = itemMatch.path;
          if (typeof segment == "string") continue;
          let index;
          typeof segment == "number" && (index = segment), typeof index == "number" && index < 0 && (index = arr.length + index), isKeySegment(segment) && (index = getIndexForKey(arr, segment._key)), typeof index == "number" && index > position && (position = index);
        }
        if (position === -1 / 0) continue;
        arr = [...arr.slice(0, position + 1), ...items, ...arr.slice(position + 1, arr.length)];
        break;
      }
      default:
        continue;
    }
    result = setDeep(result, path, arr);
  }
  return ensureArrayKeysDeep(result);
}
function inc(input, pathExpressionValues) {
  const result = Object.entries(pathExpressionValues).flatMap(
    ([pathExpression, valueToAdd]) => Array.from(jsonMatch(input, pathExpression)).map((matchEntry) => ({
      ...matchEntry,
      valueToAdd
    }))
  ).filter(
    (matchEntry) => typeof matchEntry.value == "number"
  ).reduce((acc, { path, value, valueToAdd }) => setDeep(acc, path, value + valueToAdd), input);
  return ensureArrayKeysDeep(result);
}
function dec(input, pathExpressionValues) {
  const result = inc(
    input,
    Object.fromEntries(
      Object.entries(pathExpressionValues).filter(([, value]) => typeof value == "number").map(([key, value]) => [key, -value])
    )
  );
  return ensureArrayKeysDeep(result);
}
function diffMatchPatch(input, pathExpressionValues) {
  const result = Object.entries(pathExpressionValues).flatMap(
    ([pathExpression, dmp]) => Array.from(jsonMatch(input, pathExpression)).map((m) => ({ ...m, dmp }))
  ).filter((i) => i.value !== void 0).map(({ path, value, dmp }) => {
    if (typeof value != "string")
      throw new Error(
        `Can't diff-match-patch \`${JSON.stringify(value)}\` at path \`${stringifyPath(path)}\`, because it is not a string`
      );
    const [nextValue] = applyPatches(parsePatch(dmp), value);
    return { path, value: nextValue };
  }).reduce((acc, { path, value }) => setDeep(acc, path, value), input);
  return ensureArrayKeysDeep(result);
}
function ifRevisionID(input, revisionId) {
  const inputRev = typeof input == "object" && input && "_rev" in input && typeof input._rev == "string" ? input._rev : void 0;
  if (typeof inputRev != "string")
    throw new Error("Patch specified `ifRevisionID` but could not find document's revision ID.");
  if (revisionId !== inputRev)
    throw new Error(
      `Patch's \`ifRevisionID\` \`${revisionId}\` does not match document's revision ID \`${inputRev}\``
    );
  return input;
}
function setDeep(input, path, value) {
  const [currentSegment, ...restOfPath] = path;
  if (currentSegment === void 0) return value;
  if (typeof input != "object" || input === null) {
    if (typeof currentSegment == "string")
      return { [currentSegment]: setDeep(null, restOfPath, value) };
    let index;
    if (isKeySegment(currentSegment))
      index = 0;
    else if (typeof currentSegment == "number" && currentSegment >= 0)
      index = currentSegment;
    else
      return input;
    return [
      // fill until index
      ...Array.from({ length: index }).fill(null),
      // then set deep here
      setDeep(null, restOfPath, value)
    ];
  }
  if (Array.isArray(input)) {
    let index;
    return isKeySegment(currentSegment) ? index = getIndexForKey(input, currentSegment._key) ?? input.length : typeof currentSegment == "number" && (index = currentSegment < 0 ? input.length + currentSegment : currentSegment), index === void 0 ? input : index in input ? input.map(
      (nestedInput, i) => i === index ? setDeep(nestedInput, restOfPath, value) : nestedInput
    ) : [
      ...input,
      ...Array.from({ length: index - input.length }).fill(null),
      setDeep(null, restOfPath, value)
    ];
  }
  return typeof currentSegment == "object" ? input : currentSegment in input ? Object.fromEntries(
    Object.entries(input).map(
      ([key, nestedInput]) => key === currentSegment ? [key, setDeep(nestedInput, restOfPath, value)] : [key, nestedInput]
    )
  ) : { ...input, [currentSegment]: setDeep(null, restOfPath, value) };
}
function unsetDeep(input, path) {
  const [currentSegment, ...restOfPath] = path;
  if (currentSegment === void 0 || typeof input != "object" || input === null) return input;
  let _segment;
  if (isKeySegment(currentSegment) ? _segment = getIndexForKey(input, currentSegment._key) : (typeof currentSegment == "string" || typeof currentSegment == "number") && (_segment = currentSegment), _segment === void 0) return input;
  let segment = _segment;
  return typeof segment == "number" && Array.isArray(input) && (segment = segment < 0 ? input.length + segment : segment), segment in input ? restOfPath.length ? Array.isArray(input) ? input.map(
    (nestedInput, index) => index === segment ? unsetDeep(nestedInput, restOfPath) : nestedInput
  ) : Object.fromEntries(
    Object.entries(input).map(
      ([key, value]) => key === segment ? [key, unsetDeep(value, restOfPath)] : [key, value]
    )
  ) : Array.isArray(input) ? input.filter((_nestedInput, index) => index !== segment) : Object.fromEntries(Object.entries(input).filter(([key]) => key !== segment.toString())) : input;
}
const patchOperations = {
  ifRevisionID,
  set,
  setIfMissing,
  unset,
  inc,
  dec,
  insert,
  diffMatchPatch
};
function getId(id) {
  return !id || typeof id != "string" ? crypto.randomUUID() : id.endsWith(".") ? `${id}${crypto.randomUUID()}` : id;
}
function getDocumentIds(selection) {
  if ("id" in selection) {
    const ids = (Array.isArray(selection.id) ? selection.id : [selection.id]).filter((id) => typeof id == "string");
    return Array.from(new Set(ids));
  }
  if ("query" in selection)
    throw new Error("'query' in mutations is not supported.");
  return [];
}
function processMutations({
  documents,
  mutations,
  transactionId,
  timestamp
}) {
  if (!mutations.length) return documents;
  const dataset = { ...documents }, now = timestamp || (/* @__PURE__ */ new Date()).toISOString();
  for (const mutation of mutations) {
    if ("create" in mutation) {
      const id = getId(mutation.create._id);
      if (dataset[id])
        throw new Error(
          `Cannot create document with \`_id\` \`${id}\` because another document with the same ID already exists.`
        );
      const document = {
        // > `_createdAt` and `_updatedAt` may be submitted and will override
        // > the default which is of course the current time. This can be used
        // > to reconstruct a data-set with its timestamp structure intact.
        // >
        // > [- source](https://www.sanity.io/docs/http-mutations#c732f27330a4)
        _createdAt: now,
        _updatedAt: now,
        ...mutation.create,
        // prefer the user's `_createdAt` and `_updatedAt`
        _rev: transactionId,
        _id: id
      };
      dataset[id] = document;
      continue;
    }
    if ("createOrReplace" in mutation) {
      const id = getId(mutation.createOrReplace._id), prev = dataset[id], document = {
        ...mutation.createOrReplace,
        // otherwise, if the mutation provided, a `_createdAt` time, use it,
        // otherwise default to now
        _createdAt: (
          // if there was an existing document, use the previous `_createdAt`
          // since we're replacing the current document
          prev?._createdAt || // if there was no previous document, then we're creating this
          // document for the first time so we should use the `_createdAt` from
          // the mutation if the user included it
          typeof mutation.createOrReplace._createdAt == "string" && mutation.createOrReplace._createdAt || // otherwise, default to now
          now
        ),
        _updatedAt: (
          // if there was an existing document, then set the `_updatedAt` to now
          // since we're replacing the current document
          prev ? now : (
            // otherwise, we're creating this document for the first time,
            // in that case, use the user's `_updatedAt` if included in the
            // mutation
            typeof mutation.createOrReplace._updatedAt == "string" && mutation.createOrReplace._updatedAt || // otherwise default to now
            now
          )
        ),
        _rev: transactionId,
        _id: id
      };
      dataset[id] = document;
      continue;
    }
    if ("createIfNotExists" in mutation) {
      const id = getId(mutation.createIfNotExists._id);
      if (dataset[id]) continue;
      const document = {
        // same logic as `create`:
        // prefer the user's `_createdAt` and `_updatedAt`
        _createdAt: now,
        _updatedAt: now,
        ...mutation.createIfNotExists,
        _rev: transactionId,
        _id: id
      };
      dataset[id] = document;
      continue;
    }
    if ("delete" in mutation) {
      for (const id of getDocumentIds(mutation.delete))
        dataset[id] = null;
      continue;
    }
    if ("patch" in mutation) {
      const { patch } = mutation, patched = getDocumentIds(patch).map((id) => {
        if (!dataset[id])
          throw new Error(`Cannot patch document with ID \`${id}\` because it was not found.`);
        return Object.entries(patchOperations).reduce((acc, [type, operation]) => patch[type] ? operation(
          acc,
          // @ts-expect-error TS doesn't handle this union very well
          patch[type]
        ) : acc, dataset[id]);
      });
      for (const result of patched)
        dataset[result._id] = {
          ...result,
          _rev: transactionId,
          _updatedAt: now
        };
      continue;
    }
  }
  return dataset;
}
const DEFAULT_MAX_BUFFER_SIZE = 20, DEFAULT_DEADLINE_MS = 3e4;
class OutOfSyncError extends Error {
  /**
   * Attach state to the error for debugging/reporting
   */
  state;
  constructor(message, state) {
    super(message), this.name = "OutOfSyncError", this.state = state;
  }
}
class DeadlineExceededError extends OutOfSyncError {
  constructor(message, state) {
    super(message, state), this.name = "DeadlineExceededError";
  }
}
class MaxBufferExceededError extends OutOfSyncError {
  constructor(message, state) {
    super(message, state), this.name = "MaxBufferExceededError";
  }
}
function sortListenerEvents(options) {
  const { resolveChainDeadline = DEFAULT_DEADLINE_MS, maxBufferSize = DEFAULT_MAX_BUFFER_SIZE } = {};
  return (input$) => input$.pipe(
    // Maintain state: current base revision, a buffer of pending mutation events,
    // and a list of events to emit.
    scan(
      (state, event) => {
        if (event.type === "sync")
          return {
            base: { revision: event.document?._rev },
            buffer: [],
            emitEvents: [event]
          };
        if (event.type === "mutation") {
          if (!state.base)
            throw new Error(
              "Invalid state. Cannot process mutation event without a base sync event"
            );
          const buffer = state.buffer.concat(event), emitEvents = [];
          let baseRevision = state.base.revision, progress = !0;
          for (; progress; ) {
            progress = !1;
            const idx = buffer.findIndex((e) => e.previousRev === baseRevision);
            if (idx !== -1) {
              const [next] = buffer.splice(idx, 1);
              emitEvents.push(next), baseRevision = next.transition === "disappear" ? void 0 : next.resultRev, progress = !0;
            }
          }
          if (buffer.length >= maxBufferSize)
            throw new MaxBufferExceededError(
              `Too many unchainable mutation events (${buffer.length}) waiting to resolve.`,
              { base: { revision: baseRevision }, buffer, emitEvents }
            );
          return {
            base: { revision: baseRevision },
            buffer,
            emitEvents
          };
        }
        return { ...state, emitEvents: [event] };
      },
      {
        base: void 0,
        buffer: [],
        emitEvents: []
      }
    ),
    switchMap((state) => state.buffer.length > 0 ? concat(
      of(state),
      timer(resolveChainDeadline).pipe(
        mergeMap(
          () => throwError(
            () => new DeadlineExceededError(
              `Did not resolve chain within a deadline of ${resolveChainDeadline}ms`,
              state
            )
          )
        )
      )
    ) : of(state)),
    // Emit all events that are ready to be applied.
    mergeMap((state) => of(...state.emitEvents))
  );
}
const listen = ({ state }, documentId) => {
  const { sharedListener, fetchDocument } = state.get();
  return sharedListener.events.pipe(
    concatMap((e) => e.type === "welcome" ? fetchDocument(documentId).pipe(
      map$1((document) => ({ type: "sync", document }))
    ) : e.type === "mutation" && e.documentId === documentId ? of(e) : EMPTY),
    sortListenerEvents(),
    withLatestFrom(
      state.observable.pipe(
        map$1((s) => s.documentStates[documentId]),
        filter$1(Boolean),
        distinctUntilChanged()
      )
    ),
    map$1(([next, documentState]) => {
      if (next.type === "sync")
        return {
          type: "sync",
          documentId,
          document: next.document,
          revision: next.document?._rev,
          timestamp: next.document?._updatedAt ?? (/* @__PURE__ */ new Date()).toISOString()
        };
      const [document] = Object.values(
        processMutations({
          documents: { [documentId]: documentState.remote },
          mutations: next.mutations,
          transactionId: next.transactionId,
          timestamp: next.timestamp
        })
      ), { previousRev, transactionId, timestamp } = next;
      return {
        type: "mutation",
        documentId,
        document: document ?? null,
        revision: transactionId,
        timestamp,
        ...previousRev && { previousRev }
      };
    })
  );
};
class MultiKeyWeakMap {
  // The root of our nested WeakMap structure.
  #rootMap = /* @__PURE__ */ new WeakMap();
  // Instead of random IDs, we use a counter for deterministic IDs.
  static #globalIdCounter = 0;
  // Each instance keeps a cache mapping a key to its assigned ID.
  #idCache = /* @__PURE__ */ new WeakMap();
  /**
   * Assigns a numeric ID to the key.
   */
  #assignId(key) {
    const cachedId = this.#idCache.get(key);
    if (cachedId !== void 0) return cachedId;
    const id = MultiKeyWeakMap.#globalIdCounter;
    return this.#idCache.set(key, id), MultiKeyWeakMap.#globalIdCounter++, id;
  }
  /**
   * Remove duplicate keys and arrange them in a consistent order
   * by sorting according to their assigned IDs.
   */
  #arrangeKeys(keys) {
    const keyed = Array.from(new Set(keys)).map((key) => [this.#assignId(key), key]);
    return keyed.sort((a, b) => a[0] - b[0]), keyed.map(([, key]) => key);
  }
  /**
   * Recursively search the nested WeakMap structure for the value.
   */
  #getDeep(keys, map2) {
    if (keys.length === 0) return;
    const [firstKey, ...restKeys] = keys, node = map2.get(firstKey);
    if (node)
      return restKeys.length === 0 ? node.value : this.#getDeep(restKeys, node.next);
  }
  /**
   * Recursively create nodes along the key chain until the final key
   * is reached, then assign the value.
   */
  #setDeep(keys, map2, value) {
    if (keys.length === 0) return;
    const [firstKey, ...restKeys] = keys;
    let node = map2.get(firstKey);
    node || (node = {
      value: void 0,
      next: /* @__PURE__ */ new WeakMap()
    }, map2.set(firstKey, node)), restKeys.length === 0 ? node.value = value : this.#setDeep(restKeys, node.next, value);
  }
  /**
   * Retrieves the value associated with the array of keys.
   * The keys are de-duplicated and sorted so that the order does not matter.
   */
  get(keys) {
    const arrangedKeys = this.#arrangeKeys(keys);
    return this.#getDeep(arrangedKeys, this.#rootMap);
  }
  /**
   * Associates the value with the given array of keys.
   */
  set(keys, value) {
    const arrangedKeys = this.#arrangeKeys(keys);
    this.#setDeep(arrangedKeys, this.#rootMap, value);
  }
}
function createGrantsLookup(datasetAcl) {
  const filtersByGrant = {
    create: /* @__PURE__ */ new Set(),
    history: /* @__PURE__ */ new Set(),
    read: /* @__PURE__ */ new Set(),
    update: /* @__PURE__ */ new Set()
  };
  for (const entry of datasetAcl)
    for (const grant of entry.permissions) {
      const set2 = filtersByGrant[grant];
      set2.add(entry.filter), filtersByGrant[grant] = set2;
    }
  return Object.fromEntries(
    Object.entries(filtersByGrant).map(([grant, filters]) => {
      const combinedFilter = Array.from(filters).map((i) => `(${i})`).join("||");
      return combinedFilter ? [grant, parse(`$document {"_": ${combinedFilter}}._`)] : [grant, parse("false")];
    })
  );
}
const documentsCache = new MultiKeyWeakMap(), actionsCache = /* @__PURE__ */ new WeakMap(), nullReplacer = {}, documentsSelector = createSelector(
  [
    ({ state: { documentStates } }) => documentStates,
    (_context, { actions }) => actions
  ],
  (documentStates, actions) => {
    const documentIds = new Set(
      actions.map((action) => {
        if (typeof action.documentId != "string") return [];
        if (action.liveEdit) return [action.documentId];
        const ids = [
          getPublishedId(DocumentId(action.documentId)),
          getDraftId(DocumentId(action.documentId))
        ];
        return isReleasePerspective(action.perspective) && ids.push(getVersionId(DocumentId(action.documentId), action.perspective.releaseName)), ids;
      }).flat()
    ), documents = {};
    for (const documentId of documentIds) {
      const local = documentStates[documentId]?.local;
      if (local === void 0) return;
      documents[documentId] = local;
    }
    const keys = Object.values(
      // value in this record will be `undefined` because
      // of the early return if undefined is found above
      documents
    ).map((doc) => doc === null ? nullReplacer : doc);
    return documentsCache.get(keys) || (documentsCache.set(keys, documents), documents);
  }
), memoizedActionsSelector = createSelector(
  [
    documentsSelector,
    (_state, { actions }) => actions
  ],
  (documents, actions) => {
    if (!documents) return;
    let nestedCache = actionsCache.get(documents);
    nestedCache || (nestedCache = /* @__PURE__ */ new Map(), actionsCache.set(documents, nestedCache));
    const normalizedActions = Array.isArray(actions) ? actions : [actions], actionsKey = JSON.stringify(normalizedActions);
    return nestedCache.get(actionsKey) || (nestedCache.set(actionsKey, normalizedActions), normalizedActions);
  }
);
function checkGrant$1(grantExpr, document) {
  const value = evaluateSync(grantExpr, { params: { document } });
  return value.type === "boolean" && value.data;
}
const enNarrowConjunction = new Intl.ListFormat("en", { style: "narrow", type: "conjunction" });
function calculatePermissions(...args) {
  return _calculatePermissions(...args);
}
const _calculatePermissions = createSelector(
  [
    ({ state: { grants } }) => grants,
    documentsSelector,
    memoizedActionsSelector
  ],
  (grants, documents, actions) => {
    if (!documents || !grants || !actions) return;
    const timestamp = (/* @__PURE__ */ new Date()).toISOString(), reasons = [];
    try {
      processActions({
        actions,
        transactionId: crypto.randomUUID(),
        working: documents,
        base: documents,
        timestamp,
        grants
      });
    } catch (error) {
      if (error instanceof PermissionActionError)
        reasons.push({
          message: error.message,
          documentId: error.documentId,
          type: "access"
        });
      else if (error instanceof ActionError)
        reasons.push({
          message: error.message,
          documentId: error.documentId,
          type: "precondition"
        });
      else
        throw error;
    }
    for (const action of actions)
      if (action.type === "document.edit" && !action.patches?.length) {
        const docId = action.documentId;
        let doc;
        action.liveEdit ? doc = documents[docId] : isReleasePerspective(action.perspective) ? doc = documents[getVersionId(DocumentId(docId), action.perspective.releaseName)] : doc = documents[getDraftId(DocumentId(docId))] ?? documents[getPublishedId(DocumentId(docId))], doc ? checkGrant$1(grants.update, doc) || reasons.push({
          type: "access",
          message: `You are not allowed to edit the document with ID "${docId}".`,
          documentId: docId
        }) : isReleasePerspective(action.perspective) ? reasons.push({
          type: "precondition",
          message: `The version document with ID "${docId}" could not be found. Please create it or add it to the release first.`,
          documentId: docId
        }) : reasons.push({
          type: "precondition",
          message: `The document with ID "${docId}" could not be found. Please check that it exists before editing.`,
          documentId: docId
        });
      }
    const allowed = reasons.length === 0;
    if (allowed) return { allowed };
    const sortedReasons = reasons.map((reason, index) => ({ ...reason, index })).sort((a, b) => a.type !== b.type ? a.type === "access" ? -1 : 1 : a.message.localeCompare(b.message, "en-US")).map(({ index: _index, ...reason }) => reason);
    return {
      allowed,
      reasons: sortedReasons,
      message: enNarrowConjunction.format(sortedReasons.map((i) => i.message))
    };
  }
);
function checkGrant(grantExpr, document) {
  const value = evaluateSync(grantExpr, { params: { document } });
  return value.type === "boolean" && value.data;
}
class ActionError extends Error {
  documentId;
  transactionId;
  constructor(options) {
    super(options.message), Object.assign(this, options);
  }
}
class PermissionActionError extends ActionError {
}
function processActions({
  actions,
  transactionId,
  working: initialWorking,
  base: initialBase,
  timestamp,
  grants
}) {
  let working = { ...initialWorking }, base = { ...initialBase };
  const outgoingActions = [], outgoingMutations = [];
  for (const action of actions)
    switch (action.type) {
      case "document.create": {
        const documentId = getId(action.documentId);
        if (action.liveEdit) {
          if (working[documentId])
            throw new ActionError({
              documentId,
              transactionId,
              message: "This document already exists."
            });
          const newDocBase2 = { _type: action.documentType, _id: documentId, ...action.initialValue }, mutations2 = [{ create: {
            _type: action.documentType,
            _id: documentId,
            ...action.initialValue
          } }];
          if (base = processMutations({
            documents: base,
            transactionId,
            mutations: [{ create: newDocBase2 }],
            timestamp
          }), working = processMutations({
            documents: working,
            transactionId,
            mutations: mutations2,
            timestamp
          }), !checkGrant(grants.create, working[documentId]))
            throw new PermissionActionError({
              documentId,
              transactionId,
              message: `You do not have permission to create document "${documentId}".`
            });
          outgoingMutations.push(...mutations2);
          continue;
        }
        const versionId = isReleasePerspective(action.perspective) ? getVersionId(DocumentId(documentId), action.perspective.releaseName) : void 0, draftId = getDraftId(DocumentId(documentId)), publishedId = getPublishedId(DocumentId(documentId));
        if (versionId ? working[versionId] : working[draftId]) {
          const errorDocType = versionId ? "release version" : "draft";
          throw new ActionError({
            documentId,
            transactionId,
            message: `A ${errorDocType} of this document already exists. Please use or discard the existing ${errorDocType} before creating a new one.`
          });
        }
        const newDocBase = {
          ...base[draftId] ?? base[publishedId],
          _type: action.documentType,
          _id: versionId ?? draftId,
          ...action.initialValue
        }, newDocWorking = {
          ...working[draftId] ?? working[publishedId],
          _type: action.documentType,
          _id: versionId ?? draftId,
          ...action.initialValue
        }, mutations = [{ create: newDocWorking }];
        if (base = processMutations({
          documents: base,
          transactionId,
          mutations: [{ create: newDocBase }],
          timestamp
        }), working = processMutations({
          documents: working,
          transactionId,
          mutations,
          timestamp
        }), versionId && !checkGrant(grants.create, working[versionId]))
          throw new PermissionActionError({
            documentId,
            transactionId,
            message: `You do not have permission to create a release version for document "${documentId}".`
          });
        if (!versionId && !checkGrant(grants.create, working[draftId]))
          throw new PermissionActionError({
            documentId,
            transactionId,
            message: `You do not have permission to create a draft for document "${documentId}".`
          });
        outgoingMutations.push(...mutations), outgoingActions.push({
          actionType: "sanity.action.document.version.create",
          publishedId,
          attributes: newDocWorking
        });
        continue;
      }
      case "document.delete": {
        const documentId = action.documentId;
        if (isReleasePerspective(action.perspective))
          throw new ActionError({
            documentId,
            transactionId,
            message: 'Cannot delete a version document. You may want to use the "unpublish" or "discard" actions instead.'
          });
        if (action.liveEdit) {
          if (!working[documentId])
            throw new ActionError({
              documentId,
              transactionId,
              message: "The document you are trying to delete does not exist."
            });
          if (!checkGrant(grants.update, working[documentId]))
            throw new PermissionActionError({
              documentId,
              transactionId,
              message: "You do not have permission to delete this document."
            });
          const mutations2 = [{ delete: { id: documentId } }];
          base = processMutations({ documents: base, transactionId, mutations: mutations2, timestamp }), working = processMutations({ documents: working, transactionId, mutations: mutations2, timestamp }), outgoingMutations.push(...mutations2);
          continue;
        }
        const draftId = getDraftId(DocumentId(documentId)), publishedId = getPublishedId(DocumentId(documentId));
        if (!working[publishedId])
          throw new ActionError({
            documentId,
            transactionId,
            message: working[draftId] ? "Cannot delete a document without a published version." : "The document you are trying to delete does not exist."
          });
        const cantDeleteDraft = working[draftId] && !checkGrant(grants.update, working[draftId]), cantDeletePublished = working[publishedId] && !checkGrant(grants.update, working[publishedId]);
        if (cantDeleteDraft || cantDeletePublished)
          throw new PermissionActionError({
            documentId,
            transactionId,
            message: "You do not have permission to delete this document."
          });
        const mutations = [{ delete: { id: publishedId } }, { delete: { id: draftId } }], includeDrafts = working[draftId] ? [draftId] : void 0;
        base = processMutations({ documents: base, transactionId, mutations, timestamp }), working = processMutations({ documents: working, transactionId, mutations, timestamp }), outgoingMutations.push(...mutations), outgoingActions.push({
          actionType: "sanity.action.document.delete",
          publishedId,
          ...includeDrafts ? { includeDrafts } : {}
        });
        continue;
      }
      case "document.discard": {
        const documentId = getId(action.documentId);
        if (action.liveEdit)
          throw new ActionError({
            documentId,
            transactionId,
            message: `Cannot discard changes for liveEdit document "${documentId}". LiveEdit documents do not support drafts.`
          });
        const versionId = isReleasePerspective(action.perspective) ? getVersionId(DocumentId(documentId), action.perspective.releaseName) : getDraftId(DocumentId(documentId)), mutations = [{ delete: { id: versionId } }];
        if (!working[versionId])
          throw new ActionError({
            documentId,
            transactionId,
            message: `There is no draft or version available to discard for document "${documentId}".`
          });
        if (!checkGrant(grants.update, working[versionId]))
          throw new PermissionActionError({
            documentId,
            transactionId,
            message: `You do not have permission to discard changes for document "${documentId}".`
          });
        base = processMutations({ documents: base, transactionId, mutations, timestamp }), working = processMutations({ documents: working, transactionId, mutations, timestamp }), outgoingMutations.push(...mutations), outgoingActions.push({
          actionType: "sanity.action.document.version.discard",
          versionId
        });
        continue;
      }
      case "document.edit": {
        const documentId = getId(action.documentId);
        if (action.liveEdit) {
          const userPatches2 = action.patches?.map((patch) => ({ patch: { id: documentId, ...patch } }));
          if (!userPatches2?.length) continue;
          if (!working[documentId] || !base[documentId])
            throw new ActionError({
              documentId,
              transactionId,
              message: "Cannot edit document because it does not exist."
            });
          const baseBefore2 = base[documentId];
          userPatches2 && (base = processMutations({
            documents: base,
            transactionId,
            mutations: userPatches2,
            timestamp
          }));
          const baseAfter2 = base[documentId], patches2 = diffValue(baseBefore2, baseAfter2), workingBefore2 = working[documentId];
          if (!checkGrant(grants.update, workingBefore2))
            throw new PermissionActionError({
              documentId,
              transactionId,
              message: `You do not have permission to edit document "${documentId}".`
            });
          const workingMutations2 = patches2.map((patch) => ({ patch: { id: documentId, ...patch } }));
          working = processMutations({
            documents: working,
            transactionId,
            mutations: workingMutations2,
            timestamp
          }), outgoingMutations.push(...workingMutations2);
          continue;
        }
        const versionId = isReleasePerspective(action.perspective) ? getVersionId(DocumentId(documentId), action.perspective.releaseName) : void 0, draftId = getDraftId(DocumentId(documentId)), publishedId = getPublishedId(DocumentId(documentId)), patchDocumentId = isReleasePerspective(action.perspective) ? versionId : draftId, userPatches = action.patches?.map((patch) => ({
          patch: { id: patchDocumentId, ...patch }
        }));
        if (!userPatches?.length) continue;
        if (isReleasePerspective(action.perspective)) {
          if (!working[versionId] && !base[versionId])
            throw new ActionError({
              documentId,
              transactionId,
              message: "This document does not exist in the release. Please create it or add it to the release first."
            });
        } else if (!working[draftId] && !working[publishedId] || !base[draftId] && !base[publishedId])
          throw new ActionError({
            documentId,
            transactionId,
            message: "Cannot edit document because it does not exist in draft or published form."
          });
        const baseMutations = [];
        !isReleasePerspective(action.perspective) && !base[draftId] && base[publishedId] && baseMutations.push({ create: { ...base[publishedId], _id: draftId } });
        const baseBefore = base[patchDocumentId] ?? base[publishedId];
        userPatches && baseMutations.push(...userPatches), base = processMutations({
          documents: base,
          transactionId,
          mutations: baseMutations,
          timestamp
        });
        const baseAfter = base[patchDocumentId], patches = diffValue(baseBefore, baseAfter), workingMutations = [];
        if (!isReleasePerspective(action.perspective) && !working[draftId] && working[publishedId]) {
          const newDraftFromPublished = { ...working[publishedId], _id: draftId };
          if (!checkGrant(grants.create, newDraftFromPublished))
            throw new PermissionActionError({
              documentId,
              transactionId,
              message: "You do not have permission to create a draft for editing this document."
            });
          workingMutations.push({ create: newDraftFromPublished });
        }
        const workingBefore = working[patchDocumentId] ?? working[publishedId];
        if (!checkGrant(grants.update, workingBefore))
          throw new PermissionActionError({
            documentId,
            transactionId,
            message: `You do not have permission to edit document "${documentId}".`
          });
        workingMutations.push(...patches.map((patch) => ({ patch: { id: patchDocumentId, ...patch } }))), working = processMutations({
          documents: working,
          transactionId,
          mutations: workingMutations,
          timestamp
        }), outgoingMutations.push(...workingMutations), outgoingActions.push(
          ...patches.map((patch) => ({
            actionType: "sanity.action.document.edit",
            draftId: patchDocumentId,
            publishedId,
            patch
          }))
        );
        continue;
      }
      case "document.publish": {
        const documentId = getId(action.documentId);
        if (action.liveEdit || isReleasePerspective(action.perspective))
          throw new ActionError({
            documentId,
            transactionId,
            message: "Cannot publish this document. Publishing is not supported for liveEdit or version (release) documents."
          });
        const draftId = getDraftId(DocumentId(documentId)), publishedId = getPublishedId(DocumentId(documentId)), workingDraft = working[draftId], baseDraft = base[draftId];
        if (!workingDraft || !baseDraft)
          throw new ActionError({
            documentId,
            transactionId,
            message: `Cannot publish because no draft version was found for document "${documentId}".`
          });
        if (!isDeepEqual(workingDraft, baseDraft))
          throw new ActionError({
            documentId,
            transactionId,
            message: "Publish aborted: The document has changed elsewhere. Please try again."
          });
        const newPublishedFromDraft = { ...strengthenOnPublish(workingDraft), _id: publishedId }, mutations = [
          { delete: { id: draftId } },
          { createOrReplace: newPublishedFromDraft }
        ];
        if (working[draftId] && !checkGrant(grants.update, working[draftId]))
          throw new PermissionActionError({
            documentId,
            transactionId,
            message: `Publish failed: You do not have permission to update the draft for "${documentId}".`
          });
        if (working[publishedId] && !checkGrant(grants.update, newPublishedFromDraft))
          throw new PermissionActionError({
            documentId,
            transactionId,
            message: `Publish failed: You do not have permission to update the published version of "${documentId}".`
          });
        if (!working[publishedId] && !checkGrant(grants.create, newPublishedFromDraft))
          throw new PermissionActionError({
            documentId,
            transactionId,
            message: `Publish failed: You do not have permission to publish a new version of "${documentId}".`
          });
        base = processMutations({ documents: base, transactionId, mutations, timestamp }), working = processMutations({ documents: working, transactionId, mutations, timestamp }), outgoingMutations.push(...mutations), outgoingActions.push({
          actionType: "sanity.action.document.publish",
          draftId,
          publishedId
        });
        continue;
      }
      case "document.unpublish": {
        const documentId = getId(action.documentId);
        if (action.liveEdit || isReleasePerspective(action.perspective))
          throw new ActionError({
            documentId,
            transactionId,
            message: "Cannot unpublish this document. Unpublishing is not supported for liveEdit or version (release) documents."
          });
        const draftId = getDraftId(DocumentId(documentId)), publishedId = getPublishedId(DocumentId(documentId));
        if (!working[publishedId] && !base[publishedId])
          throw new ActionError({
            documentId,
            transactionId,
            message: `Cannot unpublish because the document "${documentId}" is not currently published.`
          });
        const sourceDoc = working[publishedId] ?? base[publishedId], newDraftFromPublished = { ...sourceDoc, _id: draftId }, mutations = [
          { delete: { id: publishedId } },
          { createIfNotExists: newDraftFromPublished }
        ];
        if (!checkGrant(grants.update, sourceDoc))
          throw new PermissionActionError({
            documentId,
            transactionId,
            message: `You do not have permission to unpublish the document "${documentId}".`
          });
        if (!working[draftId] && !checkGrant(grants.create, newDraftFromPublished))
          throw new PermissionActionError({
            documentId,
            transactionId,
            message: `You do not have permission to create a draft from the published version of "${documentId}".`
          });
        base = processMutations({
          documents: base,
          transactionId,
          mutations: [
            { delete: { id: publishedId } },
            { createIfNotExists: { ...base[publishedId] ?? sourceDoc, _id: draftId } }
          ],
          timestamp
        }), working = processMutations({ documents: working, transactionId, mutations, timestamp }), outgoingMutations.push(...mutations), outgoingActions.push({
          actionType: "sanity.action.document.unpublish",
          draftId,
          publishedId
        });
        continue;
      }
      default:
        throw new Error(
          `Unknown action type: "${// @ts-expect-error invalid input
          action.type}". Please contact support if this issue persists.`
        );
    }
  const previousRevs = Object.fromEntries(
    Object.entries(initialWorking).map(([id, doc]) => [id, doc?._rev])
  );
  return {
    working,
    outgoingActions,
    outgoingMutations,
    previous: initialWorking,
    previousRevs
  };
}
function strengthenOnPublish(draft) {
  const isStrengthenReference = (value) => "_strengthenOnPublish" in value;
  function strengthen(value) {
    if (typeof value != "object" || !value) return value;
    if (isStrengthenReference(value)) {
      const { _strengthenOnPublish, _weak, ...rest } = value;
      return {
        ...rest,
        ..._strengthenOnPublish.weak && { _weak: !0 }
      };
    }
    return Array.isArray(value) ? value.map(strengthen) : Object.fromEntries(Object.entries(value).map(([k, v]) => [k, strengthen(v)]));
  }
  return strengthen(draft);
}
const EMPTY_REVISIONS = {};
function queueTransaction(prev, transaction) {
  const { transactionId, actions } = transaction;
  return {
    ...getDocumentIdsFromHandleLikes(actions).reduce(
      (acc, id) => addSubscriptionIdToDocument(acc, id, transactionId),
      prev
    ),
    queued: [...prev.queued, transaction]
  };
}
function removeQueuedTransaction(prev, transactionId) {
  const transaction = prev.queued.find((t) => t.transactionId === transactionId);
  return transaction ? {
    ...getDocumentIdsFromHandleLikes(transaction.actions).reduce(
      (acc, id) => removeSubscriptionIdFromDocument(acc, id, transactionId),
      prev
    ),
    queued: prev.queued.filter((t) => transactionId !== t.transactionId)
  } : prev;
}
function applyFirstQueuedTransaction(prev) {
  const queued = prev.queued.at(0);
  if (!queued || !prev.grants) return prev;
  const ids = getDocumentIdsFromHandleLikes(queued.actions);
  if (ids.some((id) => prev.documentStates[id]?.local === void 0)) return prev;
  const working = ids.reduce((acc, id) => (acc[id] = prev.documentStates[id]?.local, acc), {}), timestamp = (/* @__PURE__ */ new Date()).toISOString(), result = processActions({
    ...queued,
    working,
    base: working,
    timestamp,
    grants: prev.grants
  }), applied = {
    ...queued,
    ...result,
    base: result.previous,
    timestamp
  };
  return {
    ...prev,
    applied: [...prev.applied, applied],
    queued: prev.queued.filter((t) => t.transactionId !== queued.transactionId),
    documentStates: Object.entries(result.working).reduce(
      (acc, [id, next]) => {
        const prevDoc = acc[id];
        return prevDoc && (acc[id] = { ...prevDoc, local: next }), acc;
      },
      { ...prev.documentStates }
    )
  };
}
function batchAppliedTransactions([curr, ...rest]) {
  if (!curr) return;
  if (!curr.actions.length) return batchAppliedTransactions(rest);
  if (curr.actions.length > 1)
    return {
      ...curr,
      disableBatching: !0,
      batchedTransactionIds: [curr.transactionId]
    };
  const [action] = curr.actions;
  if (action.type !== "document.edit" || curr.disableBatching)
    return {
      ...curr,
      disableBatching: !0,
      batchedTransactionIds: [curr.transactionId]
    };
  const editAction = {
    ...curr,
    actions: [action],
    disableBatching: !1,
    batchedTransactionIds: [curr.transactionId]
  };
  if (!rest.length) return editAction;
  const next = batchAppliedTransactions(rest);
  if (next)
    return next.disableBatching || !!action.liveEdit != !!next.actions[0]?.liveEdit ? editAction : {
      disableBatching: !1,
      // Use the transactionId from the later (next) transaction.
      transactionId: next.transactionId,
      // Accumulate actions: current action first, then later ones.
      actions: [action, ...next.actions],
      // Merge outgoingActions in order.
      outgoingActions: [...curr.outgoingActions, ...next.outgoingActions],
      // Batched transaction IDs: preserve order by placing curr first.
      batchedTransactionIds: [curr.transactionId, ...next.batchedTransactionIds],
      // Merge outgoingMutations in order.
      outgoingMutations: [...curr.outgoingMutations, ...next.outgoingMutations],
      // Working state reflects the latest optimistic changes: later transactions override earlier.
      working: { ...curr.working, ...next.working },
      // Base state (base, previous, previousRevs) must reflect the original state.
      // Use curr values (the earliest transaction) to override later ones.
      previousRevs: { ...next.previousRevs, ...curr.previousRevs },
      previous: { ...next.previous, ...curr.previous },
      base: { ...next.base, ...curr.base },
      // Use the earliest timestamp from curr.
      timestamp: curr.timestamp ?? next.timestamp
    };
}
function transitionAppliedTransactionsToOutgoing(prev) {
  if (prev.outgoing) return prev;
  const transaction = batchAppliedTransactions(prev.applied);
  if (!transaction) return prev;
  const {
    transactionId,
    previousRevs,
    working,
    batchedTransactionIds: consumedTransactions
  } = transaction, timestamp = (/* @__PURE__ */ new Date()).toISOString();
  return {
    ...prev,
    outgoing: transaction,
    applied: prev.applied.filter((i) => !consumedTransactions.includes(i.transactionId)),
    documentStates: Object.entries(previousRevs).reduce(
      (acc, [documentId, previousRev]) => {
        if (working[documentId]?._rev === previousRev) return acc;
        const documentState = prev.documentStates[documentId];
        return documentState && (acc[documentId] = {
          ...documentState,
          unverifiedRevisions: {
            ...documentState.unverifiedRevisions,
            // add unverified revision
            [transactionId]: { documentId, previousRev, transactionId, timestamp }
          }
        }), acc;
      },
      { ...prev.documentStates }
    )
  };
}
function cleanupOutgoingTransaction(prev) {
  const { outgoing } = prev;
  if (!outgoing) return prev;
  let next = prev;
  const ids = getDocumentIdsFromHandleLikes(outgoing.actions);
  for (const transactionId of outgoing.batchedTransactionIds)
    for (const documentId of ids)
      next = removeSubscriptionIdFromDocument(next, documentId, transactionId);
  return { ...next, outgoing: void 0 };
}
function revertOutgoingTransaction(prev) {
  if (!prev.grants) return prev;
  let working = Object.fromEntries(
    Object.entries(prev.documentStates).map(([documentId, documentState]) => [
      documentId,
      documentState?.remote
    ])
  );
  const nextApplied = [];
  for (const t of prev.applied)
    try {
      const next = processActions({ ...t, working, grants: prev.grants });
      working = next.working, nextApplied.push({ ...t, ...next });
    } catch (error) {
      if (error instanceof ActionError) continue;
      throw error;
    }
  return {
    ...prev,
    applied: nextApplied,
    outgoing: void 0,
    documentStates: Object.fromEntries(
      Object.entries(prev.documentStates).filter((e) => !!e[1]).map(([documentId, { unverifiedRevisions = {}, local, ...documentState }]) => {
        const next = {
          ...documentState,
          local: documentId in working ? working[documentId] : local,
          unverifiedRevisions: prev.outgoing && prev.outgoing.transactionId in unverifiedRevisions ? omitProperty(unverifiedRevisions, prev.outgoing.transactionId) : unverifiedRevisions
        };
        return [documentId, next];
      })
    )
  };
}
function applyRemoteDocument(prev, { document, documentId, previousRev, revision, timestamp, type }, events) {
  if (!prev.grants) return prev;
  const prevDocState = prev.documentStates[documentId];
  if (!prevDocState) return prev;
  const prevUnverifiedRevisions = prevDocState.unverifiedRevisions, revisionToVerify = revision ? prevUnverifiedRevisions?.[revision] : void 0;
  let unverifiedRevisions = prevUnverifiedRevisions ?? EMPTY_REVISIONS;
  if (revision && revisionToVerify && (unverifiedRevisions = omitProperty(prevUnverifiedRevisions, revision)), type === "sync" && (unverifiedRevisions = Object.fromEntries(
    Object.entries(unverifiedRevisions).filter(([, unverifiedRevision]) => unverifiedRevision ? new Date(timestamp).getTime() <= new Date(unverifiedRevision.timestamp).getTime() : !1)
  )), revisionToVerify && revisionToVerify.previousRev === previousRev)
    return {
      ...prev,
      documentStates: {
        ...prev.documentStates,
        [documentId]: {
          ...prevDocState,
          remote: document,
          remoteRev: revision,
          unverifiedRevisions
        }
      }
    };
  let working = { ...prev.applied.at(0)?.previous, [documentId]: document };
  const nextApplied = [];
  for (const curr of prev.applied)
    try {
      const next = processActions({ ...curr, working, grants: prev.grants });
      working = next.working, nextApplied.push({ ...curr, ...next });
    } catch (error) {
      if (error instanceof ActionError) {
        events.next({
          type: "rebase-error",
          transactionId: error.transactionId,
          documentId: error.documentId,
          message: error.message,
          error
        });
        continue;
      }
      throw error;
    }
  return {
    ...prev,
    applied: nextApplied,
    documentStates: {
      ...prev.documentStates,
      [documentId]: {
        ...prevDocState,
        remote: document,
        remoteRev: revision,
        local: working[documentId],
        unverifiedRevisions
      }
    }
  };
}
function addSubscriptionIdToDocument(prev, documentId, subscriptionId) {
  const prevDocState = prev.documentStates?.[documentId], prevSubscriptions = prevDocState?.subscriptions ?? [];
  return {
    ...prev,
    documentStates: {
      ...prev.documentStates,
      [documentId]: {
        ...prevDocState,
        id: documentId,
        subscriptions: [...prevSubscriptions, subscriptionId]
      }
    }
  };
}
function removeSubscriptionIdFromDocument(prev, documentId, subscriptionId) {
  const prevDocState = prev.documentStates?.[documentId], subscriptions = (prevDocState?.subscriptions ?? []).filter((id) => id !== subscriptionId);
  return prevDocState ? subscriptions.length ? {
    ...prev,
    documentStates: {
      ...prev.documentStates,
      [documentId]: { ...prevDocState, subscriptions }
    }
  } : { ...prev, documentStates: omitProperty(prev.documentStates, documentId) } : prev;
}
function manageSubscriberIds({ state }, handles) {
  const documentIds = getDocumentIdsFromHandleLikes(handles), subscriptionId = insecureRandomId();
  return state.set(
    "addSubscribers",
    (prev) => documentIds.reduce(
      (acc, id) => addSubscriptionIdToDocument(acc, id, subscriptionId),
      prev
    )
  ), () => {
    setCleanupTimeout(() => {
      state.set(
        "removeSubscribers",
        (prev) => documentIds.reduce(
          (acc, id) => removeSubscriptionIdFromDocument(acc, id, subscriptionId),
          prev
        )
      );
    }, DOCUMENT_STATE_CLEAR_DELAY);
  };
}
function getDocumentIdsFromHandleLikes(handles) {
  return handles.flatMap((handle) => {
    const idsForDocument = [];
    return handle.documentId ? handle.liveEdit ? [handle.documentId] : (isReleasePerspective(handle.perspective) && idsForDocument.push(
      getVersionId(DocumentId(handle.documentId), handle.perspective.releaseName)
    ), idsForDocument.push(getPublishedId(DocumentId(handle.documentId))), idsForDocument.push(getDraftId(DocumentId(handle.documentId))), idsForDocument) : [];
  });
}
function getDocumentEvents(outgoing) {
  const documentIdsByAction = Object.entries(
    outgoing.actions.reduce(
      (acc, { type, documentId }) => {
        const ids = acc[type] || /* @__PURE__ */ new Set();
        return documentId && ids.add(documentId), acc[type] = ids, acc;
      },
      {}
    )
  ), actionMap = {
    "document.create": "created",
    "document.delete": "deleted",
    "document.discard": "discarded",
    "document.edit": "edited",
    "document.publish": "published",
    "document.unpublish": "unpublished"
  };
  return documentIdsByAction.flatMap(
    ([actionType, documentIds]) => Array.from(documentIds).map(
      (documentId) => ({ type: actionMap[actionType], documentId, outgoing })
    )
  );
}
const API_VERSION$3 = "v2025-05-06";
function createSharedListener(instance, resource) {
  const dispose$ = new Subject(), events$ = getClientState(instance, {
    apiVersion: API_VERSION$3,
    resource
  }).observable.pipe(
    switchMap(
      (client) => (
        // TODO: it seems like the client.listen method is not emitting disconnected
        // events. this is important to ensure we have an up to date version of the
        // doc. probably should introduce our own events for when the user goes offline
        client.listen(
          "*",
          {},
          {
            events: ["mutation", "welcome", "reconnect"],
            includeResult: !1,
            tag: "document-listener"
            // // from manual testing, it seems like mendoza patches may be
            // // causing some ambiguity/wonkiness
            // includeMutations: false,
            // effectFormat: 'mendoza',
          }
        )
      )
    ),
    takeUntil(dispose$),
    share()
  ), [welcome$, mutation$] = partition(events$, (e) => e.type === "welcome");
  return {
    events: merge(
      // we replay the welcome event because that event kicks off fetching the document
      welcome$.pipe(shareReplay(1)),
      mutation$
    ),
    dispose: () => dispose$.next()
  };
}
function createFetchDocument(instance, resource) {
  return function(documentId) {
    return getClientState(instance, {
      apiVersion: API_VERSION$3,
      resource
    }).observable.pipe(
      switchMap((client) => createDocumentLoaderFromClient(client)(documentId)),
      map$1((result) => {
        if (!result.accessible) {
          if (result.reason === "existence") return null;
          throw new Error(`Document with ID \`${documentId}\` is inaccessible due to permissions.`);
        }
        return result.document;
      }),
      first$1()
    );
  };
}
const documentStore = defineStore({
  name: "Document",
  getInitialState: (instance, { resource }) => ({
    documentStates: {},
    // these can be emptied on refetch
    queued: [],
    applied: [],
    sharedListener: createSharedListener(instance, resource),
    fetchDocument: createFetchDocument(instance, resource),
    events: new Subject()
  }),
  initialize(context) {
    const { sharedListener } = context.state.get(), subscriptions = [
      subscribeToQueuedAndApplyNextTransaction(context),
      subscribeToSubscriptionsAndListenToDocuments(context),
      subscribeToAppliedAndSubmitNextTransaction(context),
      subscribeToClientAndFetchDatasetAcl(context)
    ];
    return () => {
      sharedListener.dispose(), subscriptions.forEach((subscription) => subscription.unsubscribe());
    };
  }
});
function getDocumentState(...args) {
  return _getDocumentState(...args);
}
const _getDocumentState = bindActionByResource(
  documentStore,
  createStateSourceAction({
    selector: ({ state: { error, documentStates } }, options) => {
      const { documentId: docId, path, liveEdit, perspective } = options, documentId = DocumentId(docId);
      if (error) throw error;
      let document;
      if (liveEdit)
        document = documentStates[documentId]?.local;
      else {
        let version;
        if (isReleasePerspective(perspective)) {
          const versionId = getVersionId(documentId, perspective.releaseName);
          if (version = documentStates[versionId]?.local, version === void 0) return;
        }
        const draft = documentStates[getDraftId(documentId)]?.local, published = documentStates[getPublishedId(documentId)]?.local;
        if (draft === void 0 || published === void 0) return;
        document = version ?? draft ?? published;
      }
      if (!path) return document;
      const result = jsonMatch(document, path).next();
      if (result.done) return;
      const { value } = result.value;
      return value;
    },
    onSubscribe: (context, options) => manageSubscriberIds(context, [options])
  })
);
function resolveDocument(...args) {
  return _resolveDocument(...args);
}
const _resolveDocument = bindActionByResource(
  documentStore,
  ({ instance }, docHandle) => firstValueFrom(
    getDocumentState(instance, {
      ...docHandle,
      path: void 0
    }).observable.pipe(filter$1((i) => i !== void 0))
  )
), getDocumentSyncStatus = bindActionByResource(
  documentStore,
  createStateSourceAction({
    selector: ({ state: { error, documentStates: documents, outgoing, applied, queued } }, doc) => {
      const documentId = DocumentId(typeof doc == "string" ? doc : doc.documentId);
      if (error) throw error;
      if (doc.liveEdit) {
        if (documents[documentId] === void 0) return;
      } else {
        const version = isReleasePerspective(doc.perspective) ? documents[getVersionId(documentId, doc.perspective.releaseName)] : void 0;
        if (isReleasePerspective(doc.perspective) && version === void 0) return;
        const draft = documents[getDraftId(documentId)], published = documents[getPublishedId(documentId)];
        if (draft === void 0 || published === void 0) return;
      }
      return !queued.length && !applied.length && !outgoing;
    },
    onSubscribe: (context, doc) => manageSubscriberIds(context, [doc])
  })
), getPermissionsState = bindActionByResource(
  documentStore,
  createStateSourceAction({
    selector: calculatePermissions,
    onSubscribe: (context, { actions }) => {
      manageSubscriberIds(context, actions);
    }
  })
), resolvePermissions = bindActionByResource(
  documentStore,
  ({ instance }, options) => firstValueFrom(
    getPermissionsState(instance, options).observable.pipe(filter$1((i) => i !== void 0))
  )
), subscribeDocumentEvents = bindActionByResource(
  documentStore,
  ({ state }, options) => {
    const { events } = state.get(), subscription = events.subscribe(options.eventHandler);
    return () => subscription.unsubscribe();
  }
), subscribeToQueuedAndApplyNextTransaction = ({
  state
}) => {
  const { events } = state.get();
  return state.observable.pipe(
    map$1(applyFirstQueuedTransaction),
    distinctUntilChanged(),
    tap$1((next) => state.set("applyFirstQueuedTransaction", next)),
    catchError$1((error, caught) => {
      if (error instanceof ActionError)
        return state.set(
          "removeQueuedTransaction",
          (prev) => removeQueuedTransaction(prev, error.transactionId)
        ), events.next({
          type: "error",
          message: error.message,
          documentId: error.documentId,
          transactionId: error.transactionId,
          error
        }), caught;
      throw error;
    })
  ).subscribe({ error: (error) => state.set("setError", { error }) });
}, subscribeToAppliedAndSubmitNextTransaction = ({
  state,
  instance,
  key: { resource }
}) => {
  const { events } = state.get();
  return state.observable.pipe(
    throttle(
      (s) => (
        // if there is no outgoing transaction, we can throttle by the
        // initial outgoing throttle time…
        s.outgoing ? (
          // …otherwise, wait until the outgoing has been cleared
          state.observable.pipe(first$1(({ outgoing }) => !outgoing))
        ) : timer(INITIAL_OUTGOING_THROTTLE_TIME)
      ),
      { leading: !1, trailing: !0 }
    ),
    map$1(transitionAppliedTransactionsToOutgoing),
    distinctUntilChanged((a, b) => a.outgoing?.transactionId === b.outgoing?.transactionId),
    tap$1((next) => state.set("transitionAppliedTransactionsToOutgoing", next)),
    map$1((s) => s.outgoing),
    distinctUntilChanged(),
    withLatestFrom(
      getClientState(instance, {
        apiVersion: API_VERSION$4,
        resource
      }).observable
    ),
    concatMap(([outgoing, client]) => {
      if (!outgoing) return EMPTY;
      const revertOnError = catchError$1((error) => {
        state.set("revertOutgoingTransaction", revertOutgoingTransaction);
        const message = error instanceof Error ? error.message : "Request failed";
        return events.next({ type: "reverted", message, outgoing, error }), EMPTY;
      }), toResult = map$1((result) => ({
        result,
        outgoing
      }));
      return outgoing.actions.some((action) => action.liveEdit) ? client.observable.mutate(outgoing.outgoingMutations, {
        transactionId: outgoing.transactionId,
        visibility: "async",
        returnDocuments: !1,
        returnFirst: !1,
        tag: "document.mutate",
        skipCrossDatasetReferenceValidation: !0
      }).pipe(revertOnError, toResult) : client.observable.action(outgoing.outgoingActions, {
        transactionId: outgoing.transactionId,
        skipCrossDatasetReferenceValidation: !0,
        tag: "document.action"
      }).pipe(revertOnError, toResult);
    }),
    tap$1(({ outgoing, result }) => {
      state.set("cleanupOutgoingTransaction", cleanupOutgoingTransaction);
      for (const e of getDocumentEvents(outgoing)) events.next(e);
      events.next({ type: "accepted", outgoing, result });
    })
  ).subscribe({ error: (error) => state.set("setError", { error }) });
}, subscribeToSubscriptionsAndListenToDocuments = (context) => {
  const { state } = context, { events } = state.get();
  return state.observable.pipe(
    filter$1((s) => !!s.grants),
    map$1((s) => Object.keys(s.documentStates)),
    distinctUntilChanged((curr, next) => {
      if (curr.length !== next.length) return !1;
      const currSet = new Set(curr);
      return next.every((i) => currSet.has(i));
    }),
    startWith$1(/* @__PURE__ */ new Set()),
    pairwise$1(),
    switchMap((pair) => {
      const [curr, next] = pair.map((ids) => new Set(ids)), added = Array.from(next).filter((i) => !curr.has(i)), removed = Array.from(curr).filter((i) => !next.has(i)), changes = [
        ...added.map((id) => ({ id, add: !0 })),
        ...removed.map((id) => ({ id, add: !1 }))
      ].sort((a, b) => {
        const aIsDraft = a.id === getDraftId(DocumentId(a.id)), bIsDraft = b.id === getDraftId(DocumentId(b.id));
        return aIsDraft && bIsDraft ? a.id.localeCompare(b.id, "en-US") : aIsDraft ? -1 : bIsDraft ? 1 : a.id.localeCompare(b.id, "en-US");
      });
      return of(...changes);
    }),
    groupBy$1((i) => i.id),
    mergeMap$1(
      (group) => group.pipe(
        switchMap((e) => e.add ? listen(context, e.id).pipe(
          catchError$1((error) => {
            throw error instanceof OutOfSyncError && listen(context, e.id), error;
          }),
          tap$1(
            (remote) => state.set(
              "applyRemoteDocument",
              (prev) => applyRemoteDocument(prev, remote, events)
            )
          )
        ) : EMPTY)
      )
    )
  ).subscribe({ error: (error) => state.set("setError", { error }) });
}, subscribeToClientAndFetchDatasetAcl = ({
  instance,
  state,
  key: { resource }
}) => {
  const clientOptions = { apiVersion: API_VERSION$4, resource };
  let uri;
  if (resource && isDatasetResource(resource))
    uri = `/projects/${resource.projectId}/datasets/${resource.dataset}/acl`;
  else if (resource && isMediaLibraryResource(resource))
    uri = `/media-libraries/${resource.mediaLibraryId}/acl`;
  else if (resource && isCanvasResource(resource))
    uri = `/canvases/${resource.canvasId}/acl`;
  else
    throw new Error(`Received invalid resource: ${JSON.stringify(resource)}`);
  return getClientState(instance, clientOptions).observable.pipe(
    switchMap(
      (client) => client.observable.request({
        uri,
        tag: "acl.get",
        withCredentials: !0
      })
    ),
    tap$1((datasetAcl) => state.set("setGrants", { grants: createGrantsLookup(datasetAcl) }))
  ).subscribe({
    error: (error) => state.set("setError", { error })
  });
};
function applyDocumentActions(...args) {
  return boundApplyDocumentActions(...args);
}
const boundApplyDocumentActions = bindActionByResource(documentStore, _applyDocumentActions);
async function _applyDocumentActions({ state }, { actions, transactionId = crypto.randomUUID(), disableBatching }) {
  const { events } = state.get(), transaction = {
    transactionId,
    actions,
    ...disableBatching && { disableBatching }
  }, fatalError$ = state.observable.pipe(
    map$1((s) => s.error),
    first$1(Boolean),
    map$1((error) => ({ type: "error", error }))
  ), transactionError$ = events.pipe(
    filter$1((e) => e.type === "error"),
    first$1((e) => e.transactionId === transactionId)
  ), appliedTransaction$ = state.observable.pipe(
    map$1((s) => s.applied),
    distinctUntilChanged(),
    map$1((applied) => applied.find((t) => t.transactionId === transactionId)),
    first$1(Boolean)
  ), successfulTransaction$ = events.pipe(
    filter$1((e) => e.type === "accepted"),
    first$1((e) => e.outgoing.batchedTransactionIds.includes(transactionId))
  ), rejectedTransaction$ = events.pipe(
    filter$1((e) => e.type === "reverted"),
    first$1((e) => e.outgoing.batchedTransactionIds.includes(transactionId))
  ), appliedTransactionOrError = firstValueFrom(
    race([fatalError$, transactionError$, appliedTransaction$])
  ), acceptedOrRejectedTransaction = firstValueFrom(
    race([successfulTransaction$, rejectedTransaction$, transactionError$])
  );
  state.set("queueTransaction", (prev) => queueTransaction(prev, transaction));
  const result = await appliedTransactionOrError;
  if ("type" in result && result.type === "error") throw result.error;
  const { working: documents, previous, previousRevs } = result, existingIds = new Set(
    Object.entries(previous).filter(([, value]) => !!value).map(([key]) => key)
  ), resultingIds = new Set(
    Object.entries(documents).filter(([, value]) => !!value).map(([key]) => key)
  ), allIds = /* @__PURE__ */ new Set([...existingIds, ...resultingIds]), updated = [], appeared = [], disappeared = [];
  for (const id of allIds)
    existingIds.has(id) && resultingIds.has(id) ? updated.push(id) : !existingIds.has(id) && resultingIds.has(id) ? appeared.push(id) : !resultingIds.has(id) && existingIds.has(id) && disappeared.push(id);
  async function submitted() {
    const raceResult = await acceptedOrRejectedTransaction;
    if (raceResult.type !== "accepted") throw raceResult.error;
    return raceResult.result;
  }
  return {
    transactionId,
    documents,
    previous,
    previousRevs,
    appeared,
    updated,
    disappeared,
    submitted
  };
}
function createFavoriteKey(context) {
  return `${context.documentId}:${context.documentType}:${context.resourceId}:${context.resourceType}${context.schemaName ? `:${context.schemaName}` : ""}`;
}
const favorites = createFetcherStore({
  name: "Favorites",
  getKey: (_instance, context) => createFavoriteKey(context),
  fetcher: (instance) => (context) => {
    const nodeStateSource = getNodeState(instance, {
      name: SDK_NODE_NAME,
      connectTo: SDK_CHANNEL_NAME
    }), payload = {
      document: {
        id: context.documentId,
        type: context.documentType,
        resource: {
          id: context.resourceId,
          type: context.resourceType,
          schemaName: context.schemaName
        }
      }
    };
    return nodeStateSource.observable.pipe(
      filter$1((nodeState) => !!nodeState),
      // Only proceed when connected
      shareReplay(1),
      switchMap((nodeState) => {
        const node = nodeState.node;
        return from(
          node.fetch(
            // @ts-expect-error -- getOrCreateNode should be refactored to take type arguments
            "dashboard/v1/events/favorite/query",
            payload
          )
        ).pipe(
          map$1((response) => ({ isFavorited: response.isFavorited })),
          catchError$1((err) => (console.error("Favorites service connection error", err), of({ isFavorited: !1 })))
        );
      })
    );
  }
}), getFavoritesState = favorites.getState, resolveFavoritesState = favorites.resolveState, API_VERSION$2 = "v2025-02-19";
function resolveOrganizationId(options) {
  const organizationId = options?.organizationId;
  if (!organizationId)
    throw new Error("An organizationId is required to use the organization API.");
  return organizationId;
}
function normalizeOrganizationOptions$1(options) {
  return {
    includeMembers: options?.includeMembers ?? !1,
    includeFeatures: options?.includeFeatures ?? !1
  };
}
function getOrganizationCacheKey(_instance, options) {
  const organizationId = resolveOrganizationId(options), { includeMembers, includeFeatures } = normalizeOrganizationOptions$1(options);
  return `organization:${organizationId}${includeMembers ? ":members" : ""}${includeFeatures ? ":features" : ""}`;
}
const organization = createFetcherStore({
  name: "Organization",
  getKey: getOrganizationCacheKey,
  fetcher: (instance) => (options) => {
    const organizationId = resolveOrganizationId(options);
    return getClientState(instance, {
      apiVersion: API_VERSION$2,
      scope: "global"
    }).observable.pipe(
      switchMap((client) => {
        const normalized = normalizeOrganizationOptions$1(options), query = Object.fromEntries(
          Object.entries(normalized).filter(([, value]) => value !== void 0).map(([key, value]) => [key, String(value)])
        );
        return client.observable.request({
          uri: `/organizations/${organizationId}`,
          query,
          tag: "organization.get"
        });
      })
    );
  }
}), getOrganizationState = organization.getState, resolveOrganization = organization.resolveState, API_VERSION$1 = "v2025-02-19";
function normalizeOrganizationOptions(options) {
  return {
    includeImplicitMemberships: options?.includeImplicitMemberships ?? !1,
    includeMembers: options?.includeMembers ?? !1,
    includeFeatures: options?.includeFeatures ?? !1
  };
}
function getOrganizationsCacheKey(_instance, options) {
  const { includeMembers, includeFeatures, includeImplicitMemberships } = normalizeOrganizationOptions(options);
  return `organizations${includeMembers ? ":members" : ""}${includeFeatures ? ":features" : ""}${includeImplicitMemberships ? ":implicit" : ""}`;
}
const organizations = createFetcherStore({
  name: "Organizations",
  getKey: getOrganizationsCacheKey,
  fetcher: (instance) => (options) => getClientState(instance, {
    apiVersion: API_VERSION$1,
    scope: "global"
  }).observable.pipe(
    switchMap((client) => {
      const normalized = normalizeOrganizationOptions(options), query = Object.fromEntries(
        Object.entries(normalized).filter(([, value]) => value !== void 0).map(([key, value]) => [key, String(value)])
      );
      return client.observable.request({
        uri: "/organizations",
        query,
        tag: "organizations.get"
      });
    })
  )
}), getOrganizationsState = organizations.getState, resolveOrganizations = organizations.resolveState, usersStore = defineStore({
  name: "UsersStore",
  getInitialState: () => ({ users: {} }),
  initialize: (context) => {
    const subscription = listenForLoadMoreAndFetch(context);
    return () => subscription.unsubscribe();
  }
}), errorHandler = (state) => (error) => state.set("setError", { error }), listenForLoadMoreAndFetch = ({ state, instance }) => state.observable.pipe(
  map$1((s) => new Set(Object.keys(s.users))),
  distinctUntilChanged((curr, next) => curr.size !== next.size ? !1 : Array.from(next).every((i) => curr.has(i))),
  startWith$1(/* @__PURE__ */ new Set()),
  pairwise$1(),
  mergeMap$1(([curr, next]) => {
    const added = Array.from(next).filter((i) => !curr.has(i)), removed = Array.from(curr).filter((i) => !next.has(i));
    return [
      ...added.map((key) => ({ key, added: !0 })),
      ...removed.map((key) => ({ key, added: !1 }))
    ];
  }),
  groupBy$1((i) => i.key),
  mergeMap$1(
    (group$) => group$.pipe(
      switchMap((e) => {
        if (!e.added) return EMPTY;
        const { userId, batchSize, ...options } = parseUsersKey(group$.key);
        if (userId) {
          if (userId.startsWith("p"))
            return getClient(instance, {
              apiVersion: PROJECT_API_VERSION,
              // this is a global store, so we need to use the projectId from the options when we're fetching
              // users from a project subdomain
              projectId: options.projectId,
              useProjectHostname: !0
            }).observable.request({
              method: "GET",
              uri: `/users/${userId}`,
              tag: "users.get"
            }).pipe(
              map$1((user) => ({
                data: [{
                  sanityUserId: user.sanityUserId,
                  profile: {
                    id: user.id,
                    displayName: user.displayName,
                    familyName: user.familyName ?? void 0,
                    givenName: user.givenName ?? void 0,
                    middleName: user.middleName ?? void 0,
                    imageUrl: user.imageUrl ?? void 0,
                    createdAt: user.createdAt,
                    updatedAt: user.updatedAt,
                    isCurrentUser: user.isCurrentUser,
                    email: user.email,
                    provider: user.provider
                  },
                  memberships: []
                }],
                totalCount: 1,
                nextCursor: null
              })),
              catchError$1((error) => (state.set("setUsersError", setUsersError(group$.key, error)), EMPTY)),
              tap$1(
                (response) => state.set("setUsersData", setUsersData(group$.key, response))
              )
            );
          const scope = userId.startsWith("g") ? "global" : void 0, client = getClient(instance, {
            scope,
            apiVersion: API_VERSION$8
          }), resourceType2 = options.resourceType || "project", resourceId = resourceType2 === "organization" ? options.organizationId : options.projectId;
          return resourceId ? client.observable.request({
            method: "GET",
            uri: `access/${resourceType2}/${resourceId}/users/${userId}`,
            tag: "users.get"
          }).pipe(
            map$1((response) => "sanityUserId" in response ? {
              data: [response],
              totalCount: 1,
              nextCursor: null
            } : response),
            catchError$1((error) => (state.set("setUsersError", setUsersError(group$.key, error)), EMPTY)),
            tap$1((response) => state.set("setUsersData", setUsersData(group$.key, response)))
          ) : throwError(() => new Error("An organizationId or a projectId is required"));
        }
        const projectId = options.projectId, resourceType = options.resourceType ?? (options.organizationId ? "organization" : projectId ? "project" : "organization"), organizationId$ = options.organizationId ? of(options.organizationId) : getDashboardOrganizationId$1(instance).observable.pipe(
          filter$1((i) => typeof i == "string")
        ), resource$ = resourceType === "project" ? projectId ? of({ type: "project", id: projectId }) : throwError(() => new Error("Project ID required for this API.")) : organizationId$.pipe(map$1((id) => ({ type: "organization", id }))), client$ = getClientState(instance, {
          scope: "global",
          apiVersion: API_VERSION$8
        }).observable, loadMore$ = state.observable.pipe(
          map$1((s) => s.users[group$.key]?.lastLoadMoreRequest),
          distinctUntilChanged()
        ), cursor$ = state.observable.pipe(
          map$1((s) => s.users[group$.key]?.nextCursor),
          distinctUntilChanged(),
          filter$1((cursor) => cursor !== null)
        );
        return combineLatest([resource$, client$, loadMore$]).pipe(
          withLatestFrom(cursor$),
          switchMap(
            ([[resource, client], cursor]) => client.observable.request({
              method: "GET",
              uri: `access/${resource.type}/${resource.id}/users`,
              tag: "users.list",
              query: cursor ? { nextCursor: cursor, limit: batchSize.toString() } : { limit: batchSize.toString() }
            })
          ),
          catchError$1((error) => (state.set("setUsersError", setUsersError(group$.key, error)), EMPTY)),
          tap$1((response) => state.set("setUsersData", setUsersData(group$.key, response)))
        );
      })
    )
  )
).subscribe({ error: errorHandler(state) }), getUsersState = bindActionGlobally(
  usersStore,
  createStateSourceAction({
    selector: createSelector(
      [
        ({ instance, state }, options) => state.error ?? state.users[getUsersKey(instance, options)]?.error,
        ({ instance, state }, options) => state.users[getUsersKey(instance, options)]?.users,
        ({ instance, state }, options) => state.users[getUsersKey(instance, options)]?.totalCount,
        ({ instance, state }, options) => state.users[getUsersKey(instance, options)]?.nextCursor
      ],
      (error, data, totalCount, nextCursor) => {
        if (error) throw error;
        if (!(data === void 0 || totalCount === void 0 || nextCursor === void 0))
          return { data, totalCount, hasMore: nextCursor !== null };
      }
    ),
    onSubscribe: ({ instance, state }, options) => {
      const subscriptionId = insecureRandomId(), key = getUsersKey(instance, options);
      return state.set("addSubscription", addSubscription(subscriptionId, key)), () => {
        setCleanupTimeout(
          () => state.set("removeSubscription", removeSubscription(subscriptionId, key)),
          USERS_STATE_CLEAR_DELAY
        );
      };
    }
  })
), resolveUsers = bindActionGlobally(
  usersStore,
  async ({ state, instance }, { signal, ...options }) => {
    const key = getUsersKey(instance, options), { getCurrent } = getUsersState(instance, options), aborted$ = signal ? new Observable((observer) => {
      const cleanup = () => {
        signal.removeEventListener("abort", listener);
      }, listener = () => {
        observer.error(new DOMException("The operation was aborted.", "AbortError")), observer.complete(), cleanup();
      };
      return signal.addEventListener("abort", listener), cleanup;
    }).pipe(
      catchError$1((error) => {
        throw error instanceof Error && error.name === "AbortError" && state.set("cancelRequest", cancelRequest(key)), error;
      })
    ) : NEVER;
    state.set("initializeRequest", initializeRequest(key));
    const resolved$ = state.observable.pipe(
      map$1(getCurrent),
      first$1((i) => i !== void 0)
    );
    return firstValueFrom(race([resolved$, aborted$]));
  }
), loadMoreUsers = bindActionGlobally(
  usersStore,
  async ({ state, instance }, options) => {
    const key = getUsersKey(instance, options), users = getUsersState(instance, options), usersState = users.getCurrent();
    if (!usersState)
      throw new Error("Users not loaded for specified resource. Please call resolveUsers first.");
    if (!usersState.hasMore)
      throw new Error("No more users available to load for this resource.");
    const promise = firstValueFrom(
      users.observable.pipe(
        filter$1((i) => i !== void 0),
        skip(1)
      )
    ), timestamp = (/* @__PURE__ */ new Date()).toISOString();
    return state.set("updateLastLoadMoreRequest", updateLastLoadMoreRequest(timestamp, key)), await promise;
  }
), getUserState = bindActionGlobally(
  usersStore,
  ({ instance }, { userId, ...options }) => getUsersState(instance, { userId, ...options }).observable.pipe(
    map$1((res) => res?.data[0]),
    distinctUntilChanged((a, b) => a?.profile.updatedAt === b?.profile.updatedAt)
  )
), resolveUser = bindActionGlobally(
  usersStore,
  async ({ instance }, { signal, ...options }) => (await resolveUsers(instance, {
    signal,
    ...options
  }))?.data[0]
);
function getBifurClient(client, token$) {
  const bifurVersionedClient = client.withConfig({ apiVersion: "2022-06-30" }), {
    resource,
    dataset,
    url: baseUrl,
    requestTagPrefix = "sanity.sdk.presence"
  } = bifurVersionedClient.config();
  let resourcePath;
  if (resource?.type === "canvas")
    resourcePath = `canvases/${resource.id}`;
  else if (dataset)
    resourcePath = dataset;
  else
    throw new Error("Unable to determine presence URL: no canvas resource or dataset configured");
  const urlWithTag = `${`${baseUrl}/socket/${resourcePath}`.replace(/^http/, "ws")}?tag=${requestTagPrefix}`;
  return fromUrl(urlWithTag, { token$ });
}
const handleIncomingMessage = (event) => {
  switch (event.type) {
    case "rollCall":
      return {
        type: "rollCall",
        userId: event.i,
        sessionId: event.session
      };
    case "state": {
      const { sessionId, locations } = event.m;
      return {
        type: "state",
        userId: event.i,
        sessionId,
        timestamp: (/* @__PURE__ */ new Date()).toISOString(),
        locations
      };
    }
    case "disconnect":
      return {
        type: "disconnect",
        userId: event.i,
        sessionId: event.m.session,
        timestamp: (/* @__PURE__ */ new Date()).toISOString()
      };
    default:
      throw new Error(`Got unknown presence event: ${JSON.stringify(event)}`);
  }
}, createBifurTransport = (options) => {
  const { client, token$, sessionId } = options, bifur = getBifurClient(client, token$), incomingEvents$ = bifur.listen("presence").pipe(map(handleIncomingMessage)), dispatchMessage = (message) => {
    switch (message.type) {
      case "rollCall":
        return bifur.request("presence_rollcall", { session: sessionId });
      case "state":
        return bifur.request("presence_announce", {
          data: { locations: message.locations, sessionId }
        });
      case "disconnect":
        return bifur.request("presence_disconnect", { session: sessionId });
      default:
        return EMPTY;
    }
  };
  return typeof window < "u" && fromEvent(window, "beforeunload").pipe(switchMap$1(() => dispatchMessage({ type: "disconnect" }))).subscribe(), [incomingEvents$.pipe(share$1()), dispatchMessage];
}, PRESENCE_API_VERSION = "2026-03-30", getInitialState = () => ({
  locations: /* @__PURE__ */ new Map(),
  users: {}
}), presenceStore = defineStore({
  name: "presence",
  getInitialState,
  initialize: (context) => {
    const {
      instance,
      state,
      key: { resource }
    } = context;
    if (isMediaLibraryResource(resource))
      throw new Error("Presence is not supported for media library resources.");
    const sessionId = crypto.randomUUID(), client = isDatasetResource(resource) ? getClient(instance, {
      apiVersion: PRESENCE_API_VERSION,
      projectId: resource.projectId,
      dataset: resource.dataset,
      useProjectHostname: !0
    }) : getClient(instance, {
      apiVersion: PRESENCE_API_VERSION,
      resource
    }), token$ = getTokenState(instance).observable.pipe(distinctUntilChanged()), [incomingEvents$, dispatch] = createBifurTransport({
      client,
      token$,
      sessionId
    }), subscription = new Subscription();
    if (subscription.add(
      incomingEvents$.subscribe((event) => {
        "sessionId" in event && event.sessionId === sessionId || (event.type === "state" ? state.set("presence/state", (prevState) => {
          const newLocations = new Map(prevState.locations);
          return newLocations.set(event.sessionId, {
            userId: event.userId,
            locations: event.locations
          }), {
            ...prevState,
            locations: newLocations
          };
        }) : event.type === "disconnect" && state.set("presence/disconnect", (prevState) => {
          const newLocations = new Map(prevState.locations);
          return newLocations.delete(event.sessionId), { ...prevState, locations: newLocations };
        }));
      })
    ), dispatch({ type: "rollCall" }).subscribe(), isCanvasResource(resource)) {
      const globalClient = getClient(instance, { apiVersion: PRESENCE_API_VERSION });
      subscription.add(
        globalClient.observable.request({
          uri: `/canvases/${resource.canvasId}`,
          tag: "canvases.get"
        }).pipe(catchError$1(() => EMPTY)).subscribe(({ organizationId }) => {
          state.set("presence/organizationId", (prev) => ({ ...prev, organizationId }));
        })
      );
    }
    return () => {
      dispatch({ type: "disconnect" }).subscribe(), subscription.unsubscribe();
    };
  }
}), selectLocations = (state) => state.locations, selectUsers = (state) => state.users, selectPresence = createSelector(
  selectLocations,
  selectUsers,
  (locations, users) => Array.from(locations.entries()).map(([sessionId, { userId, locations: locs }]) => ({
    user: users[userId] || {
      id: userId,
      displayName: "Unknown user",
      name: "Unknown user",
      email: ""
    },
    sessionId,
    locations: locs
  }))
), _getPresence = bindActionByResource(
  presenceStore,
  createStateSourceAction({
    selector: (context) => selectPresence(context.state),
    onSubscribe: (context) => {
      const resource = context.key.resource, userIds$ = context.state.observable.pipe(
        map$1(
          (state) => Array.from(state.locations.values()).map((l) => l.userId).filter((id) => !!id)
        ),
        distinctUntilChanged((a, b) => a.length === b.length && a.every((v, i) => v === b[i]))
      ), organizationId$ = isCanvasResource(resource) ? context.state.observable.pipe(
        map$1((s) => s.organizationId),
        filter$1((id) => id !== void 0),
        first$1()
      ) : of(void 0), subscription = combineLatest([userIds$, organizationId$]).pipe(
        switchMap(([userIds, organizationId]) => {
          if (userIds.length === 0)
            return of([]);
          const userObservables = userIds.map(
            (userId) => getUserState(context.instance, {
              userId,
              ...isDatasetResource(resource) ? { resourceType: "project", projectId: resource.projectId } : { resourceType: "organization", organizationId }
            }).pipe(filter$1((v) => !!v))
          );
          return combineLatest(userObservables);
        })
      ).subscribe((users) => {
        context.state.set("presence/users", (prevState) => ({
          ...prevState,
          users: {
            ...prevState.users,
            ...users.reduce((acc, user) => (user && (acc[user.profile.id] = user), acc), {})
          }
        }));
      });
      return () => subscription.unsubscribe();
    }
  })
);
function getPresence(instance, params) {
  return _getPresence(instance, params ?? {});
}
function hashString(str) {
  let hash = 0;
  for (let i = 0; i < str.length; i++)
    hash = (hash * 31 + str.charCodeAt(i)) % 2147483647;
  return Math.abs(hash).toString(16).padStart(8, "0");
}
const PROJECTION_TAG = "projection", PROJECTION_STATE_CLEAR_DELAY = 1e3, STABLE_EMPTY_PROJECTION = {
  data: null,
  isPending: !1
};
function validateProjection(projection) {
  if (!projection.startsWith("{") || !projection.endsWith("}"))
    throw new Error(
      `Invalid projection format: "${projection}". Projections must be enclosed in curly braces, e.g. "{title, 'author': author.name}"`
    );
  return projection;
}
function createProjectionQuery(documentIds, documentProjections) {
  const projections = Array.from(documentIds).flatMap((id) => {
    const projectionsForDoc = documentProjections[id];
    return projectionsForDoc ? Object.entries(projectionsForDoc).map(([projectionHash, projection]) => ({
      documentId: id,
      projection: validateProjection(projection),
      projectionHash
    })) : [];
  }).reduce((acc, { documentId, projection, projectionHash }) => {
    const obj = acc[projectionHash] ?? { documentIds: /* @__PURE__ */ new Set(), projection };
    return obj.documentIds.add(documentId), acc[projectionHash] = obj, acc;
  }, {}), query = `[${Object.entries(projections).map(([projectionHash, { projection }]) => `...*[_id in $__ids_${projectionHash}]{_id,_type,_updatedAt,"__projectionHash":"${projectionHash}","result":{...${projection}}}`).join(",")}]`, params = Object.fromEntries(
    Object.entries(projections).map(([projectionHash, value]) => {
      const idsInProjection = Array.from(value.documentIds).flatMap((id) => DocumentId(id));
      return [`__ids_${projectionHash}`, Array.from(idsInProjection)];
    })
  );
  return { query, params };
}
function processProjectionQuery({
  ids,
  results,
  documentStatuses
}) {
  const groupedResults = {};
  for (const result of results) {
    const originalId = getPublishedId(DocumentId(result._id)), hash = result.__projectionHash;
    ids.has(originalId) && (groupedResults[originalId] || (groupedResults[originalId] = {}), groupedResults[originalId][hash] || (groupedResults[originalId][hash] = void 0), groupedResults[originalId][hash] = result);
  }
  const finalValues = {};
  for (const originalId of ids) {
    finalValues[originalId] = {};
    const projectionsForDoc = groupedResults[originalId];
    if (projectionsForDoc)
      for (const hash in projectionsForDoc) {
        const projectionResultData = projectionsForDoc[hash]?.result;
        if (!projectionResultData) {
          finalValues[originalId][hash] = { data: null, isPending: !1 };
          continue;
        }
        const statusFromStore = documentStatuses?.[originalId];
        finalValues[originalId][hash] = {
          data: { ...projectionResultData, _status: statusFromStore },
          isPending: !1
        };
      }
  }
  return finalValues;
}
function buildStatusQueryIds(documentIds, perspective) {
  const ids = [], releaseName = isReleasePerspective(perspective) ? perspective.releaseName : null;
  for (const id of documentIds) {
    const publishedId = getPublishedId(DocumentId(id)), draftId = getDraftId(publishedId);
    ids.push(draftId, publishedId), releaseName && ids.push(getVersionId(publishedId, releaseName));
  }
  return ids;
}
function processStatusQueryResults(results) {
  const documentStatuses = {};
  for (const result of results) {
    const id = DocumentId(result._id), updatedAt = result._updatedAt, publishedId = getPublishedId(id), statusData = documentStatuses[publishedId] ?? {};
    isDraftId(id) ? statusData.lastEditedDraftAt = updatedAt : isVersionId(id) ? statusData.lastEditedVersionAt = updatedAt : isPublishedId(id) && (statusData.lastEditedPublishedAt = updatedAt), documentStatuses[publishedId] = statusData;
  }
  return documentStatuses;
}
const BATCH_DEBOUNCE_TIME = 50, isSetEqual = (a, b) => a.size === b.size && Array.from(a).every((i) => b.has(i)), subscribeToStateAndFetchBatches = ({
  state,
  instance,
  key: { resource, perspective }
}) => {
  const documentProjections$ = state.observable.pipe(
    map$1((s) => s.documentProjections),
    distinctUntilChanged(isDeepEqual)
  ), activeDocumentIds$ = state.observable.pipe(
    map$1(({ subscriptions }) => new Set(Object.keys(subscriptions).map((id) => DocumentId(id)))),
    distinctUntilChanged(isSetEqual)
  ), pendingUpdateSubscription = activeDocumentIds$.pipe(
    debounceTime(BATCH_DEBOUNCE_TIME),
    startWith$1(/* @__PURE__ */ new Set()),
    pairwise$1(),
    tap$1(([prevIds, currIds]) => {
      const newIds = [...currIds].filter((id) => !prevIds.has(id));
      newIds.length !== 0 && state.set("updatingPending", (prev) => {
        const nextValues = { ...prev.values };
        for (const id of newIds) {
          const projectionsForDoc = prev.documentProjections[id];
          if (!projectionsForDoc) continue;
          const updatedValuesForDoc = { ...prev.values[id] ?? {} };
          for (const hash in projectionsForDoc) {
            const currentValue = updatedValuesForDoc[hash];
            updatedValuesForDoc[hash] = {
              data: currentValue?.data ?? null,
              isPending: !0
            };
          }
          nextValues[id] = updatedValuesForDoc;
        }
        return { values: nextValues };
      });
    })
  ).subscribe(), queryExecutionSubscription = combineLatest([activeDocumentIds$, documentProjections$]).pipe(
    debounceTime(BATCH_DEBOUNCE_TIME),
    distinctUntilChanged(isDeepEqual)
  ).pipe(
    switchMap(([ids, documentProjections]) => {
      if (!ids.size) return EMPTY;
      const { query, params } = createProjectionQuery(ids, documentProjections), controller = new AbortController(), statusQueryIds = buildStatusQueryIds(ids, perspective), statusQuery = "*[_id in $statusIds]{_id, _updatedAt}", statusParams = { statusIds: statusQueryIds }, projectionQuery$ = new Observable((observer) => {
        const { getCurrent, observable } = getQueryState(instance, {
          query,
          params,
          tag: PROJECTION_TAG,
          perspective,
          resource
        }), subscription = defer(() => getCurrent() === void 0 ? from(
          resolveQuery(instance, {
            query,
            params,
            tag: PROJECTION_TAG,
            signal: controller.signal,
            perspective,
            resource
          })
        ).pipe(switchMap(() => observable)) : observable).pipe(filter$1((result) => result !== void 0)).subscribe(observer);
        return () => {
          controller.signal.aborted || controller.abort(), subscription.unsubscribe();
        };
      }), statusQuery$ = new Observable((observer) => {
        const { getCurrent, observable } = getQueryState(instance, {
          query: statusQuery,
          params: statusParams,
          tag: PROJECTION_TAG,
          perspective: "raw",
          resource
        }), subscription = defer(() => getCurrent() === void 0 ? from(
          resolveQuery(instance, {
            query: statusQuery,
            params: statusParams,
            tag: PROJECTION_TAG,
            signal: controller.signal,
            perspective: "raw",
            resource
          })
        ).pipe(switchMap(() => observable)) : observable).pipe(filter$1((result) => result !== void 0)).subscribe(observer);
        return () => {
          subscription.unsubscribe();
        };
      });
      return combineLatest([projectionQuery$, statusQuery$]).pipe(
        filter$1(
          (pair) => pair[0] !== void 0 && pair[1] !== void 0
        ),
        map$1(([projection, status]) => ({
          data: projection,
          ids,
          statusResults: status
        }))
      );
    }),
    map$1(({ ids, data, statusResults }) => {
      const documentStatuses = processStatusQueryResults(statusResults);
      state.set("updateStatuses", (prev) => ({
        documentStatuses: {
          ...prev.documentStatuses,
          ...documentStatuses
        }
      }));
      const currentState = state.get();
      return processProjectionQuery({
        ids,
        results: data,
        documentStatuses: currentState.documentStatuses
      });
    })
  ).subscribe({
    next: (processedValues) => {
      state.set("updateResult", (prev) => {
        const nextValues = { ...prev.values };
        for (const docId in processedValues)
          processedValues[docId] && (nextValues[docId] = {
            ...prev.values[docId] ?? {},
            ...processedValues[docId]
          });
        return { values: nextValues };
      });
    },
    error: (err) => {
      console.error("Error fetching projection batches:", err);
    }
  });
  return new Subscription(() => {
    pendingUpdateSubscription.unsubscribe(), queryExecutionSubscription.unsubscribe();
  });
}, projectionStore = defineStore({
  name: "Projection",
  getInitialState() {
    return {
      values: {},
      documentProjections: {},
      subscriptions: {},
      documentStatuses: {}
    };
  },
  initialize(context) {
    const batchSubscription = subscribeToStateAndFetchBatches(context);
    return () => batchSubscription.unsubscribe();
  }
});
function getProjectionState(...args) {
  return _getProjectionState(...args);
}
const _getProjectionState = bindActionByResourceAndPerspective(
  projectionStore,
  createStateSourceAction({
    selector: ({ state }, options) => {
      const documentId = getPublishedId(DocumentId(options.documentId)), projectionHash = hashString(options.projection);
      return state.values[documentId]?.[projectionHash] ?? STABLE_EMPTY_PROJECTION;
    },
    onSubscribe: ({ state }, options) => {
      const { projection, ...docHandle } = options, subscriptionId = insecureRandomId(), documentId = getPublishedId(DocumentId(docHandle.documentId)), validProjection = validateProjection(projection), projectionHash = hashString(validProjection);
      return state.set("addSubscription", (prev) => ({
        documentProjections: {
          ...prev.documentProjections,
          [documentId]: {
            ...prev.documentProjections[documentId],
            [projectionHash]: validProjection
          }
        },
        subscriptions: {
          ...prev.subscriptions,
          [documentId]: {
            ...prev.subscriptions[documentId],
            [projectionHash]: {
              ...prev.subscriptions[documentId]?.[projectionHash],
              [subscriptionId]: !0
            }
          }
        }
      })), () => {
        setCleanupTimeout(() => {
          state.set("removeSubscription", (prev) => {
            const documentSubscriptionsForHash = omitProperty(
              prev.subscriptions[documentId]?.[projectionHash],
              subscriptionId
            ), hasSubscribersForProjection = !!Object.keys(documentSubscriptionsForHash).length, nextSubscriptions = { ...prev.subscriptions }, nextDocumentProjections = { ...prev.documentProjections }, nextValues = { ...prev.values };
            if (hasSubscribersForProjection)
              nextSubscriptions[documentId] && (nextSubscriptions[documentId][projectionHash] = documentSubscriptionsForHash);
            else {
              delete nextSubscriptions[documentId][projectionHash], delete nextDocumentProjections[documentId][projectionHash];
              const currentProjectionValue = prev.values[documentId]?.[projectionHash];
              currentProjectionValue && nextValues[documentId] && (nextValues[documentId][projectionHash] = {
                data: currentProjectionValue.data,
                isPending: !1
              });
            }
            return Object.values(
              nextSubscriptions[documentId] ?? {}
            ).some((subs) => Object.keys(subs).length > 0) || (delete nextSubscriptions[documentId], delete nextDocumentProjections[documentId]), {
              subscriptions: nextSubscriptions,
              documentProjections: nextDocumentProjections,
              values: nextValues
            };
          });
        }, PROJECTION_STATE_CLEAR_DELAY);
      };
    }
  })
);
function getPreviewState(instance, options) {
  const projectionState = getProjectionState(instance, {
    ...options,
    projection: PREVIEW_PROJECTION
  }), transformResult = (current) => !current || current.data === null ? { data: null, isPending: current?.isPending ?? !1 } : {
    data: transformProjectionToPreview(instance, current.data, options.resource),
    isPending: current.isPending
  };
  return {
    getCurrent: () => transformResult(projectionState.getCurrent()),
    subscribe: (callback) => projectionState.subscribe(callback),
    observable: projectionState.observable.pipe(map$1(transformResult))
  };
}
function resolveProjection(...args) {
  return _resolveProjection(...args);
}
const _resolveProjection = bindActionByResourceAndPerspective(
  projectionStore,
  ({ instance }, options) => firstValueFrom(
    getProjectionState(instance, options).observable.pipe(
      filter$1((state) => !!state?.data)
    )
  )
);
async function resolvePreview(instance, options) {
  const projectionResult = await resolveProjection(instance, {
    ...options,
    projection: PREVIEW_PROJECTION
  });
  return projectionResult.data ? {
    data: transformProjectionToPreview(
      instance,
      projectionResult.data,
      options.resource
    ),
    isPending: projectionResult.isPending
  } : { data: null, isPending: projectionResult.isPending };
}
const API_VERSION = "v2025-02-19";
function normalizeProjectsOptions(options) {
  return {
    organizationId: options?.organizationId,
    includeMembers: options?.includeMembers ?? !1,
    includeFeatures: options?.includeFeatures ?? !0,
    onlyExplicitMembership: options?.onlyExplicitMembership ?? !1
  };
}
function getProjectsCacheKey(_instance, options) {
  const { organizationId, includeMembers, includeFeatures, onlyExplicitMembership } = normalizeProjectsOptions(options);
  return `projects${organizationId ? `:org:${organizationId}` : ""}${includeMembers ? ":members" : ""}${includeFeatures ? ":features" : ""}${onlyExplicitMembership ? ":explicit" : ""}`;
}
const projects = createFetcherStore({
  name: "Projects",
  getKey: getProjectsCacheKey,
  fetcher: (instance) => (options) => getClientState(instance, {
    apiVersion: API_VERSION,
    scope: "global"
  }).observable.pipe(
    switchMap((client) => {
      const normalized = normalizeProjectsOptions(options), query = Object.fromEntries(
        Object.entries(normalized).filter(([, value]) => value !== void 0).map(([key, value]) => [key, String(value)])
      );
      return client.observable.request({
        uri: "/projects",
        query,
        tag: "projects.get"
      });
    })
  )
}), getProjectsState = projects.getState, resolveProjects = projects.resolveState;
function defineIntent(intent) {
  if (!intent.id)
    throw new Error("Intent must have an id");
  if (!intent.action)
    throw new Error("Intent must have an action");
  if (!intent.title)
    throw new Error("Intent must have a title");
  if (!Array.isArray(intent.filters))
    throw new Error("Intent must have a filters array");
  if (intent.filters.length === 0)
    throw new Error(
      "Intent must have at least one filter. If you want to match everything, use {types: ['*']}"
    );
  return intent.filters.forEach((filter2, index) => {
    validateFilter(filter2, index);
  }), intent;
}
function validateFilter(filter2, index) {
  const filterContext = `Filter at index ${index}`;
  if (!filter2 || typeof filter2 != "object")
    throw new Error(`${filterContext} must be an object`);
  if (filter2.types === void 0)
    throw new Error(
      `${filterContext} must have a types property. Use ['*'] to match all document types.`
    );
  if (filter2.projectId !== void 0) {
    if (typeof filter2.projectId != "string")
      throw new Error(`${filterContext}: projectId must be a string`);
    if (filter2.projectId.trim() === "")
      throw new Error(`${filterContext}: projectId cannot be empty`);
  }
  if (filter2.dataset !== void 0) {
    if (typeof filter2.dataset != "string")
      throw new Error(`${filterContext}: dataset must be a string`);
    if (filter2.dataset.trim() === "")
      throw new Error(`${filterContext}: dataset cannot be empty`);
    if (filter2.projectId === void 0)
      throw new Error(`${filterContext}: dataset cannot be specified without projectId`);
  }
  if (!Array.isArray(filter2.types))
    throw new Error(`${filterContext}: types must be an array`);
  if (filter2.types.length === 0)
    throw new Error(`${filterContext}: types array cannot be empty`);
  if (filter2.types.forEach((type, typeIndex) => {
    if (typeof type != "string")
      throw new Error(`${filterContext}: types[${typeIndex}] must be a string`);
    if (type.trim() === "")
      throw new Error(`${filterContext}: types[${typeIndex}] cannot be empty`);
  }), filter2.types.includes("*") && filter2.types.length > 1)
    throw new Error(
      `${filterContext}: when using wildcard '*', it must be the only type in the array`
    );
}
function getCorsErrorProjectId(error) {
  if (!(error instanceof CorsOriginError)) return null;
  const projMatch = (error.message || "").match(/manage\/project\/([^/?#]+)/);
  return projMatch ? projMatch[1] : null;
}
function isImportError(error) {
  if (!(error instanceof Error)) return !1;
  if (error.name === "ChunkLoadError") return !0;
  const message = error.message || "";
  return /Loading chunk [\w-]+ failed/i.test(message) || /Failed to fetch dynamically imported module/i.test(message) || /error loading dynamically imported module/i.test(message) || /Import(?:ing)? a module script failed/i.test(message) || /Unable to preload CSS/i.test(message);
}
export {
  AuthStateType,
  CORE_SDK_VERSION,
  PREVIEW_PROJECTION,
  agentGenerate,
  agentPatch,
  agentPrompt,
  agentTransform,
  agentTranslate,
  applyDocumentActions,
  configureLogging,
  createDatasetHandle,
  createDocument,
  createDocumentHandle,
  createDocumentTypeHandle,
  createGroqSearchFilter,
  createProjectHandle,
  createSanityInstance,
  defineIntent,
  deleteDocument,
  destroyController,
  discardDocument,
  editDocument,
  getActiveReleasesState,
  getAuthState,
  getClient,
  getClientErrorApiBody,
  getClientErrorApiDescription,
  getClientErrorApiType,
  getClientState,
  getCorsErrorProjectId,
  getCurrentUserState,
  getDashboardOrganizationId$1 as getDashboardOrganizationId,
  getDatasetsState,
  getDocumentState,
  getDocumentSyncStatus,
  getFavoritesState,
  getIndexForKey2 as getIndexForKey,
  getIsInDashboardState,
  getLoginUrlState,
  getNodeState,
  getOrCreateChannel,
  getOrCreateController,
  getOrCreateNode,
  getOrganizationState,
  getOrganizationsState,
  getPathDepth,
  getPermissionsState,
  getPerspectiveState,
  getPresence,
  getPreviewState,
  getProjectState,
  getProjectionState,
  getProjectsState,
  getQueryKey,
  getQueryState,
  getTokenState,
  getUserState,
  getUsersKey,
  getUsersState,
  handleAuthCallback,
  isCanvasResource,
  isCanvasSource,
  isDatasetResource,
  isDatasetSource,
  isImportError,
  isMediaLibraryResource,
  isMediaLibrarySource,
  isProjectUserNotFoundClientError,
  isStudioConfig,
  joinPaths,
  jsonMatch2 as jsonMatch,
  loadMoreUsers,
  logout,
  observeOrganizationVerificationState,
  parseQueryKey,
  parseUsersKey,
  publishDocument,
  releaseChannel,
  releaseNode,
  resolveDatasets,
  resolveDocument,
  resolveFavoritesState,
  resolveOrganization,
  resolveOrganizations,
  resolvePermissions,
  resolvePreview,
  resolveProject,
  resolveProjection,
  resolveProjects,
  resolveQuery,
  resolveUser,
  resolveUsers,
  setAuthToken,
  slicePath2 as slicePath,
  stringifyPath2 as stringifyPath,
  subscribeDocumentEvents,
  transformProjectionToPreview,
  unpublishDocument
};
