// @sanity/sdk/dist/_chunks-es/telemetryManager.js
import { createBatchedStore } from "@sanity/telemetry";
import { createLogger } from "./createGroqSearchFilter.js";
import { CORE_SDK_VERSION } from "./version.js";
import { SDKDevSessionEnded, SDKDevError, SDKHookMounted, SDKDevSessionStarted } from "./_internal.js";
const FLUSH_INTERVAL_MS = 3e4, CONSENT_TAG = "telemetry-consent.sdk", BATCH_TAG = "telemetry.batch", log = createLogger("telemetry");
function createTelemetryManager(options) {
  const { sessionId, getClient, projectId } = options, startedAt = Date.now(), emittedHooks = /* @__PURE__ */ new Set();
  let cachedConsent = null;
  const resolveConsent = async () => {
    if (cachedConsent) return cachedConsent;
    try {
      cachedConsent = await getClient().request({
        uri: "/intake/telemetry-status",
        tag: CONSENT_TAG
      });
    } catch {
      cachedConsent = { status: "undetermined" };
    }
    return cachedConsent;
  }, enrichBatch = (batch) => batch.map((event) => ({
    ...event,
    context: {
      version: CORE_SDK_VERSION,
      environment: "development",
      origin: typeof window < "u" ? window.location.origin : "node"
    }
  })), store = createBatchedStore(
    sessionId,
    {
      flushInterval: FLUSH_INTERVAL_MS,
      resolveConsent,
      sendEvents: async (batch) => {
        const client = getClient();
        return log.debug("sending event batch", { batchSize: batch.length }), client.request({
          uri: "/intake/batch",
          method: "POST",
          body: { projectId, batch: enrichBatch(batch) },
          tag: BATCH_TAG
        });
      }
    }
  ), logger = store.logger;
  return {
    async checkConsent() {
      const { status } = await resolveConsent();
      return status === "granted";
    },
    logSessionStarted(data) {
      log.debug("event: SDK Dev Session Started", {
        projectId: data.projectId,
        perspective: data.perspective,
        authMethod: data.authMethod,
        version: CORE_SDK_VERSION
      }), logger.log(SDKDevSessionStarted, {
        version: CORE_SDK_VERSION,
        ...data
      });
    },
    logHookFirstUsed(hookName) {
      emittedHooks.has(hookName) || (emittedHooks.add(hookName), log.debug("event: SDK Hook Mounted", { hookName }), logger.log(SDKHookMounted, { hookName }));
    },
    logDevError(errorType, hookName) {
      log.debug("event: SDK Dev Error", { errorType, hookName }), logger.log(SDKDevError, { errorType, hookName });
    },
    endSession() {
      const durationSeconds = Math.round((Date.now() - startedAt) / 1e3);
      log.debug("event: SDK Dev Session Ended", {
        durationSeconds,
        hooksUsed: [...emittedHooks]
      }), logger.log(SDKDevSessionEnded, {
        durationSeconds,
        hooksUsed: [...emittedHooks]
      }), store.flush().catch(() => {
      }), store.end();
    },
    dispose() {
      store.end();
    },
    get hooksUsed() {
      return emittedHooks;
    }
  };
}
export {
  createTelemetryManager
};
