// @sanity/runtime-cli/dist/cores/blueprints/doctor.js
import { readFileSync } from 'node:fs';
import { arch, cwd, env, version as nodeVersion, platform } from 'node:process';
import * as resolve from 'empathic/resolve';
import ora from 'ora';
import { loadAndParseBlueprint, } from '../../actions/blueprints/blueprint.js';
import { readConfigFile } from '../../actions/blueprints/config.js';
import { resolveIds } from '../../actions/blueprints/resolve.js';
import { getStack } from '../../actions/blueprints/stacks.js';
import config, { RUNTIME_CLI_VERSION } from '../../config.js';
import { formatDateTime } from '../../utils/display/dates.js';
import { check, filePathRelativeToCwd, niceId, renderSection, severe, unsure, } from '../../utils/display/presenters.js';
import { styleText } from '../../utils/style-text.js';
import { createTracedFetch } from '../../utils/traced-fetch.js';
import { validTokenOrErrorMessage } from '../../utils/validated-token.js';
import { blueprintConfigCore } from './config.js';
const diagLookup = {
    online: 'Host online',
    tokenValid: 'Authenticated',
    blueprintValid: 'Manifest valid',
    stackReady: 'Stack ready',
    userHasAccess: 'User has access',
};
function sourceLabel(source) {
    switch (source) {
        case 'env':
            return 'environment';
        case 'module':
            return 'manifest module';
        case 'config':
            return '.sanity/blueprint.config.json';
        case 'inferred':
            return 'inferred';
        default:
            return 'unknown';
    }
}
export async function blueprintDoctorCore(options) {
    const { bin, log, token, validateResources, flags: { verbose: v, path: p, fix }, } = options;
    const path = p || cwd();
    let tokenOrError;
    const diagnostics = {};
    for (const key in diagLookup) {
        diagnostics[key] = { status: null };
    }
    const envRows = [['Directory', p ? filePathRelativeToCwd(path) : path]];
    if (bin && bin !== 'sanity')
        envRows.push(['CLI', bin]);
    const configRows = [];
    const stackRows = [];
    const spinner = ora('Running diagnostics...').start();
    // --- ONLINE ---
    const fetchFn = createTracedFetch(log);
    try {
        const res = await fetchFn(config.apiUrl);
        if (res.ok) {
            diagnostics.online = { status: true };
        }
        else {
            diagnostics.online = { status: false, detail: `${res.status} ${res.statusText}` };
        }
    }
    catch (err) {
        const reason = err instanceof Error ? err.message : 'unknown error';
        diagnostics.online = { status: false, detail: reason };
    }
    // --- TOKEN ---
    if (token) {
        tokenOrError = await validTokenOrErrorMessage(log, token);
        if (tokenOrError.ok) {
            diagnostics.tokenValid = { status: true };
        }
        else {
            diagnostics.tokenValid = { status: false, detail: tokenOrError.error.message };
        }
    }
    else {
        diagnostics.tokenValid = { status: false, detail: 'no token found' };
    }
    // --- BLUEPRINT ---
    let localBlueprint;
    try {
        const blueprint = await loadAndParseBlueprint(path, {
            validateResources: options.validateResources || false,
        });
        const blueprintConfig = readConfigFile(blueprint.fileInfo.blueprintFilePath);
        const resolved = resolveIds({
            module: blueprint.module,
            config: blueprintConfig,
        });
        localBlueprint = {
            ...blueprint,
            blueprintConfig,
            ...resolved,
        };
        envRows.push(['Manifest', filePathRelativeToCwd(blueprint.fileInfo.blueprintFilePath)]);
        if (blueprint.errors.length === 0) {
            diagnostics.blueprintValid = { status: true };
        }
        else {
            diagnostics.blueprintValid = {
                status: false,
                detail: `${blueprint.errors.length} error(s)`,
            };
        }
    }
    catch {
        diagnostics.blueprintValid = { status: false, detail: 'unable to read file' };
    }
    if (env.SANITY_INTERNAL_ENV)
        envRows.push(['SANITY_INTERNAL_ENV', env.SANITY_INTERNAL_ENV]);
    envRows.push(['API URL', config.apiUrl]);
    if (env.POPULUS_API_URL)
        envRows.push(['POPULUS API URL', config.populusApiUrl]);
    envRows.push(['Runtime', `Node.js ${nodeVersion} (${platform}-${arch})`]);
    if (RUNTIME_CLI_VERSION) {
        envRows.push(['Internals', `v${RUNTIME_CLI_VERSION}`]);
    }
    const sanityCliPkgPath = resolve.from(path, '@sanity/cli/package.json', true);
    if (sanityCliPkgPath) {
        try {
            const sanityCliPkg = JSON.parse(readFileSync(sanityCliPkgPath, 'utf8'));
            envRows.push(['Sanity CLI', sanityCliPkg.version]);
        }
        catch { }
    }
    // --- CONFIGURATION ---
    if (localBlueprint) {
        const { scopeType, scopeId, stackId, sources, blueprintConfig } = localBlueprint;
        if (blueprintConfig && 'configPath' in blueprintConfig && blueprintConfig.configPath) {
            configRows.push(['Source', filePathRelativeToCwd(blueprintConfig.configPath)]);
        }
        else if (!blueprintConfig) {
            configRows.push(['Source', styleText('dim', 'no .sanity/blueprint.config.json')]);
        }
        if (blueprintConfig?.updatedAt) {
            configRows.push(['Updated', formatDateTime(blueprintConfig.updatedAt)]);
        }
        if (token && env.SANITY_AUTH_TOKEN) {
            configRows.push(['SANITY_AUTH_TOKEN', `***  ${styleText('dim', 'environment')}`]);
        }
        if (scopeType && scopeId) {
            const scopeSourceKey = scopeType === 'project' ? 'projectId' : 'organizationId';
            const label = scopeType === 'project' ? 'Project' : 'Organization';
            configRows.push([
                label,
                `${niceId(scopeId)}  ${styleText('dim', sourceLabel(sources?.[scopeSourceKey]))}`,
            ]);
        }
        if (stackId) {
            configRows.push([
                'Stack',
                `${niceId(stackId)}  ${styleText('dim', sourceLabel(sources?.stackId))}`,
            ]);
        }
        // --- STACK + ACCESS ---
        if (diagnostics.online.status &&
            diagnostics.tokenValid.status &&
            token &&
            stackId &&
            scopeType &&
            scopeId) {
            const stackResponse = await getStack({
                auth: { token, scopeType, scopeId },
                stackId,
                logger: log,
            });
            if (stackResponse.ok) {
                diagnostics.stackReady = { status: true };
                diagnostics.userHasAccess = { status: true };
                const stack = stackResponse.stack;
                if (stack) {
                    const label = stack.name ? `"${stack.name}" ${niceId(stackId)}` : niceId(stackId);
                    stackRows.push(['Stack', label]);
                    if (stack.recentOperation) {
                        const op = stack.recentOperation;
                        const time = op.completedAt || op.createdAt;
                        const timestamp = time ? `  ${styleText('dim', formatDateTime(time))}` : '';
                        stackRows.push(['Operation', `${op.status}${timestamp}`]);
                        if (op.userMessage)
                            stackRows.push(['  Message', `"${op.userMessage}"`]);
                    }
                    if (stack.resources) {
                        stackRows.push(['Resources', `${stack.resources.length}`]);
                    }
                }
            }
            else if (stackResponse.response?.status === 404) {
                diagnostics.stackReady = {
                    status: false,
                    detail: `Stack ${niceId(stackId)} not found (404)`,
                };
            }
            else if (stackResponse.response?.status === 403 || stackResponse.response?.status === 401) {
                diagnostics.userHasAccess = {
                    status: false,
                    detail: `no access to Stack ${niceId(stackId)} (${stackResponse.response.status})`,
                };
            }
            else {
                const statusSuffix = stackResponse.response?.status
                    ? ` (${stackResponse.response.status})`
                    : '';
                diagnostics.stackReady = {
                    status: null,
                    detail: (stackResponse.error || 'unknown error') + statusSuffix,
                };
                diagnostics.userHasAccess = { status: null, detail: `unknown error${statusSuffix}` };
            }
        }
        else if (stackId && scopeType && scopeId) {
            diagnostics.stackReady = { status: null, detail: 'requires online + authenticated' };
            diagnostics.userHasAccess = { status: null, detail: 'requires online + authenticated' };
        }
        else if (!stackId && !scopeType && !scopeId) {
            diagnostics.stackReady = { status: null, detail: 'missing configuration' };
            diagnostics.userHasAccess = { status: null, detail: 'missing configuration' };
        }
        else {
            const missing = [];
            if (!stackId)
                missing.push('Stack ID');
            if (!scopeType)
                missing.push('Scope Type');
            if (!scopeId) {
                const scopeLabel = scopeType === 'project'
                    ? 'Project ID'
                    : scopeType === 'organization'
                        ? 'Organization ID'
                        : 'Scope ID';
                missing.push(scopeLabel);
            }
            const detail = `missing ${missing.join(', ')}`;
            diagnostics.stackReady = { status: null, detail };
            diagnostics.userHasAccess = { status: null, detail };
        }
    }
    // --- RENDER REPORT ---
    spinner.stop();
    // Environment (verbose)
    log.verbose(renderSection('Environment', envRows));
    log.verbose('');
    // Configuration (verbose)
    if (configRows.length > 0) {
        log.verbose(renderSection('Configuration', configRows));
        log.verbose('');
    }
    // Deployment (verbose)
    if (stackRows.length > 0) {
        log.verbose(renderSection('Deployment', stackRows));
        log.verbose('');
    }
    // Checks (always visible)
    const maxLabel = Math.max(...Object.values(diagLookup).map((l) => l.length));
    log(styleText('bold', 'Checks'));
    let allGood = true;
    for (const [key, entry] of Object.entries(diagnostics)) {
        const label = diagLookup[key].padEnd(maxLabel);
        const detail = entry.detail ? `  ${styleText('dim', entry.detail)}` : '';
        switch (entry.status) {
            case true:
                log(`  ${check(label)}${detail}`);
                break;
            case false:
                allGood = false;
                log(`  ${severe(label)}${detail}`);
                break;
            case null:
                allGood = false;
                log(`  ${unsure(label)}${detail}`);
                break;
        }
    }
    // Result
    const flatDiagnostics = {};
    for (const [key, entry] of Object.entries(diagnostics)) {
        flatDiagnostics[key] = entry.status;
    }
    log('');
    const errorMessage = 'One or more checks failed';
    if (allGood) {
        log(styleText(['bold', 'green'], 'All checks passed'));
        if (fix)
            log(styleText(['bold', 'yellow'], 'Nothing to fix; --fix flag is ignored'));
        return {
            success: true,
            json: { diagnostics: flatDiagnostics },
            data: { diagnostics: flatDiagnostics },
        };
    }
    if (fix) {
        if (p) {
            return {
                success: false,
                error: `${errorMessage}. --fix cannot be used with --path`,
                data: { diagnostics: flatDiagnostics },
            };
        }
        if (!tokenOrError) {
            return {
                success: false,
                error: `${errorMessage}. Unable to fix: Missing authentication token`,
                suggestions: ['Run `npx @sanity/cli login` to authenticate, then try again.'],
                data: { diagnostics: flatDiagnostics },
            };
        }
        if (tokenOrError?.ok === false) {
            return {
                success: false,
                error: `${errorMessage}. Unable to fix: ${tokenOrError.error.message}`,
                suggestions: ['Run `npx @sanity/cli login` to re-authenticate.'],
                data: { diagnostics: flatDiagnostics },
            };
        }
        if (!localBlueprint) {
            return {
                success: false,
                error: `${errorMessage}. Unable to fix: Blueprint manifest is missing or invalid`,
                suggestions: [`Run \`npx ${bin} blueprints init\` to create a new Blueprint manifest.`],
                data: { diagnostics: flatDiagnostics },
            };
        }
        return blueprintConfigCore({
            bin,
            log,
            token: tokenOrError.value,
            blueprint: localBlueprint,
            validateResources,
            flags: { edit: true, verbose: v },
        });
    }
    return {
        success: false,
        error: errorMessage,
        suggestions: [`Run \`npx ${bin} blueprints doctor --fix\` to fix configuration issues.`],
        data: { diagnostics: flatDiagnostics },
    };
}
