// @sanity/runtime-cli/dist/cores/blueprints/mint-deploy-token.js
import { hostname } from 'node:os';
import { confirm, input, select } from '@inquirer/prompts';
import { createRobotToken } from '../../actions/sanity/access.js';
import { ClipboardUnavailableError, write as clipboardWrite } from '../../utils/clipboard.js';
import { formatDate, formatDateTime } from '../../utils/display/dates.js';
import { styleText } from '../../utils/style-text.js';
const ROLE_BY_SCOPE = {
    project: 'blueprints-deployer',
    organization: 'blueprints-deployer-robot',
};
// Matches Access API's "Role not found: <name>" 404 message.
export function isRoleUnavailableError(error) {
    return /^\s*role not found\b/i.test(error);
}
export function defaultLabel(scopeType) {
    const date = formatDate(new Date());
    let host = 'unknown-host';
    try {
        host = hostname();
    }
    catch { }
    return `blueprints-deployer @ ${host} ${date} (${scopeType})`;
}
export async function blueprintMintDeployTokenCore(options) {
    const { log, auth, scopeType, scopeId, flags } = options;
    const label = flags.label ?? defaultLabel(scopeType);
    const roleName = ROLE_BY_SCOPE[scopeType];
    const spinner = log.ora('Minting deploy token...').start();
    const { ok, error, robot } = await createRobotToken({
        auth,
        resourceType: scopeType,
        resourceId: scopeId,
        body: {
            label,
            memberships: [{ resourceType: scopeType, resourceId: scopeId, roleNames: [roleName] }],
        },
        logger: log,
    });
    if (!ok || !robot) {
        spinner.fail('Failed to mint deploy token');
        if (error && isRoleUnavailableError(error)) {
            return {
                success: false,
                error: `The "${roleName}" role is not available in this ${scopeType}.`,
                code: 'ROLE_UNAVAILABLE',
                suggestions: [],
            };
        }
        const suggestions = [];
        if (error && /access|permission|forbidden|role/i.test(error)) {
            suggestions.push(scopeType === 'organization'
                ? 'Verify you have organization admin access.'
                : 'Verify you have project admin access.');
        }
        return { success: false, error: error || 'Failed to mint deploy token', suggestions };
    }
    spinner.stop().clear();
    if (flags.print) {
        log(robot.token);
        return { success: true, json: serializeRobot(robot) };
    }
    if (flags.json) {
        return { success: true, json: serializeRobot(robot) };
    }
    log(styleText('green', `Minted "${robot.label}" (${robot.id})`));
    if (robot.expiresAt) {
        log(styleText('dim', `Expires: ${formatDateTime(robot.expiresAt)}`));
    }
    const action = await select({
        message: 'What would you like to do with the token?',
        choices: [
            { name: 'Copy to clipboard', value: 'copy' },
            { name: 'Print to terminal', value: 'print' },
            { name: 'Exit (token will be lost)', value: 'exit' },
        ],
        default: 'copy',
    });
    if (action === 'copy') {
        try {
            await clipboardWrite(robot.token);
            log(styleText('green', 'Token copied to clipboard.'));
        }
        catch (err) {
            const msg = err instanceof ClipboardUnavailableError
                ? err.message
                : err instanceof Error
                    ? err.message
                    : String(err);
            log.warn(styleText('yellow', `Could not copy to clipboard: ${msg}`));
            const printInstead = await confirm({
                message: 'Print the token to the terminal instead?',
                default: false,
            });
            if (printInstead) {
                await printAndConfirm(log, robot.token);
            }
            else {
                log(styleText('yellow', 'Token discarded. You can mint a new one anytime.'));
            }
        }
    }
    else if (action === 'print') {
        await printAndConfirm(log, robot.token);
    }
    else {
        log(styleText('yellow', 'Token discarded. You can mint a new one anytime.'));
    }
    return { success: true, json: serializeRobot(robot) };
}
/**
 * Show the token inside an inquirer prompt and rely on `clearPromptOnDone`
 * so inquirer wipes the entire rendered block (token included) on Enter.
 */
async function printAndConfirm(log, token) {
    await input({
        message: `\n${token}\n\nPress Enter once you have copied the token (it will be erased):`,
        default: '',
        theme: { prefix: '' },
    }, { clearPromptOnDone: true });
    log(styleText('dim', 'Token erased.'));
}
function serializeRobot(robot) {
    return robot;
}
