// @sanity/runtime-cli/dist/cores/blueprints/promote.js
import { confirm } from '@inquirer/prompts';
import { patchConfigFile } from '../../actions/blueprints/config.js';
import { promoteStack } from '../../actions/blueprints/stacks.js';
import { niceId } from '../../utils/display/presenters.js';
export async function blueprintPromoteCore(options) {
    const { bin = 'sanity', log, stackId, auth, flags, deployedStack, blueprint } = options;
    const newStackName = flags['new-stack-name'];
    if (newStackName && deployedStack.scopeType === 'organization') {
        return {
            success: false,
            error: '--new-stack-name can only be used when promoting a project-scoped Stack to organization scope.',
            suggestions: [
                `Stack "${deployedStack.name}" ${niceId(deployedStack.id)} is already org-scoped.`,
            ],
        };
    }
    let message = `"${deployedStack.name}" ${niceId(deployedStack.id)}`;
    if (deployedStack.scopeType === 'organization') {
        message = `Stack ${message} is already org-scoped. Promote again?`;
    }
    else if (newStackName) {
        message = `Promote Stack ${message} from project to organization scope and rename to "${newStackName}"? You will need admin access to your organization.`;
    }
    else {
        message = `Promote Stack ${message} from project to organization scope? You will need admin access to your organization.`;
    }
    if (!flags.force) {
        const confirmed = await confirm({ message });
        if (!confirmed) {
            return { success: false, error: 'Promotion cancelled by user' };
        }
    }
    try {
        const spinner = log.ora('Promoting Stack...').start();
        const { ok, error, stack, response } = await promoteStack({
            stackId,
            auth,
            logger: log,
            name: newStackName,
        });
        if (!ok) {
            spinner.fail('Failed to promote Stack');
            const suggestions = [];
            if (response?.status === 409) {
                suggestions.push('A Stack with this name already exists in the target organization. Use --new-stack-name <name> to rename while promoting.');
            }
            if (error && /access|permission|forbidden/i.test(error)) {
                suggestions.push('Verify you have organization admin access.');
            }
            suggestions.push(`Run \`npx ${bin} blueprints info\` to check the current Stack scope.`);
            return { success: false, error: error || 'Failed to promote Stack', suggestions };
        }
        spinner.stop().clear();
        log(`Stack "${stack.name}" ${niceId(stack.id)} promoted successfully`);
        const { blueprintFilePath } = blueprint.fileInfo;
        try {
            patchConfigFile(blueprintFilePath, {
                stackId: stack.id,
                organizationId: stack.scopeId,
                projectId: null,
            });
            log('Blueprint config updated (.sanity/blueprint.config.json)');
        }
        catch {
            return {
                success: false,
                error: 'Stack promoted, but could not update Blueprint config (.sanity/blueprint.config.json).',
                suggestions: [
                    `Run \`npx ${bin} blueprints config --edit\` to update your Blueprint config manually.`,
                ],
            };
        }
        return { success: true, json: { stack }, data: { stack } };
    }
    catch (error) {
        const errorMessage = error instanceof Error ? error.message : String(error);
        return { success: false, error: errorMessage };
    }
}
