// @sanity/runtime-cli/dist/cores/blueprints/stacks.js
import { listStacks } from '../../actions/blueprints/stacks.js';
import { listProjects } from '../../actions/sanity/projects.js';
import { formatStacksListing } from '../../utils/display/blueprints-formatting.js';
import { capitalize, niceId, warn } from '../../utils/display/presenters.js';
import { styleText } from '../../utils/style-text.js';
/** Max projects to query for stacks during --include-projects */
const PROJECT_LIMIT = 50;
export async function blueprintStacksCore(options) {
    const { log, token, blueprint, flags } = options;
    const { scopeType: blueprintScopeType, scopeId: blueprintScopeId, stackId: blueprintStackId, } = blueprint;
    const { 'project-id': flagProjectId, 'organization-id': flagOrganizationId, 'include-projects': includeProjects = false, verbose: _verbose = false, } = flags;
    if (flagOrganizationId && flagProjectId) {
        log.error('Cannot specify both --organization-id and --project-id');
        return { success: false, error: 'Cannot specify both --organization-id and --project-id' };
    }
    let scopeType = blueprintScopeType;
    let scopeId = blueprintScopeId;
    if (flagOrganizationId) {
        scopeType = 'organization';
        scopeId = flagOrganizationId;
    }
    if (flagProjectId) {
        scopeType = 'project';
        scopeId = flagProjectId;
    }
    if (!scopeType || !scopeId) {
        return {
            success: false,
            error: 'Unable to determine scope for Blueprint Stacks',
            suggestions: ['Run in a Blueprint directory or provide a project with --project-id.'],
        };
    }
    if (includeProjects && scopeType === 'organization') {
        return deepListStacks({ organizationId: scopeId, token, log, blueprintStackId });
    }
    try {
        const spinner = log.ora('Fetching Stacks...').start();
        const { ok, stacks, error } = await listStacks({ token, scopeType, scopeId }, log);
        if (!ok) {
            spinner.fail('Failed to list Stacks');
            const suggestions = [];
            if (error && /access|permission|forbidden/i.test(error)) {
                suggestions.push('Verify you have access to this organization or project.', `Try a different scope with --project-id or --organization-id.`);
            }
            return {
                success: false,
                error: error || 'Failed to list stacks',
                suggestions: suggestions.length ? suggestions : undefined,
            };
        }
        spinner.stop().clear();
        if (!stacks || stacks.length === 0) {
            log('No stacks found');
            return { success: true };
        }
        log(`${styleText('bold', capitalize(scopeType))} ${niceId(scopeId)} ${styleText('bold', 'Stacks')}:\n`);
        log(formatStacksListing(stacks, blueprintStackId));
        return { success: true, json: { stacks, scopeType, scopeId } };
    }
    catch (error) {
        const errorMessage = error instanceof Error ? error.message : String(error);
        log.error(`Error: ${errorMessage}`);
        return { success: false, error: errorMessage };
    }
}
async function deepListStacks(options) {
    const { organizationId, token, log, blueprintStackId } = options;
    const allStacks = [];
    const spinner = log.ora('Fetching Stacks...').start();
    try {
        // 1. Org-scoped stacks
        const orgResult = await listStacks({ token, scopeType: 'organization', scopeId: organizationId }, log);
        const orgStacks = orgResult.ok ? orgResult.stacks : [];
        // 2. Get projects in the organization
        const projectsResult = await listProjects({ token, organizationId, logger: log });
        if (!projectsResult.ok) {
            spinner.fail('Failed to list projects');
            return { success: false, error: projectsResult.error || 'Failed to list projects' };
        }
        const projects = projectsResult.projects;
        const totalProjects = projects.length;
        const cappedProjects = projects.slice(0, PROJECT_LIMIT);
        // 3. Fetch stacks for each project
        const projectGroups = [];
        for (let i = 0; i < cappedProjects.length; i++) {
            const project = cappedProjects[i];
            spinner.text = `Fetching Stacks... (project ${i + 1}/${cappedProjects.length})`;
            const result = await listStacks({ token, scopeType: 'project', scopeId: project.id }, log);
            if (result.ok && result.stacks.length > 0) {
                projectGroups.push({
                    projectId: project.id,
                    displayName: project.displayName,
                    stacks: result.stacks,
                });
                allStacks.push(...result.stacks);
            }
        }
        spinner.stop().clear();
        // 4. Print results
        if (orgStacks.length > 0) {
            log(`${styleText('bold', 'Organization')} ${niceId(organizationId)} ${styleText('bold', 'Stacks')}:\n`);
            log(formatStacksListing(orgStacks, blueprintStackId));
            if (projectGroups.length > 0)
                log(''); // blank line between sections
        }
        for (const group of projectGroups) {
            log(`${styleText('bold', 'Project')} "${group.displayName}" ${niceId(group.projectId)} ${styleText('bold', 'Stacks')}:`);
            log(formatStacksListing(group.stacks, blueprintStackId));
            log('');
        }
        if (totalProjects > PROJECT_LIMIT) {
            log(warn(`Organization has ${totalProjects} projects, only checked the first ${PROJECT_LIMIT}.`));
        }
        if (orgStacks.length === 0 && allStacks.length === 0) {
            log('No stacks found');
            return { success: true };
        }
        allStacks.push(...orgStacks);
        return {
            success: true,
            json: { stacks: allStacks, scopeType: 'organization', scopeId: organizationId },
        };
    }
    catch (error) {
        spinner.stop().clear();
        const errorMessage = error instanceof Error ? error.message : String(error);
        log.error(`Error: ${errorMessage}`);
        return { success: false, error: errorMessage };
    }
}
