// @sanity/runtime-cli/dist/baseCommands.js
// * https://oclif.io/docs/base_class
import { Command, CommandHelp, Flags as OclifFlags } from '@oclif/core';
import { findBlueprintFile, loadBlueprintFile, parseBlueprintContent, } from './actions/blueprints/blueprint.js';
import { backfillProjectBasedStackId, readConfigFile } from './actions/blueprints/config.js';
import { resolveIds } from './actions/blueprints/resolve.js';
import { getStack, resolveStackIdByNameOrId } from './actions/blueprints/stacks.js';
import { presentBlueprintParserErrors } from './utils/display/errors.js';
import { Logger } from './utils/logger.js';
import { validTokenOrErrorMessage } from './utils/validated-token.js';
/**
 * Fallback error-to-hint patterns for RuntimeCommand.catch().
 * Each entry maps a regex to a function that returns suggestion strings.
 * These only apply when the core did not set suggestions in the result.
 */
const CATCH_HINT_PATTERNS = [
// Add patterns here as needed, for example:
// [/scope/i, (bin) => [`Run \`npx ${bin} blueprints promote\` to change the Stack scope.`]],
];
export const pathFlagConfig = {
    description: 'Path to a Blueprint file or directory containing one',
    env: 'SANITY_BLUEPRINT_PATH',
    aliases: ['blueprint-path'],
    char: 'p',
};
export const baseFlags = {
    json: OclifFlags.boolean({
        // override defaults from oclif's built-in --json flag
        description: 'Format output as json',
    }),
    path: OclifFlags.string({
        ...pathFlagConfig,
        hidden: true,
    }),
    trace: OclifFlags.boolean({
        description: 'Trace output',
        default: false,
        hidden: true,
    }),
    'validate-resources': OclifFlags.boolean({
        description: 'Validate resources',
        default: true,
        allowNo: true,
        hidden: true,
    }),
    verbose: OclifFlags.boolean({
        description: 'Verbose output',
        default: false,
        hidden: true,
    }),
};
export const stackFlagConfig = {
    description: 'Stack name or ID to use instead of the locally configured Stack',
};
export const projectIdFlagConfig = {
    description: 'Sanity project ID used to scope Blueprint and Stack',
    aliases: ['project', 'projectId'],
    exclusive: ['organization-id'],
};
export const organizationIdFlagConfig = {
    description: 'Sanity organization ID used to scope Blueprint and Stack',
    aliases: ['organization', 'organizationId', 'org'],
    exclusive: ['project-id'],
};
/**
 * Guarantees flags and args.
 * Also centralizes baseFlags and enables oclif's built-in --json for all subclasses.
 * @extends Command
 */
export class RuntimeCommand extends Command {
    flags;
    args;
    static baseFlags = baseFlags;
    static enableJsonFlag = true; // oclif's built-in --json flag
    /**
     * Generates help text for this command class
     * @param bin - The bin name to use for the help text; e.g. 'sanity', defaults to 'sanity-run'
     * @param commandId - The sub-command to use for the help text; optional but recommended for better help text
     * @returns The oclif-styled help text for the command
     * @example const helpText = InfoCommand.getHelpText('sanity', 'blueprints info')
     */
    static getHelpText(bin = 'sanity-run', commandId) {
        // biome-ignore lint/complexity/noThisInStatic: this is this command class - not confusing
        const thisClass = this;
        const rawArgs = thisClass.args || {};
        const args = Object.entries(rawArgs).reduce((acc, [name, arg]) => {
            acc[name] = { ...arg, name };
            return acc;
        }, {});
        const commandLoadable = {
            //! required properties
            id: commandId || thisClass.id || '',
            flags: thisClass.flags || {},
            args,
            aliases: thisClass.aliases || [],
            hidden: thisClass.hidden || false,
            hiddenAliases: thisClass.hiddenAliases || [],
            load: async () => thisClass,
            //* optional properties
            // bin, // seems to not be used in the CommandHelp class
            // thisClass, // doesn't change the help text
            help: thisClass.help,
            usage: thisClass.usage,
            examples: thisClass.examples,
            description: thisClass.description,
            summary: thisClass.summary,
            state: thisClass.state,
            deprecateAliases: thisClass.deprecateAliases,
            deprecationOptions: thisClass.deprecationOptions,
            strict: thisClass.strict,
            hasDynamicHelp: thisClass.hasDynamicHelp,
        };
        const commandHelp = new CommandHelp(commandLoadable, { bin }, {
            maxWidth: 120,
        });
        return commandHelp.generate();
    }
    async init() {
        const { args, flags } = await this.parse({
            flags: this.ctor.flags,
            baseFlags: super.ctor.baseFlags,
            enableJsonFlag: this.ctor.enableJsonFlag,
            args: this.ctor.args,
            strict: this.ctor.strict,
        });
        this.flags = flags;
        this.args = args;
        await super.init();
    }
    /**
     * Throws an oclif error from a failed CoreResult.
     * forwarding suggestions, ref, or code to oclif's PrettyPrintableError rendering.
     */
    coreError(result) {
        this.error(result.error, {
            suggestions: result.suggestions,
            ref: result.ref,
            code: result.code,
        });
    }
    /**
     * If an error doesn't have suggestions, match against known patterns to
     * attach relevant suggestions before oclif renders the error.
     *
     * Add patterns to `CATCH_HINT_PATTERNS` as needed.
     */
    async catch(err) {
        const pErr = err;
        if (!pErr.suggestions?.length) {
            for (const [pattern, hints] of CATCH_HINT_PATTERNS) {
                if (pattern.test(err.message)) {
                    pErr.suggestions = hints(this.config.bin);
                    break;
                }
            }
        }
        return super.catch(err);
    }
}
/**
 * Base command that resolves context declaratively from a `static needs` array.
 * Commands declare what they *directly use*; the base class resolves transitive
 * dependencies automatically (e.g. 'deployedStack' implies token, scope, stackId).
 *
 * When a command needs only remote context (e.g. `['deployedStack']`),
 * no local blueprint file is required -- flags and config are sufficient.
 *
 * @example
 * ```ts
 * class InfoCommand extends ResolvedCommand<typeof InfoCommand> {
 *   static needs = ['deployedStack'] as const
 *   // token, scope, stackId resolved automatically
 * }
 * ```
 *
 * @extends RuntimeCommand
 */
export class ResolvedCommand extends RuntimeCommand {
    static needs = [];
    static baseFlags = {
        ...baseFlags,
        stack: OclifFlags.string({ ...stackFlagConfig, hidden: true }),
        'project-id': OclifFlags.string({ ...projectIdFlagConfig, hidden: true }),
        'organization-id': OclifFlags.string({ ...organizationIdFlagConfig, hidden: true }),
    };
    // Populated by init() according to `needs`
    sanityToken;
    blueprint;
    scopeType;
    scopeId;
    stackId;
    auth;
    deployedStack;
    async init() {
        await super.init(); // RuntimeCommand: parse flags + args
        const needs = new Set(this.constructor.needs);
        if (needs.size === 0)
            return;
        const log = Logger(this.log.bind(this), this.flags);
        const needsToken = needs.has('token') || needs.has('scope') || needs.has('stackId') || needs.has('deployedStack');
        const needsScope = needs.has('scope') || needs.has('deployedStack');
        const needsStack = needs.has('stackId') || needs.has('deployedStack');
        const needsBlueprint = needs.has('blueprint');
        const needsDeployedStack = needs.has('deployedStack');
        // 1. Token
        if (needsToken) {
            const check = await validTokenOrErrorMessage(log);
            if (!check.ok) {
                this.error(check.error.message, {
                    suggestions: ['Run `npx @sanity/cli login` to authenticate.'],
                });
            }
            this.sanityToken = check.value;
        }
        // 2. Read config file (cheap single-file JSON read).
        // Anchor to --path flag or the blueprint file location if found, else cwd.
        const blueprintFileInfo = findBlueprintFile(this.flags.path);
        const blueprintConfig = readConfigFile(blueprintFileInfo?.blueprintFilePath);
        // 3. Blueprint (if needed)
        let blueprintModule;
        if (needsBlueprint) {
            if (!blueprintFileInfo) {
                this.error('Could not find a Blueprint manifest (sanity.blueprint.ts, .js, or .json).', {
                    suggestions: [
                        `Run \`${this.config.bin} blueprints init\` to create one.`,
                        `Run \`${this.config.bin} blueprints doctor\` to check your configuration.`,
                    ],
                });
            }
            const loaded = await loadBlueprintFile(blueprintFileInfo);
            const parsed = parseBlueprintContent(loaded.rawBlueprint, {
                validateResources: this.flags['validate-resources'],
            });
            if (parsed.errors.length > 0) {
                log(presentBlueprintParserErrors(parsed.errors));
                this.error('Blueprint manifest contains errors.', {
                    suggestions: [
                        'Fix the manifest errors listed above.',
                        `Run \`${this.config.bin} blueprints doctor\` to check your configuration.`,
                    ],
                });
            }
            blueprintModule = loaded.module;
            // Assemble the full ReadBlueprintResult -- will be completed with resolved IDs below
            this.blueprint = {
                fileInfo: blueprintFileInfo,
                blueprintConfig,
                rawBlueprint: loaded.rawBlueprint,
                ...parsed,
            };
        }
        // 4. Resolve IDs
        const resolved = resolveIds({
            flags: {
                projectId: this.flags['project-id'],
                organizationId: this.flags['organization-id'],
            },
            module: blueprintModule,
            config: blueprintConfig,
        });
        // Legacy stack ID inference (only when we have a blueprint file path for the write side-effect)
        if (!resolved.stackId && resolved.projectId && blueprintFileInfo) {
            try {
                const inferred = await backfillProjectBasedStackId({
                    blueprintFilePath: blueprintFileInfo.blueprintFilePath,
                    projectId: resolved.projectId,
                    logger: log,
                });
                if (inferred) {
                    resolved.stackId = inferred;
                    resolved.sources.stackId = 'inferred';
                }
            }
            catch {
                // assumption was wrong; leave stackId undefined
            }
        }
        // Backfill resolved IDs onto the blueprint result if it was loaded
        if (this.blueprint) {
            Object.assign(this.blueprint, {
                organizationId: resolved.organizationId,
                projectId: resolved.projectId,
                stackId: resolved.stackId,
                scopeType: resolved.scopeType,
                scopeId: resolved.scopeId,
                sources: resolved.sources,
            });
        }
        // 5. Scope
        if (needsScope) {
            if (!resolved.scopeType || !resolved.scopeId) {
                this.error('Missing scope: provide --project-id or --organization-id, or run `blueprints init` to create a Blueprint config (.sanity/blueprint.config.json).', {
                    suggestions: [
                        `Run \`${this.config.bin} blueprints doctor\` to check your configuration.`,
                    ],
                });
            }
            this.scopeType = resolved.scopeType;
            this.scopeId = resolved.scopeId;
        }
        // 6. Stack
        if (needsStack) {
            let { stackId } = resolved;
            const stackOverride = this.flags.stack;
            if (stackOverride) {
                const auth = {
                    token: this.sanityToken,
                    scopeType: this.scopeType,
                    scopeId: this.scopeId,
                };
                stackId = await resolveStackIdByNameOrId(stackOverride, auth, log);
            }
            if (!stackId) {
                this.error(`Missing Stack: provide --stack, or set a Stack in your Blueprint config (\`${this.config.bin} blueprints config --edit\`).`, {
                    suggestions: [
                        `Run \`${this.config.bin} blueprints doctor\` to check your configuration.`,
                    ],
                });
            }
            this.stackId = stackId;
        }
        // 7. Auth
        if (needsScope) {
            this.auth = { token: this.sanityToken, scopeType: this.scopeType, scopeId: this.scopeId };
        }
        // 8. Deployed Stack
        if (needsDeployedStack) {
            const spinner = log.ora('Loading Stack deployment...').start();
            const response = await getStack({ stackId: this.stackId, auth: this.auth, logger: log });
            if (!response.ok) {
                spinner.fail('Could not load Stack deployment');
                this.error('Missing Stack deployment', {
                    suggestions: [
                        `Run \`${this.config.bin} blueprints doctor\` to check your configuration.`,
                    ],
                });
            }
            spinner.stop().clear();
            this.deployedStack = response.stack;
        }
    }
}
