// @sanity/runtime-cli/dist/constants.js
export const SANITY_ACCESS_ROLE = 'sanity.access.role';
export const SANITY_ACCESS_ROBOT = 'sanity.access.robot';
export const SANITY_FUNCTION_PREFIX = 'sanity.function.';
export const SANITY_FUNCTION_DOCUMENT = 'sanity.function.document';
export const SANITY_FUNCTION_MEDIA_LIBRARY_ASSET = 'sanity.function.media-library.asset';
export const SANITY_FUNCTION_SCHEDULED = 'sanity.function.cron';
export const SANITY_FUNCTION_SYNC_TAG_INVALIDATE = 'sanity.function.sync-tag-invalidate';
export const SANITY_PROJECT = 'sanity.project';
export const SANITY_PROJECT_CORS = 'sanity.project.cors';
export const SANITY_PROJECT_DATASET = 'sanity.project.dataset';
export const SANITY_PROJECT_WEBHOOK = 'sanity.project.webhook';
export const EVENT_DOCUMENT_PUBLISH = 'document-publish';
export const EVENT_DOCUMENT_CREATE = 'document-create';
export const EVENT_DOCUMENT_UPDATE = 'document-update';
export const EVENT_DOCUMENT_DELETE = 'document-delete';
export const EVENT_MEDIA_LIBRARY_ASSET_CREATE = 'media-library-asset-create';
export const EVENT_MEDIA_LIBRARY_ASSET_UPDATE = 'media-library-asset-update';
export const EVENT_MEDIA_LIBRARY_ASSET_DELETE = 'media-library-asset-delete';
export const EVENT_SCHEDULED = 'scheduled-function';
export const EVENT_SYNC_TAG_INVALIDATE = 'sync-tag-invalidate';
export const MAP_EVENT_TO_FUNCTION_TYPE = {
    [EVENT_DOCUMENT_PUBLISH]: SANITY_FUNCTION_DOCUMENT,
    [EVENT_DOCUMENT_CREATE]: SANITY_FUNCTION_DOCUMENT,
    [EVENT_DOCUMENT_DELETE]: SANITY_FUNCTION_DOCUMENT,
    [EVENT_DOCUMENT_UPDATE]: SANITY_FUNCTION_DOCUMENT,
    [EVENT_MEDIA_LIBRARY_ASSET_CREATE]: SANITY_FUNCTION_MEDIA_LIBRARY_ASSET,
    [EVENT_MEDIA_LIBRARY_ASSET_DELETE]: SANITY_FUNCTION_MEDIA_LIBRARY_ASSET,
    [EVENT_MEDIA_LIBRARY_ASSET_UPDATE]: SANITY_FUNCTION_MEDIA_LIBRARY_ASSET,
    [EVENT_SCHEDULED]: SANITY_FUNCTION_SCHEDULED,
    [EVENT_SYNC_TAG_INVALIDATE]: SANITY_FUNCTION_SYNC_TAG_INVALIDATE,
};
export const FUNCTION_TYPES = Object.keys(MAP_EVENT_TO_FUNCTION_TYPE);
export const PROJECT_SCOPED_FUNCTION_TYPES = new Set([
    SANITY_FUNCTION_DOCUMENT,
    SANITY_FUNCTION_MEDIA_LIBRARY_ASSET,
    SANITY_FUNCTION_SYNC_TAG_INVALIDATE,
]);
export const ORGANIZATION_SCOPED_FUNCTION_TYPES = new Set([
    SANITY_FUNCTION_SCHEDULED,
]);
export const MAX_ASSET_SIZE = 209_715_200; // 200 MB in bytes
export const CONVERT_BYTES_TO_MB = 1_048_576; // Used to convert bytes to megabytes (1024 * 1024)
export const LABEL_DOCUMENT_CREATE = 'Document Create';
export const LABEL_DOCUMENT_UPDATE = 'Document Update';
export const LABEL_DOCUMENT_DELETE = 'Document Delete';
export const LABEL_MEDIA_LIBRARY_ASSET_CREATE = 'Media Library Asset Create';
export const LABEL_MEDIA_LIBRARY_ASSET_UPDATE = 'Media Library Asset Update';
export const LABEL_MEDIA_LIBRARY_ASSET_DELETE = 'Media Library Asset Delete';
export const LABEL_SCHEDULED = 'Scheduled';
export const LABEL_SYNC_TAG_INVALIDATE = 'Sync Tag Invalidate';
